{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "import joblib\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['text', 'segments', 'language'])\n",
      "[{'id': 0, 'seek': 0, 'start': 0.0, 'end': 1.0, 'text': ' Thank you all.', 'tokens': [50363, 6952, 345, 477, 13, 50413], 'temperature': 0.0, 'avg_logprob': -0.45480142982260696, 'compression_ratio': 1.611353711790393, 'no_speech_prob': 0.2748943865299225}, {'id': 1, 'seek': 0, 'start': 1.0, 'end': 2.0, 'text': ' Good afternoon.', 'tokens': [50413, 4599, 6672, 13, 50463], 'temperature': 0.0, 'avg_logprob': -0.45480142982260696, 'compression_ratio': 1.611353711790393, 'no_speech_prob': 0.2748943865299225}, {'id': 2, 'seek': 0, 'start': 2.0, 'end': 3.0, 'text': ' How are you guys doing?', 'tokens': [50463, 1374, 389, 345, 3730, 1804, 30, 50513], 'temperature': 0.0, 'avg_logprob': -0.45480142982260696, 'compression_ratio': 1.611353711790393, 'no_speech_prob': 0.2748943865299225}, {'id': 3, 'seek': 0, 'start': 3.0, 'end': 4.0, 'text': ' It looks fun.', 'tokens': [50513, 632, 3073, 1257, 13, 50563], 'temperature': 0.0, 'avg_logprob': -0.45480142982260696, 'compression_ratio': 1.611353711790393, 'no_speech_prob': 0.2748943865299225}, {'id': 4, 'seek': 0, 'start': 4.0, 'end': 7.0, 'text': ' How many fans have done themselves last week?', 'tokens': [50563, 1374, 867, 3296, 423, 1760, 2405, 938, 1285, 30, 50713], 'temperature': 0.0, 'avg_logprob': -0.45480142982260696, 'compression_ratio': 1.611353711790393, 'no_speech_prob': 0.2748943865299225}, {'id': 5, 'seek': 0, 'start': 7.0, 'end': 8.0, 'text': ' Too easy?', 'tokens': [50713, 14190, 2562, 30, 50763], 'temperature': 0.0, 'avg_logprob': -0.45480142982260696, 'compression_ratio': 1.611353711790393, 'no_speech_prob': 0.2748943865299225}, {'id': 6, 'seek': 0, 'start': 8.0, 'end': 16.0, 'text': ' At least, maybe you think, well, applied the concepts that we learned in the past.', 'tokens': [50763, 1629, 1551, 11, 3863, 345, 892, 11, 880, 11, 5625, 262, 10838, 326, 356, 4499, 287, 262, 1613, 13, 51163], 'temperature': 0.0, 'avg_logprob': -0.45480142982260696, 'compression_ratio': 1.611353711790393, 'no_speech_prob': 0.2748943865299225}, {'id': 7, 'seek': 0, 'start': 16.0, 'end': 21.0, 'text': ' I hope that we all have relevant problems, for modern students like you guys.', 'tokens': [51163, 314, 2911, 326, 356, 477, 423, 5981, 2761, 11, 329, 3660, 2444, 588, 345, 3730, 13, 51413], 'temperature': 0.0, 'avg_logprob': -0.45480142982260696, 'compression_ratio': 1.611353711790393, 'no_speech_prob': 0.2748943865299225}, {'id': 8, 'seek': 0, 'start': 21.0, 'end': 27.0, 'text': ' At least, I hope you had some fun from doing it and not finding it too challenging.', 'tokens': [51413, 1629, 1551, 11, 314, 2911, 345, 550, 617, 1257, 422, 1804, 340, 290, 407, 4917, 340, 1165, 9389, 13, 51713], 'temperature': 0.0, 'avg_logprob': -0.45480142982260696, 'compression_ratio': 1.611353711790393, 'no_speech_prob': 0.2748943865299225}, {'id': 9, 'seek': 2700, 'start': 27.0, 'end': 36.0, 'text': ' The final exam questions, they will be largely in the context of problem solving like this.', 'tokens': [50363, 383, 2457, 2814, 2683, 11, 484, 481, 307, 5688, 287, 262, 4732, 286, 1917, 18120, 588, 428, 13, 50813], 'temperature': 0.0, 'avg_logprob': -0.28589046833127046, 'compression_ratio': 1.7281553398058251, 'no_speech_prob': 0.6804109215736389}, {'id': 10, 'seek': 2700, 'start': 36.0, 'end': 40.0, 'text': ' And the level of challenge will be suddenly a little bit less, perhaps,', 'tokens': [50813, 843, 262, 1241, 286, 4427, 481, 307, 6451, 257, 1310, 1643, 1342, 11, 3737, 11, 51013], 'temperature': 0.0, 'avg_logprob': -0.28589046833127046, 'compression_ratio': 1.7281553398058251, 'no_speech_prob': 0.6804109215736389}, {'id': 11, 'seek': 2700, 'start': 40.0, 'end': 43.0, 'text': ' because in the final exam you are kind of limited on it.', 'tokens': [51013, 780, 287, 262, 2457, 2814, 345, 389, 1611, 286, 3614, 319, 340, 13, 51163], 'temperature': 0.0, 'avg_logprob': -0.28589046833127046, 'compression_ratio': 1.7281553398058251, 'no_speech_prob': 0.6804109215736389}, {'id': 12, 'seek': 2700, 'start': 43.0, 'end': 47.0, 'text': ' So the highest level of challenge you will see, and at least my point in the course,', 'tokens': [51163, 1406, 262, 4511, 1241, 286, 4427, 345, 481, 766, 11, 290, 379, 1551, 616, 966, 287, 262, 1781, 11, 51363], 'temperature': 0.0, 'avg_logprob': -0.28589046833127046, 'compression_ratio': 1.7281553398058251, 'no_speech_prob': 0.6804109215736389}, {'id': 13, 'seek': 2700, 'start': 47.0, 'end': 53.0, 'text': ' will be represented by the problem sets questions.', 'tokens': [51363, 481, 307, 7997, 416, 262, 1917, 5621, 2683, 13, 51663], 'temperature': 0.0, 'avg_logprob': -0.28589046833127046, 'compression_ratio': 1.7281553398058251, 'no_speech_prob': 0.6804109215736389}, {'id': 14, 'seek': 5300, 'start': 53.0, 'end': 59.0, 'text': ' So if you feel that we understand them and you can tackle them, especially after the session,', 'tokens': [50363, 1406, 611, 345, 1254, 326, 356, 1833, 606, 290, 345, 460, 9761, 606, 11, 2592, 706, 262, 6246, 11, 50663], 'temperature': 0.0, 'avg_logprob': -0.2227006477884727, 'compression_ratio': 1.592057761732852, 'no_speech_prob': 0.33141064643859863}, {'id': 15, 'seek': 5300, 'start': 59.0, 'end': 64.0, 'text': ' and you are engaged in the class and you really should not need to worry suddenly from the', 'tokens': [50663, 290, 345, 389, 7953, 287, 262, 1398, 290, 345, 1107, 815, 407, 761, 284, 5490, 6451, 422, 262, 50913], 'temperature': 0.0, 'avg_logprob': -0.2227006477884727, 'compression_ratio': 1.592057761732852, 'no_speech_prob': 0.33141064643859863}, {'id': 16, 'seek': 5300, 'start': 64.0, 'end': 69.0, 'text': \" three lecture series that I'm very happy to provide here for this course.\", 'tokens': [50913, 1115, 19143, 2168, 326, 314, 1101, 845, 3772, 284, 2148, 994, 329, 428, 1781, 13, 51163], 'temperature': 0.0, 'avg_logprob': -0.2227006477884727, 'compression_ratio': 1.592057761732852, 'no_speech_prob': 0.33141064643859863}, {'id': 17, 'seek': 5300, 'start': 69.0, 'end': 72.0, 'text': \" So let's continue to today's topic too.\", 'tokens': [51163, 1406, 1309, 338, 2555, 284, 1909, 338, 7243, 1165, 13, 51313], 'temperature': 0.0, 'avg_logprob': -0.2227006477884727, 'compression_ratio': 1.592057761732852, 'no_speech_prob': 0.33141064643859863}, {'id': 18, 'seek': 5300, 'start': 72.0, 'end': 76.0, 'text': \" We're going to learn about designing and applying these RNA switches.\", 'tokens': [51313, 775, 821, 1016, 284, 2193, 546, 18492, 290, 11524, 777, 25897, 18225, 13, 51513], 'temperature': 0.0, 'avg_logprob': -0.2227006477884727, 'compression_ratio': 1.592057761732852, 'no_speech_prob': 0.33141064643859863}, {'id': 19, 'seek': 5300, 'start': 76.0, 'end': 80.0, 'text': ' You probably know synthetic biologists are very much inspired by nature,', 'tokens': [51513, 921, 2192, 760, 18512, 43149, 389, 845, 881, 7867, 416, 3450, 11, 51713], 'temperature': 0.0, 'avg_logprob': -0.2227006477884727, 'compression_ratio': 1.592057761732852, 'no_speech_prob': 0.33141064643859863}, {'id': 20, 'seek': 8000, 'start': 80.0, 'end': 86.0, 'text': ' by how nature designs these circuits and controls cell decision making and cell responses,', 'tokens': [50363, 416, 703, 3450, 9824, 777, 24907, 290, 6973, 2685, 2551, 1642, 290, 2685, 9109, 11, 50663], 'temperature': 0.0, 'avg_logprob': -0.2217906399777061, 'compression_ratio': 1.5270935960591132, 'no_speech_prob': 0.17531506717205048}, {'id': 21, 'seek': 8000, 'start': 86.0, 'end': 88.0, 'text': ' all the way at the animal level.', 'tokens': [50663, 477, 262, 835, 379, 262, 5044, 1241, 13, 50763], 'temperature': 0.0, 'avg_logprob': -0.2217906399777061, 'compression_ratio': 1.5270935960591132, 'no_speech_prob': 0.17531506717205048}, {'id': 22, 'seek': 8000, 'start': 88.0, 'end': 96.0, 'text': ' And when we use these ideas, we really be able to control back cells and living culture models.', 'tokens': [50763, 843, 618, 356, 779, 777, 4213, 11, 356, 1107, 307, 1498, 284, 1630, 736, 4778, 290, 2877, 3968, 4981, 13, 51163], 'temperature': 0.0, 'avg_logprob': -0.2217906399777061, 'compression_ratio': 1.5270935960591132, 'no_speech_prob': 0.17531506717205048}, {'id': 23, 'seek': 8000, 'start': 96.0, 'end': 99.0, 'text': ' So we will focus on RNA level switches.', 'tokens': [51163, 1406, 356, 481, 2962, 319, 25897, 1241, 18225, 13, 51313], 'temperature': 0.0, 'avg_logprob': -0.2217906399777061, 'compression_ratio': 1.5270935960591132, 'no_speech_prob': 0.17531506717205048}, {'id': 24, 'seek': 8000, 'start': 99.0, 'end': 104.0, 'text': ' Last week it was therapeutics, as you guys recall.', 'tokens': [51313, 4586, 1285, 340, 373, 20094, 48063, 11, 355, 345, 3730, 10014, 13, 51563], 'temperature': 0.0, 'avg_logprob': -0.2217906399777061, 'compression_ratio': 1.5270935960591132, 'no_speech_prob': 0.17531506717205048}, {'id': 25, 'seek': 10400, 'start': 105.0, 'end': 111.0, 'text': ' So the key aspects that we will target will be first going through the principles of natural regulatory', 'tokens': [50413, 1406, 262, 1994, 7612, 326, 356, 481, 2496, 481, 307, 717, 1016, 832, 262, 7811, 286, 3288, 11344, 50713], 'temperature': 0.0, 'avg_logprob': -0.13671333010834041, 'compression_ratio': 1.6666666666666667, 'no_speech_prob': 0.0655023604631424}, {'id': 26, 'seek': 10400, 'start': 111.0, 'end': 117.0, 'text': \" RNA. That's kind of important fundamentals because we need to know how nature does these mechanisms,\", 'tokens': [50713, 25897, 13, 1320, 338, 1611, 286, 1593, 33099, 780, 356, 761, 284, 760, 703, 3450, 857, 777, 11701, 11, 51013], 'temperature': 0.0, 'avg_logprob': -0.13671333010834041, 'compression_ratio': 1.6666666666666667, 'no_speech_prob': 0.0655023604631424}, {'id': 27, 'seek': 10400, 'start': 117.0, 'end': 122.0, 'text': ' and how we can hijack these principles to really control the cell decision making.', 'tokens': [51013, 290, 703, 356, 460, 16836, 441, 777, 7811, 284, 1107, 1630, 262, 2685, 2551, 1642, 13, 51263], 'temperature': 0.0, 'avg_logprob': -0.13671333010834041, 'compression_ratio': 1.6666666666666667, 'no_speech_prob': 0.0655023604631424}, {'id': 28, 'seek': 10400, 'start': 122.0, 'end': 127.0, 'text': ' And there will be three courses that we will concern with for our course level.', 'tokens': [51263, 843, 612, 481, 307, 1115, 10902, 326, 356, 481, 2328, 351, 329, 674, 1781, 1241, 13, 51513], 'temperature': 0.0, 'avg_logprob': -0.13671333010834041, 'compression_ratio': 1.6666666666666667, 'no_speech_prob': 0.0655023604631424}, {'id': 29, 'seek': 10400, 'start': 127.0, 'end': 132.0, 'text': ' This will be regulators, ribozymes, RNA-based enzymes, and riboswitches.', 'tokens': [51513, 770, 481, 307, 17199, 11, 12183, 8590, 22009, 11, 25897, 12, 3106, 29120, 11, 290, 12183, 418, 86, 9249, 13, 51763], 'temperature': 0.0, 'avg_logprob': -0.13671333010834041, 'compression_ratio': 1.6666666666666667, 'no_speech_prob': 0.0655023604631424}, {'id': 30, 'seek': 13200, 'start': 132.0, 'end': 139.0, 'text': ' And lastly, connected to the second topic, it will be small molecule ligand or drug-guided RNA conformational gating', 'tokens': [50363, 843, 938, 306, 11, 5884, 284, 262, 1218, 7243, 11, 340, 481, 307, 1402, 27756, 26106, 392, 393, 2563, 12, 23657, 25897, 17216, 864, 308, 803, 50713], 'temperature': 0.0, 'avg_logprob': -0.13429405319858606, 'compression_ratio': 1.519047619047619, 'no_speech_prob': 0.006292225327342749}, {'id': 31, 'seek': 13200, 'start': 139.0, 'end': 146.0, 'text': ' that will allow you to achieve this reversible regulatory control at the level of these RNA switches.', 'tokens': [50713, 326, 481, 1249, 345, 284, 4620, 428, 48287, 11344, 1630, 379, 262, 1241, 286, 777, 25897, 18225, 13, 51063], 'temperature': 0.0, 'avg_logprob': -0.13429405319858606, 'compression_ratio': 1.519047619047619, 'no_speech_prob': 0.006292225327342749}, {'id': 32, 'seek': 13200, 'start': 148.0, 'end': 156.0, 'text': \" And first, to cover everyone's kind of basic fundamental principles on ribosome protein translation.\", 'tokens': [51163, 843, 717, 11, 284, 3002, 2506, 338, 1611, 286, 4096, 7531, 7811, 319, 12183, 418, 462, 7532, 11059, 13, 51563], 'temperature': 0.0, 'avg_logprob': -0.13429405319858606, 'compression_ratio': 1.519047619047619, 'no_speech_prob': 0.006292225327342749}, {'id': 33, 'seek': 15600, 'start': 156.0, 'end': 163.0, 'text': ' Pretty much last week, we talked about this central dogma, right, starting from transcription and translation,', 'tokens': [50363, 20090, 881, 938, 1285, 11, 356, 6619, 546, 428, 4318, 40937, 11, 826, 11, 3599, 422, 26955, 290, 11059, 11, 50713], 'temperature': 0.0, 'avg_logprob': -0.16720433351470204, 'compression_ratio': 1.6724137931034482, 'no_speech_prob': 0.05509773641824722}, {'id': 34, 'seek': 15600, 'start': 163.0, 'end': 168.0, 'text': \" and of course there's also this reverse transcription method that's happened in nature in certain contexts.\", 'tokens': [50713, 290, 286, 1781, 612, 338, 635, 428, 9575, 26955, 2446, 326, 338, 3022, 287, 3450, 287, 1728, 26307, 13, 50963], 'temperature': 0.0, 'avg_logprob': -0.16720433351470204, 'compression_ratio': 1.6724137931034482, 'no_speech_prob': 0.05509773641824722}, {'id': 35, 'seek': 15600, 'start': 168.0, 'end': 174.0, 'text': ' But what about this second step to the I mentioned here, protein translation.', 'tokens': [50963, 887, 644, 546, 428, 1218, 2239, 284, 262, 314, 4750, 994, 11, 7532, 11059, 13, 51263], 'temperature': 0.0, 'avg_logprob': -0.16720433351470204, 'compression_ratio': 1.6724137931034482, 'no_speech_prob': 0.05509773641824722}, {'id': 36, 'seek': 15600, 'start': 174.0, 'end': 183.0, 'text': ' You guys know what machinery in any organism that does this translation, what is it called?', 'tokens': [51263, 921, 3730, 760, 644, 20230, 287, 597, 26433, 326, 857, 428, 11059, 11, 644, 318, 340, 1444, 30, 51713], 'temperature': 0.0, 'avg_logprob': -0.16720433351470204, 'compression_ratio': 1.6724137931034482, 'no_speech_prob': 0.05509773641824722}, {'id': 37, 'seek': 18300, 'start': 183.0, 'end': 190.0, 'text': \" Machinery, biomolecular machine that does this translation. It's actually known as ribosome.\", 'tokens': [50363, 7080, 15451, 11, 27488, 2305, 10440, 4572, 326, 857, 428, 11059, 13, 632, 338, 1682, 1900, 355, 12183, 418, 462, 13, 50713], 'temperature': 0.0, 'avg_logprob': -0.1364655918545193, 'compression_ratio': 1.6355140186915889, 'no_speech_prob': 0.012160781770944595}, {'id': 38, 'seek': 18300, 'start': 190.0, 'end': 196.0, 'text': \" So we'll talk a little bit about ribosome. So as you can see in the footnote, it's a protein synthesis machine in the cell.\", 'tokens': [50713, 1406, 356, 1183, 1561, 257, 1310, 1643, 546, 12183, 418, 462, 13, 1406, 355, 345, 460, 766, 287, 262, 38512, 11, 340, 338, 257, 7532, 21263, 4572, 287, 262, 2685, 13, 51013], 'temperature': 0.0, 'avg_logprob': -0.1364655918545193, 'compression_ratio': 1.6355140186915889, 'no_speech_prob': 0.012160781770944595}, {'id': 39, 'seek': 18300, 'start': 196.0, 'end': 207.0, 'text': ' It is a molecular, macromolecular complex made up of protein subunits together with certain RNAs that you will see is a catalytic RNA', 'tokens': [51013, 632, 318, 257, 18955, 11, 8352, 398, 2305, 10440, 3716, 925, 510, 286, 7532, 850, 41667, 1978, 351, 1728, 43112, 1722, 326, 345, 481, 766, 318, 257, 36745, 13370, 25897, 51563], 'temperature': 0.0, 'avg_logprob': -0.1364655918545193, 'compression_ratio': 1.6355140186915889, 'no_speech_prob': 0.012160781770944595}, {'id': 40, 'seek': 20700, 'start': 207.0, 'end': 216.0, 'text': ' that essentially together, this machine allows you to convert the mRNA sequence to your protein sequence.', 'tokens': [50363, 326, 6986, 1978, 11, 428, 4572, 3578, 345, 284, 10385, 262, 47227, 8379, 284, 534, 7532, 8379, 13, 50813], 'temperature': 0.0, 'avg_logprob': -0.16713451385498046, 'compression_ratio': 1.6323529411764706, 'no_speech_prob': 0.3385063707828522}, {'id': 41, 'seek': 20700, 'start': 216.0, 'end': 225.0, 'text': ' Proteins are made up of amino acids that essentially is a long polymer of individual 20-coding amino acids.', 'tokens': [50813, 1041, 660, 1040, 389, 925, 510, 286, 23206, 17045, 326, 6986, 318, 257, 890, 30528, 286, 1981, 1160, 12, 66, 7656, 23206, 17045, 13, 51263], 'temperature': 0.0, 'avg_logprob': -0.16713451385498046, 'compression_ratio': 1.6323529411764706, 'no_speech_prob': 0.3385063707828522}, {'id': 42, 'seek': 20700, 'start': 225.0, 'end': 233.0, 'text': \" So you can imagine this machine, it doesn't have in all cells, including bacteria and bacteria, those bacteria like us,\", 'tokens': [51263, 1406, 345, 460, 5967, 428, 4572, 11, 340, 1595, 470, 423, 287, 477, 4778, 11, 1390, 11492, 290, 11492, 11, 883, 11492, 588, 514, 11, 51663], 'temperature': 0.0, 'avg_logprob': -0.16713451385498046, 'compression_ratio': 1.6323529411764706, 'no_speech_prob': 0.3385063707828522}, {'id': 43, 'seek': 23300, 'start': 233.0, 'end': 243.0, 'text': \" essentially it's a polymerization machine if you think about it, because you're going to allow amino acids, individual amino acids to be covalently linked.\", 'tokens': [50363, 6986, 340, 338, 257, 30528, 1634, 4572, 611, 345, 892, 546, 340, 11, 780, 345, 821, 1016, 284, 1249, 23206, 17045, 11, 1981, 23206, 17045, 284, 307, 269, 8325, 1473, 6692, 13, 50863], 'temperature': 0.0, 'avg_logprob': -0.20949992679414295, 'compression_ratio': 1.60352422907489, 'no_speech_prob': 0.07042592763900757}, {'id': 44, 'seek': 23300, 'start': 243.0, 'end': 256.0, 'text': ' In peptide bonds, so-called A-migons, you have these long chain proteins that you know that proteins are absolutely essential macromolecules that actually drive the cellular engine, just like in a car engine.', 'tokens': [50863, 554, 34337, 485, 13100, 11, 523, 12, 7174, 317, 12, 76, 328, 684, 11, 345, 423, 777, 890, 6333, 15568, 326, 345, 760, 326, 15568, 389, 5543, 6393, 8352, 398, 2305, 13930, 326, 1682, 3708, 262, 19824, 3113, 11, 655, 588, 287, 257, 1097, 3113, 13, 51513], 'temperature': 0.0, 'avg_logprob': -0.20949992679414295, 'compression_ratio': 1.60352422907489, 'no_speech_prob': 0.07042592763900757}, {'id': 45, 'seek': 25600, 'start': 256.0, 'end': 263.0, 'text': ' So that is a very essential aspect of protein synthesis is a ribosome.', 'tokens': [50363, 1406, 326, 318, 257, 845, 6393, 4843, 286, 7532, 21263, 318, 257, 12183, 418, 462, 13, 50713], 'temperature': 0.0, 'avg_logprob': -0.16153527469169804, 'compression_ratio': 1.616580310880829, 'no_speech_prob': 0.34936168789863586}, {'id': 46, 'seek': 25600, 'start': 263.0, 'end': 270.0, 'text': ' So before I go there, just give you an idea of, you know, you may wonder, where does protein synthesis happen?', 'tokens': [50713, 1406, 878, 314, 467, 612, 11, 655, 1577, 345, 281, 2126, 286, 11, 345, 760, 11, 345, 743, 4240, 11, 810, 857, 7532, 21263, 1645, 30, 51063], 'temperature': 0.0, 'avg_logprob': -0.16153527469169804, 'compression_ratio': 1.616580310880829, 'no_speech_prob': 0.34936168789863586}, {'id': 47, 'seek': 25600, 'start': 270.0, 'end': 274.0, 'text': \" What I'm showing you here is the architecture of a eukaryotic cell.\", 'tokens': [51063, 1867, 314, 1101, 4478, 345, 994, 318, 262, 10959, 286, 257, 304, 2724, 560, 6210, 2685, 13, 51263], 'temperature': 0.0, 'avg_logprob': -0.16153527469169804, 'compression_ratio': 1.616580310880829, 'no_speech_prob': 0.34936168789863586}, {'id': 48, 'seek': 25600, 'start': 274.0, 'end': 277.0, 'text': ' Why do you think protein synthesis happens?', 'tokens': [51263, 4162, 466, 345, 892, 7532, 21263, 4325, 30, 51413], 'temperature': 0.0, 'avg_logprob': -0.16153527469169804, 'compression_ratio': 1.616580310880829, 'no_speech_prob': 0.34936168789863586}, {'id': 49, 'seek': 25600, 'start': 277.0, 'end': 284.0, 'text': ' Which compartment?', 'tokens': [51413, 9022, 26247, 30, 51763], 'temperature': 0.0, 'avg_logprob': -0.16153527469169804, 'compression_ratio': 1.616580310880829, 'no_speech_prob': 0.34936168789863586}, {'id': 50, 'seek': 28400, 'start': 284.0, 'end': 292.0, 'text': ' In a very specific location, the majority is on the ER surface, but largely speaking, does it happen in the nucleus of the cytoplasm?', 'tokens': [50363, 554, 257, 845, 2176, 4067, 11, 262, 3741, 318, 319, 262, 13793, 4417, 11, 475, 5688, 5486, 11, 857, 340, 1645, 287, 262, 29984, 286, 262, 3075, 4852, 75, 8597, 30, 50763], 'temperature': 0.0, 'avg_logprob': -0.1610228157043457, 'compression_ratio': 1.7649769585253456, 'no_speech_prob': 0.410399466753006}, {'id': 51, 'seek': 28400, 'start': 292.0, 'end': 293.0, 'text': ' Which compartment?', 'tokens': [50763, 9022, 26247, 30, 50813], 'temperature': 0.0, 'avg_logprob': -0.1610228157043457, 'compression_ratio': 1.7649769585253456, 'no_speech_prob': 0.410399466753006}, {'id': 52, 'seek': 28400, 'start': 293.0, 'end': 295.0, 'text': ' Cytoplasm, right?', 'tokens': [50813, 5934, 4852, 75, 8597, 11, 826, 30, 50913], 'temperature': 0.0, 'avg_logprob': -0.1610228157043457, 'compression_ratio': 1.7649769585253456, 'no_speech_prob': 0.410399466753006}, {'id': 53, 'seek': 28400, 'start': 295.0, 'end': 299.0, 'text': ' Whereas mRNA synthesis, where does it happen in this eukaryotic cell?', 'tokens': [50913, 28016, 47227, 21263, 11, 810, 857, 340, 1645, 287, 428, 304, 2724, 560, 6210, 2685, 30, 51113], 'temperature': 0.0, 'avg_logprob': -0.1610228157043457, 'compression_ratio': 1.7649769585253456, 'no_speech_prob': 0.410399466753006}, {'id': 54, 'seek': 28400, 'start': 299.0, 'end': 302.0, 'text': ' In the nucleus, right?', 'tokens': [51113, 554, 262, 29984, 11, 826, 30, 51263], 'temperature': 0.0, 'avg_logprob': -0.1610228157043457, 'compression_ratio': 1.7649769585253456, 'no_speech_prob': 0.410399466753006}, {'id': 55, 'seek': 28400, 'start': 302.0, 'end': 309.0, 'text': ' So there is this really compartmentalized segregation in eukaryotic system where mRNA synthesis happens in the nucleus,', 'tokens': [51263, 1406, 612, 318, 428, 1107, 26247, 282, 1143, 26718, 287, 304, 2724, 560, 6210, 1080, 810, 47227, 21263, 4325, 287, 262, 29984, 11, 51613], 'temperature': 0.0, 'avg_logprob': -0.1610228157043457, 'compression_ratio': 1.7649769585253456, 'no_speech_prob': 0.410399466753006}, {'id': 56, 'seek': 30900, 'start': 309.0, 'end': 318.0, 'text': ' and this mRNA undergoes maturation processing, ultimately get exported outside the nucleus into the cytoplasm,', 'tokens': [50363, 290, 428, 47227, 17777, 274, 285, 36921, 7587, 11, 6165, 651, 29050, 2354, 262, 29984, 656, 262, 3075, 4852, 75, 8597, 11, 50813], 'temperature': 0.0, 'avg_logprob': -0.1150209880111241, 'compression_ratio': 1.619047619047619, 'no_speech_prob': 0.23736900091171265}, {'id': 57, 'seek': 30900, 'start': 318.0, 'end': 328.0, 'text': ' where it will meet this machinery that I mentioned, ribosome, that will catalyze the polypeptide, i.e. amino acids polymerization, to give you the protein.', 'tokens': [50813, 810, 340, 481, 1826, 428, 20230, 326, 314, 4750, 11, 12183, 418, 462, 11, 326, 481, 36745, 2736, 262, 7514, 431, 457, 485, 11, 1312, 13, 68, 13, 23206, 17045, 30528, 1634, 11, 284, 1577, 345, 262, 7532, 13, 51313], 'temperature': 0.0, 'avg_logprob': -0.1150209880111241, 'compression_ratio': 1.619047619047619, 'no_speech_prob': 0.23736900091171265}, {'id': 58, 'seek': 30900, 'start': 328.0, 'end': 337.0, 'text': ' So these individual black dots, you may see better in your slides, these are actually ribosome, these sort of a complex of RNAs and proteins.', 'tokens': [51313, 1406, 777, 1981, 2042, 22969, 11, 345, 743, 766, 1365, 287, 534, 19392, 11, 777, 389, 1682, 12183, 418, 462, 11, 777, 3297, 286, 257, 3716, 286, 43112, 1722, 290, 15568, 13, 51763], 'temperature': 0.0, 'avg_logprob': -0.1150209880111241, 'compression_ratio': 1.619047619047619, 'no_speech_prob': 0.23736900091171265}, {'id': 59, 'seek': 33700, 'start': 337.0, 'end': 345.0, 'text': ' They are essentially type complex. They are individual ribosome complex molecules that are diffusing throughout the cytoplasm.', 'tokens': [50363, 1119, 389, 6986, 2099, 3716, 13, 1119, 389, 1981, 12183, 418, 462, 3716, 17745, 326, 389, 814, 3500, 3690, 262, 3075, 4852, 75, 8597, 13, 50763], 'temperature': 0.0, 'avg_logprob': -0.12742719532530986, 'compression_ratio': 1.580188679245283, 'no_speech_prob': 0.06587988883256912}, {'id': 60, 'seek': 33700, 'start': 345.0, 'end': 358.0, 'text': \" So protein synthesis actually happens in the cytoplasm. It is true, the student over there said, it's also concentrated on the ER, so these tubules, tubular green structures you see are endoplasmic reticulum.\", 'tokens': [50763, 1406, 7532, 21263, 1682, 4325, 287, 262, 3075, 4852, 75, 8597, 13, 632, 318, 2081, 11, 262, 3710, 625, 612, 531, 11, 340, 338, 635, 17298, 319, 262, 13793, 11, 523, 777, 12202, 5028, 11, 12202, 934, 4077, 8573, 345, 766, 389, 886, 20106, 8597, 291, 1005, 291, 14452, 13, 51413], 'temperature': 0.0, 'avg_logprob': -0.12742719532530986, 'compression_ratio': 1.580188679245283, 'no_speech_prob': 0.06587988883256912}, {'id': 61, 'seek': 35800, 'start': 358.0, 'end': 372.0, 'text': \" So we won't worry so much in our cost level, but essentially the key point is that protein synthesis happens outside the nucleus, broadly speaking in the cytoplasm, which is very different from bacteria system like E. coli.\", 'tokens': [50363, 1406, 356, 1839, 470, 5490, 523, 881, 287, 674, 1575, 1241, 11, 475, 6986, 262, 1994, 966, 318, 326, 7532, 21263, 4325, 2354, 262, 29984, 11, 18633, 5486, 287, 262, 3075, 4852, 75, 8597, 11, 543, 318, 845, 1180, 422, 11492, 1080, 588, 412, 13, 40395, 13, 51063], 'temperature': 0.0, 'avg_logprob': -0.1492058535416921, 'compression_ratio': 1.630952380952381, 'no_speech_prob': 0.6753475666046143}, {'id': 62, 'seek': 35800, 'start': 372.0, 'end': 385.0, 'text': \" In E. coli, do we worry about cytoplasm versus nucleus? The answer is no, right? There's essentially translation and transcription couple process in E. coli, as we talked about last week.\", 'tokens': [51063, 554, 412, 13, 40395, 11, 466, 356, 5490, 546, 3075, 4852, 75, 8597, 9051, 29984, 30, 383, 3280, 318, 645, 11, 826, 30, 1318, 338, 6986, 11059, 290, 26955, 3155, 1429, 287, 412, 13, 40395, 11, 355, 356, 6619, 546, 938, 1285, 13, 51713], 'temperature': 0.0, 'avg_logprob': -0.1492058535416921, 'compression_ratio': 1.630952380952381, 'no_speech_prob': 0.6753475666046143}, {'id': 63, 'seek': 38500, 'start': 385.0, 'end': 400.0, 'text': ' So you may wonder, right, how fast are these processes? If you think about it, you know, one second is very fast, right? And actually ribosome can polymerize 10, on average, 10 amino acids in a second, which is really very rapid, right?', 'tokens': [50363, 1406, 345, 743, 4240, 11, 826, 11, 703, 3049, 389, 777, 7767, 30, 1002, 345, 892, 546, 340, 11, 345, 760, 11, 530, 1218, 318, 845, 3049, 11, 826, 30, 843, 1682, 12183, 418, 462, 460, 30528, 1096, 838, 11, 319, 2811, 11, 838, 23206, 17045, 287, 257, 1218, 11, 543, 318, 1107, 845, 5801, 11, 826, 30, 51113], 'temperature': 0.0, 'avg_logprob': -0.20049346923828126, 'compression_ratio': 1.6352459016393444, 'no_speech_prob': 0.20434410870075226}, {'id': 64, 'seek': 38500, 'start': 400.0, 'end': 409.0, 'text': \" As I'm talking about, even before I finish the sentence, it will be several seconds, right, perhaps. And ribosome has already made 10 amino acids linked together.\", 'tokens': [51113, 1081, 314, 1101, 3375, 546, 11, 772, 878, 314, 5461, 262, 6827, 11, 340, 481, 307, 1811, 4201, 11, 826, 11, 3737, 13, 843, 12183, 418, 462, 468, 1541, 925, 838, 23206, 17045, 6692, 1978, 13, 51563], 'temperature': 0.0, 'avg_logprob': -0.20049346923828126, 'compression_ratio': 1.6352459016393444, 'no_speech_prob': 0.20434410870075226}, {'id': 65, 'seek': 40900, 'start': 409.0, 'end': 418.0, 'text': ' And average protein is about 300 amino acids linked. So, if you compare ribosome rate that I mentioned, protein translation rate.', 'tokens': [50363, 843, 2811, 7532, 318, 546, 5867, 23206, 17045, 6692, 13, 1406, 11, 611, 345, 8996, 12183, 418, 462, 2494, 326, 314, 4750, 11, 7532, 11059, 2494, 13, 50813], 'temperature': 0.0, 'avg_logprob': -0.11784065530655231, 'compression_ratio': 1.675, 'no_speech_prob': 0.5664843320846558}, {'id': 66, 'seek': 40900, 'start': 418.0, 'end': 434.0, 'text': \" So what I'm showing you here, you don't need to memorize these numbers, but at least good to know at the level of, you know, back of the envelope number, what kind of rate that are we talking about, when you think about RNA polymerization versus amino acid polymerization.\", 'tokens': [50813, 1406, 644, 314, 1101, 4478, 345, 994, 11, 345, 836, 470, 761, 284, 16181, 1096, 777, 3146, 11, 475, 379, 1551, 922, 284, 760, 379, 262, 1241, 286, 11, 345, 760, 11, 736, 286, 262, 22878, 1271, 11, 644, 1611, 286, 2494, 326, 389, 356, 3375, 546, 11, 618, 345, 892, 546, 25897, 30528, 1634, 9051, 23206, 7408, 30528, 1634, 13, 51613], 'temperature': 0.0, 'avg_logprob': -0.11784065530655231, 'compression_ratio': 1.675, 'no_speech_prob': 0.5664843320846558}, {'id': 67, 'seek': 43400, 'start': 434.0, 'end': 450.0, 'text': ' RNA polymerization gives you mRNA from DNA, amino acid polymerization gives you protein from mRNA. So what types of enzyme that we saw last week, what does the mRNA synthesis?', 'tokens': [50363, 25897, 30528, 1634, 3607, 345, 47227, 422, 7446, 11, 23206, 7408, 30528, 1634, 3607, 345, 7532, 422, 47227, 13, 1406, 644, 3858, 286, 27679, 326, 356, 2497, 938, 1285, 11, 644, 857, 262, 47227, 21263, 30, 51163], 'temperature': 0.0, 'avg_logprob': -0.1995506167411804, 'compression_ratio': 1.4112903225806452, 'no_speech_prob': 0.42428576946258545}, {'id': 68, 'seek': 45000, 'start': 450.0, 'end': 469.0, 'text': \" So a machinery that produces mRNA, what we call these types of enzyme as, are polymerases and the RNA polymerases, right, machinery that makes DNA will be DNA polymerases, right, because it's all polymerizing the biomolecular polymers.\", 'tokens': [50363, 1406, 257, 20230, 326, 11073, 47227, 11, 644, 356, 869, 777, 3858, 286, 27679, 355, 11, 389, 30528, 1386, 290, 262, 25897, 30528, 1386, 11, 826, 11, 20230, 326, 1838, 7446, 481, 307, 7446, 30528, 1386, 11, 826, 11, 780, 340, 338, 477, 30528, 2890, 262, 27488, 2305, 10440, 7514, 11056, 13, 51313], 'temperature': 0.0, 'avg_logprob': -0.1921318428856986, 'compression_ratio': 1.5666666666666667, 'no_speech_prob': 0.4728154242038727}, {'id': 69, 'seek': 46900, 'start': 469.0, 'end': 489.0, 'text': ' If you think about the rates in terms of actual DNA replication rate is in the bacteria world on the E. coli, for instance, versus mammalian cells such as, you know, cancer cell type of cancer cell line, HeLa cells, the replication timescale are quite different, right.', 'tokens': [50363, 1002, 345, 892, 546, 262, 3965, 287, 2846, 286, 4036, 7446, 30330, 2494, 318, 287, 262, 11492, 995, 319, 262, 412, 13, 40395, 11, 329, 4554, 11, 9051, 46177, 4778, 884, 355, 11, 345, 760, 11, 4890, 2685, 2099, 286, 4890, 2685, 1627, 11, 679, 14772, 4778, 11, 262, 30330, 1661, 38765, 389, 2407, 1180, 11, 826, 13, 51363], 'temperature': 0.0, 'avg_logprob': -0.23467828381446101, 'compression_ratio': 1.5197740112994351, 'no_speech_prob': 0.4383728504180908}, {'id': 70, 'seek': 48900, 'start': 489.0, 'end': 501.0, 'text': \" But at the fundamental level, at the nucleotide level, the actual rate of transcription is, largely speaking, identical. You can make per second, if I'm taking 10 to 100 nucleotides.\", 'tokens': [50363, 887, 379, 262, 7531, 1241, 11, 379, 262, 17751, 45608, 1241, 11, 262, 4036, 2494, 286, 26955, 318, 11, 5688, 5486, 11, 10411, 13, 921, 460, 787, 583, 1218, 11, 611, 314, 1101, 2263, 838, 284, 1802, 17751, 313, 1460, 13, 50963], 'temperature': 0.0, 'avg_logprob': -0.12031697063911252, 'compression_ratio': 1.543778801843318, 'no_speech_prob': 0.4211328625679016}, {'id': 71, 'seek': 48900, 'start': 501.0, 'end': 512.0, 'text': ' But why then? Why is it that our cell division in human cells is almost a day, whereas cell division in E. coli and many bacteria are less than an hour?', 'tokens': [50963, 887, 1521, 788, 30, 4162, 318, 340, 326, 674, 2685, 7297, 287, 1692, 4778, 318, 2048, 257, 1110, 11, 9472, 2685, 7297, 287, 412, 13, 40395, 290, 867, 11492, 389, 1342, 621, 281, 1711, 30, 51513], 'temperature': 0.0, 'avg_logprob': -0.12031697063911252, 'compression_ratio': 1.543778801843318, 'no_speech_prob': 0.4211328625679016}, {'id': 72, 'seek': 51200, 'start': 512.0, 'end': 522.0, 'text': ' Why is it that DNA replication overall process takes much longer in our cells compared to E. coli?', 'tokens': [50363, 4162, 318, 340, 326, 7446, 30330, 4045, 1429, 2753, 881, 2392, 287, 674, 4778, 3688, 284, 412, 13, 40395, 30, 50863], 'temperature': 0.0, 'avg_logprob': -0.13612428307533264, 'compression_ratio': 1.0888888888888888, 'no_speech_prob': 0.3630462884902954}, {'id': 73, 'seek': 52200, 'start': 522.0, 'end': 541.0, 'text': \" Yes, certainly it's more complicated. You have to do the group checking, editing, etc. But still, the key point also is the fact that in the gene size, on average gene size, human gene is a large, much larger than bacterial gene.\", 'tokens': [50363, 3363, 11, 3729, 340, 338, 517, 8253, 13, 921, 423, 284, 466, 262, 1448, 10627, 11, 12857, 11, 3503, 13, 887, 991, 11, 262, 1994, 966, 635, 318, 262, 1109, 326, 287, 262, 9779, 2546, 11, 319, 2811, 9779, 2546, 11, 1692, 9779, 318, 257, 1588, 11, 881, 4025, 621, 23462, 9779, 13, 51313], 'temperature': 0.0, 'avg_logprob': -0.2405266845435427, 'compression_ratio': 1.506578947368421, 'no_speech_prob': 0.6162317395210266}, {'id': 74, 'seek': 54100, 'start': 541.0, 'end': 555.0, 'text': ' Can you guys think about the fundamental reason why that is the case? We talked a bit about that last week. What is it that is intrinsically, as a stark difference in the genome level between E. coli genome and human genome?', 'tokens': [50363, 1680, 345, 3730, 892, 546, 262, 7531, 1738, 1521, 326, 318, 262, 1339, 30, 775, 6619, 257, 1643, 546, 326, 938, 1285, 13, 1867, 318, 340, 326, 318, 48676, 11, 355, 257, 19278, 3580, 287, 262, 19270, 1241, 1022, 412, 13, 40395, 19270, 290, 1692, 19270, 30, 51063], 'temperature': 0.0, 'avg_logprob': -0.15858878341375612, 'compression_ratio': 1.4736842105263157, 'no_speech_prob': 0.6303730607032776}, {'id': 75, 'seek': 55500, 'start': 555.0, 'end': 570.0, 'text': \" Yeah, presence of non-coding genetic regions, so-called introns. Intons and exons, you don't have to worry about that in prokaryotic systems like E. coli or bacteria.\", 'tokens': [50363, 9425, 11, 4931, 286, 1729, 12, 66, 7656, 8513, 7652, 11, 523, 12, 7174, 9913, 684, 13, 2558, 684, 290, 409, 684, 11, 345, 836, 470, 423, 284, 5490, 546, 326, 287, 386, 74, 560, 6210, 3341, 588, 412, 13, 40395, 393, 11492, 13, 51113], 'temperature': 0.0, 'avg_logprob': -0.2008612354596456, 'compression_ratio': 1.2671755725190839, 'no_speech_prob': 0.868340015411377}, {'id': 76, 'seek': 57000, 'start': 570.0, 'end': 585.0, 'text': \" But we don't have introns in E. coli, whereas we have introns. So that means that for a given gene, even if ultimately the protein sequence is identical between E. coli protein and human protein, the actual starting genomes will be quite different.\", 'tokens': [50363, 887, 356, 836, 470, 423, 9913, 684, 287, 412, 13, 40395, 11, 9472, 356, 423, 9913, 684, 13, 1406, 326, 1724, 326, 329, 257, 1813, 9779, 11, 772, 611, 6165, 262, 7532, 8379, 318, 10411, 1022, 412, 13, 40395, 7532, 290, 1692, 7532, 11, 262, 4036, 3599, 42136, 481, 307, 2407, 1180, 13, 51113], 'temperature': 0.0, 'avg_logprob': -0.12480886358963816, 'compression_ratio': 1.5403726708074534, 'no_speech_prob': 0.8399501442909241}, {'id': 77, 'seek': 58500, 'start': 585.0, 'end': 600.0, 'text': ' Because in origin of the DNA in the human cell, we have introns and exons. Ultimately, introns are spliced out inside the nucleus, right, we talked about that last week, before you ultimately get to translated protein.', 'tokens': [50363, 4362, 287, 8159, 286, 262, 7446, 287, 262, 1692, 2685, 11, 356, 423, 9913, 684, 290, 409, 684, 13, 24199, 11, 9913, 684, 389, 4328, 3711, 503, 2641, 262, 29984, 11, 826, 11, 356, 6619, 546, 326, 938, 1285, 11, 878, 345, 6165, 651, 284, 14251, 7532, 13, 51113], 'temperature': 0.0, 'avg_logprob': -0.19976058373084435, 'compression_ratio': 1.4342105263157894, 'no_speech_prob': 0.8339758515357971}, {'id': 78, 'seek': 60000, 'start': 600.0, 'end': 615.0, 'text': ' So even if the amino acid level sequence is identical, at the DNA level, there will be introns in the human cells and eukaryotic system that will make the whole trans DNA synthesis step taking much longer.', 'tokens': [50363, 1406, 772, 611, 262, 23206, 7408, 1241, 8379, 318, 10411, 11, 379, 262, 7446, 1241, 11, 612, 481, 307, 9913, 684, 287, 262, 1692, 4778, 290, 304, 2724, 560, 6210, 1080, 326, 481, 787, 262, 2187, 1007, 7446, 21263, 2239, 2263, 881, 2392, 13, 51113], 'temperature': 0.0, 'avg_logprob': -0.11145347502173447, 'compression_ratio': 1.6221198156682028, 'no_speech_prob': 0.7764801979064941}, {'id': 79, 'seek': 60000, 'start': 615.0, 'end': 626.0, 'text': ' So replication process will be a lot longer. So that is certainly one key fundamental aspect of the eukaryotic cell, having a presence of introns.', 'tokens': [51113, 1406, 30330, 1429, 481, 307, 257, 1256, 2392, 13, 1406, 326, 318, 3729, 530, 1994, 7531, 4843, 286, 262, 304, 2724, 560, 6210, 2685, 11, 1719, 257, 4931, 286, 9913, 684, 13, 51663], 'temperature': 0.0, 'avg_logprob': -0.11145347502173447, 'compression_ratio': 1.6221198156682028, 'no_speech_prob': 0.7764801979064941}, {'id': 80, 'seek': 62600, 'start': 626.0, 'end': 637.0, 'text': \" Now, protein synthesis, which is the focus of today's discussion, as I said, the average translation rate is 10 amino acids given take per second.\", 'tokens': [50363, 2735, 11, 7532, 21263, 11, 543, 318, 262, 2962, 286, 1909, 338, 5114, 11, 355, 314, 531, 11, 262, 2811, 11059, 2494, 318, 838, 23206, 17045, 1813, 1011, 583, 1218, 13, 50913], 'temperature': 0.0, 'avg_logprob': -0.10098431327126244, 'compression_ratio': 1.6496350364963503, 'no_speech_prob': 0.24965786933898926}, {'id': 81, 'seek': 62600, 'start': 637.0, 'end': 646.0, 'text': \" So by that time, why is it that we don't have to care about these introns and exons anymore by the time we're talking about ribosome translation?\", 'tokens': [50913, 1406, 416, 326, 640, 11, 1521, 318, 340, 326, 356, 836, 470, 423, 284, 1337, 546, 777, 9913, 684, 290, 409, 684, 7471, 416, 262, 640, 356, 821, 3375, 546, 12183, 418, 462, 11059, 30, 51363], 'temperature': 0.0, 'avg_logprob': -0.10098431327126244, 'compression_ratio': 1.6496350364963503, 'no_speech_prob': 0.24965786933898926}, {'id': 82, 'seek': 62600, 'start': 646.0, 'end': 654.0, 'text': \" So the blue unit that I'm pointing out here is essentially a classical way to illustrate the ribosome because it has two subunits, a large one and a small one.\", 'tokens': [51363, 1406, 262, 4171, 4326, 326, 314, 1101, 10609, 503, 994, 318, 6986, 257, 15993, 835, 284, 19418, 262, 12183, 418, 462, 780, 340, 468, 734, 850, 41667, 11, 257, 1588, 530, 290, 257, 1402, 530, 13, 51763], 'temperature': 0.0, 'avg_logprob': -0.10098431327126244, 'compression_ratio': 1.6496350364963503, 'no_speech_prob': 0.24965786933898926}, {'id': 83, 'seek': 65400, 'start': 654.0, 'end': 661.0, 'text': ' So the green piece is the polypeptides of proteins that are being translated, are being synthesized.', 'tokens': [50363, 1406, 262, 4077, 3704, 318, 262, 7514, 431, 457, 1460, 286, 15568, 326, 389, 852, 14251, 11, 389, 852, 24983, 1143, 13, 50713], 'temperature': 0.0, 'avg_logprob': -0.08457424670835084, 'compression_ratio': 1.6310679611650485, 'no_speech_prob': 0.06005080044269562}, {'id': 84, 'seek': 65400, 'start': 661.0, 'end': 669.0, 'text': ' And then the red piece is the incoming mRNA that is coding for what amino acids should be in the protein sequence.', 'tokens': [50713, 843, 788, 262, 2266, 3704, 318, 262, 15619, 47227, 326, 318, 19617, 329, 644, 23206, 17045, 815, 307, 287, 262, 7532, 8379, 13, 51113], 'temperature': 0.0, 'avg_logprob': -0.08457424670835084, 'compression_ratio': 1.6310679611650485, 'no_speech_prob': 0.06005080044269562}, {'id': 85, 'seek': 65400, 'start': 669.0, 'end': 676.0, 'text': \" So given this, why do you think that at the point that we talk about translation, we don't have this difference anymore?\", 'tokens': [51113, 1406, 1813, 428, 11, 1521, 466, 345, 892, 326, 379, 262, 966, 326, 356, 1561, 546, 11059, 11, 356, 836, 470, 423, 428, 3580, 7471, 30, 51463], 'temperature': 0.0, 'avg_logprob': -0.08457424670835084, 'compression_ratio': 1.6310679611650485, 'no_speech_prob': 0.06005080044269562}, {'id': 86, 'seek': 67600, 'start': 676.0, 'end': 679.0, 'text': ' Yes, it is splicing. Absolutely.', 'tokens': [50363, 3363, 11, 340, 318, 4328, 6345, 13, 28416, 13, 50513], 'temperature': 0.0, 'avg_logprob': -0.18198345689212575, 'compression_ratio': 1.5449438202247192, 'no_speech_prob': 0.16806462407112122}, {'id': 87, 'seek': 67600, 'start': 679.0, 'end': 684.0, 'text': ' So mRNA that we talk about here is already a mature messenger RNA.', 'tokens': [50513, 1406, 47227, 326, 356, 1561, 546, 994, 318, 1541, 257, 15345, 31228, 25897, 13, 50763], 'temperature': 0.0, 'avg_logprob': -0.18198345689212575, 'compression_ratio': 1.5449438202247192, 'no_speech_prob': 0.16806462407112122}, {'id': 88, 'seek': 67600, 'start': 684.0, 'end': 699.0, 'text': ' Whereas the newly synthesized, the nascently synthesized, if you like, the mRNA up there at the point of the RNA polymerase doing the transcription will be the premature mRNA.', 'tokens': [50763, 28016, 262, 8308, 24983, 1143, 11, 262, 45419, 306, 24983, 1143, 11, 611, 345, 588, 11, 262, 47227, 510, 612, 379, 262, 966, 286, 262, 25897, 30528, 589, 1804, 262, 26955, 481, 307, 262, 19905, 47227, 13, 51513], 'temperature': 0.0, 'avg_logprob': -0.18198345689212575, 'compression_ratio': 1.5449438202247192, 'no_speech_prob': 0.16806462407112122}, {'id': 89, 'seek': 69900, 'start': 699.0, 'end': 709.0, 'text': ' You also tackled that question in the problem set last week, where you need to go through splicing and export, you know, decoration with poly-A tail, etc.', 'tokens': [50363, 921, 635, 35457, 326, 1808, 287, 262, 1917, 900, 938, 1285, 11, 810, 345, 761, 284, 467, 832, 4328, 6345, 290, 10784, 11, 345, 760, 11, 42050, 351, 7514, 12, 32, 7894, 11, 3503, 13, 50863], 'temperature': 0.0, 'avg_logprob': -0.15592191574421335, 'compression_ratio': 1.5454545454545454, 'no_speech_prob': 0.5966856479644775}, {'id': 90, 'seek': 69900, 'start': 709.0, 'end': 714.0, 'text': ' before you can actually get to the mature mRNA that is ready for ribosome translation.', 'tokens': [50863, 878, 345, 460, 1682, 651, 284, 262, 15345, 47227, 326, 318, 3492, 329, 12183, 418, 462, 11059, 13, 51113], 'temperature': 0.0, 'avg_logprob': -0.15592191574421335, 'compression_ratio': 1.5454545454545454, 'no_speech_prob': 0.5966856479644775}, {'id': 91, 'seek': 69900, 'start': 714.0, 'end': 718.0, 'text': \" So remember, this has to be mature mRNA. At that point, it's identical.\", 'tokens': [51113, 1406, 3505, 11, 428, 468, 284, 307, 15345, 47227, 13, 1629, 326, 966, 11, 340, 338, 10411, 13, 51313], 'temperature': 0.0, 'avg_logprob': -0.15592191574421335, 'compression_ratio': 1.5454545454545454, 'no_speech_prob': 0.5966856479644775}, {'id': 92, 'seek': 69900, 'start': 718.0, 'end': 722.0, 'text': ' Introns no longer matter because they have been spliced out.', 'tokens': [51313, 2558, 12212, 645, 2392, 2300, 780, 484, 423, 587, 4328, 3711, 503, 13, 51513], 'temperature': 0.0, 'avg_logprob': -0.15592191574421335, 'compression_ratio': 1.5454545454545454, 'no_speech_prob': 0.5966856479644775}, {'id': 93, 'seek': 72200, 'start': 722.0, 'end': 733.0, 'text': \" OK. So the take home message is there's differences in time scale, broadly speaking, and also in terms of the translation, efficiency is the same between the two systems.\", 'tokens': [50363, 7477, 13, 1406, 262, 1011, 1363, 3275, 318, 612, 338, 5400, 287, 640, 5046, 11, 18633, 5486, 11, 290, 635, 287, 2846, 286, 262, 11059, 11, 9332, 318, 262, 976, 1022, 262, 734, 3341, 13, 50913], 'temperature': 0.0, 'avg_logprob': -0.12045338227576816, 'compression_ratio': 1.5546875, 'no_speech_prob': 0.11256681382656097}, {'id': 94, 'seek': 72200, 'start': 733.0, 'end': 738.0, 'text': \" So let's think about bacterial mRNA and eukaryotic mRNA.\", 'tokens': [50913, 1406, 1309, 338, 892, 546, 23462, 47227, 290, 304, 2724, 560, 6210, 47227, 13, 51163], 'temperature': 0.0, 'avg_logprob': -0.12045338227576816, 'compression_ratio': 1.5546875, 'no_speech_prob': 0.11256681382656097}, {'id': 95, 'seek': 72200, 'start': 738.0, 'end': 744.0, 'text': \" What I'm showing you here is a domain structure, very simplified, but it gives you a generic feature.\", 'tokens': [51163, 1867, 314, 1101, 4478, 345, 994, 318, 257, 7386, 4645, 11, 845, 27009, 11, 475, 340, 3607, 345, 257, 14276, 3895, 13, 51463], 'temperature': 0.0, 'avg_logprob': -0.12045338227576816, 'compression_ratio': 1.5546875, 'no_speech_prob': 0.11256681382656097}, {'id': 96, 'seek': 72200, 'start': 744.0, 'end': 748.0, 'text': \" In terms of coding sequences, don't do the same for a given protein.\", 'tokens': [51463, 554, 2846, 286, 19617, 16311, 11, 836, 470, 466, 262, 976, 329, 257, 1813, 7532, 13, 51663], 'temperature': 0.0, 'avg_logprob': -0.12045338227576816, 'compression_ratio': 1.5546875, 'no_speech_prob': 0.11256681382656097}, {'id': 97, 'seek': 74800, 'start': 748.0, 'end': 754.0, 'text': \" Because we don't need to worry about introns anymore by the time you get to the mature mRNA.\", 'tokens': [50363, 4362, 356, 836, 470, 761, 284, 5490, 546, 9913, 684, 7471, 416, 262, 640, 345, 651, 284, 262, 15345, 47227, 13, 50663], 'temperature': 0.0, 'avg_logprob': -0.16942123517598193, 'compression_ratio': 1.5069124423963134, 'no_speech_prob': 0.26380160450935364}, {'id': 98, 'seek': 74800, 'start': 754.0, 'end': 761.0, 'text': ' But what other striking differences do you see between the two systems?', 'tokens': [50663, 887, 644, 584, 8871, 5400, 466, 345, 766, 1022, 262, 734, 3341, 30, 51013], 'temperature': 0.0, 'avg_logprob': -0.16942123517598193, 'compression_ratio': 1.5069124423963134, 'no_speech_prob': 0.26380160450935364}, {'id': 99, 'seek': 74800, 'start': 761.0, 'end': 770.0, 'text': \" Because we'll hijack quite a bit of these mRNA regulation and their individual domains and complements when we think about ribosome switch design and application,\", 'tokens': [51013, 4362, 356, 1183, 16836, 441, 2407, 257, 1643, 286, 777, 47227, 9001, 290, 511, 1981, 18209, 290, 1224, 902, 618, 356, 892, 546, 12183, 418, 462, 5078, 1486, 290, 3586, 11, 51463], 'temperature': 0.0, 'avg_logprob': -0.16942123517598193, 'compression_ratio': 1.5069124423963134, 'no_speech_prob': 0.26380160450935364}, {'id': 100, 'seek': 77000, 'start': 770.0, 'end': 776.0, 'text': ' the later part of the topic. So what do you see beyond the coding sequence?', 'tokens': [50363, 262, 1568, 636, 286, 262, 7243, 13, 1406, 644, 466, 345, 766, 3675, 262, 19617, 8379, 30, 50663], 'temperature': 0.0, 'avg_logprob': -0.18111299857115135, 'compression_ratio': 1.6647727272727273, 'no_speech_prob': 0.5201756358146667}, {'id': 101, 'seek': 77000, 'start': 776.0, 'end': 784.0, 'text': ' Coding sequence is the mRNA that will code for your amino acid sequence on the protein level.', 'tokens': [50663, 327, 7656, 8379, 318, 262, 47227, 326, 481, 2438, 329, 534, 23206, 7408, 8379, 319, 262, 7532, 1241, 13, 51063], 'temperature': 0.0, 'avg_logprob': -0.18111299857115135, 'compression_ratio': 1.6647727272727273, 'no_speech_prob': 0.5201756358146667}, {'id': 102, 'seek': 77000, 'start': 784.0, 'end': 789.0, 'text': ' You see untranslated region. So this is upstream untranslated region.', 'tokens': [51063, 921, 766, 1418, 26084, 17249, 3814, 13, 1406, 428, 318, 28717, 1418, 26084, 17249, 3814, 13, 51313], 'temperature': 0.0, 'avg_logprob': -0.18111299857115135, 'compression_ratio': 1.6647727272727273, 'no_speech_prob': 0.5201756358146667}, {'id': 103, 'seek': 77000, 'start': 789.0, 'end': 795.0, 'text': \" So 5' UTR, so untranslated region and downstream UTR.\", 'tokens': [51313, 1406, 642, 6, 471, 5446, 11, 523, 1418, 26084, 17249, 3814, 290, 33218, 471, 5446, 13, 51613], 'temperature': 0.0, 'avg_logprob': -0.18111299857115135, 'compression_ratio': 1.6647727272727273, 'no_speech_prob': 0.5201756358146667}, {'id': 104, 'seek': 79500, 'start': 795.0, 'end': 799.0, 'text': ' And those are present in both cases.', 'tokens': [50363, 843, 883, 389, 1944, 287, 1111, 2663, 13, 50563], 'temperature': 0.0, 'avg_logprob': -0.18963532919412132, 'compression_ratio': 1.6085106382978724, 'no_speech_prob': 0.08599299937486649}, {'id': 105, 'seek': 79500, 'start': 799.0, 'end': 805.0, 'text': \" Those components are essential for gene regulation and stability of mRNA. We'll come to that.\", 'tokens': [50563, 5845, 6805, 389, 6393, 329, 9779, 9001, 290, 10159, 286, 47227, 13, 775, 1183, 1282, 284, 326, 13, 50863], 'temperature': 0.0, 'avg_logprob': -0.18963532919412132, 'compression_ratio': 1.6085106382978724, 'no_speech_prob': 0.08599299937486649}, {'id': 106, 'seek': 79500, 'start': 805.0, 'end': 809.0, 'text': ' And that essentially happens in both bacteria and eukaryotic world.', 'tokens': [50863, 843, 326, 6986, 4325, 287, 1111, 11492, 290, 304, 2724, 560, 6210, 995, 13, 51063], 'temperature': 0.0, 'avg_logprob': -0.18963532919412132, 'compression_ratio': 1.6085106382978724, 'no_speech_prob': 0.08599299937486649}, {'id': 107, 'seek': 79500, 'start': 809.0, 'end': 814.0, 'text': \" But you see additional units such as poly-A tail and 5' prime capping.\", 'tokens': [51063, 887, 345, 766, 3224, 4991, 884, 355, 7514, 12, 32, 7894, 290, 642, 6, 6994, 1275, 2105, 13, 51313], 'temperature': 0.0, 'avg_logprob': -0.18963532919412132, 'compression_ratio': 1.6085106382978724, 'no_speech_prob': 0.08599299937486649}, {'id': 108, 'seek': 79500, 'start': 814.0, 'end': 823.0, 'text': ' That is absolutely essential to introduce this poly-A tail tagging and then cap modification on the 5 prime.', 'tokens': [51313, 1320, 318, 5543, 6393, 284, 10400, 428, 7514, 12, 32, 7894, 49620, 290, 788, 1451, 17613, 319, 262, 642, 6994, 13, 51763], 'temperature': 0.0, 'avg_logprob': -0.18963532919412132, 'compression_ratio': 1.6085106382978724, 'no_speech_prob': 0.08599299937486649}, {'id': 109, 'seek': 82300, 'start': 823.0, 'end': 831.0, 'text': ' Fourth, eukaryotic mRNA, because when you think about eukaryotic mRNA, they are first produced in the nucleus,', 'tokens': [50363, 15692, 11, 304, 2724, 560, 6210, 47227, 11, 780, 618, 345, 892, 546, 304, 2724, 560, 6210, 47227, 11, 484, 389, 717, 4635, 287, 262, 29984, 11, 50763], 'temperature': 0.0, 'avg_logprob': -0.13216607466987942, 'compression_ratio': 1.5086705202312138, 'no_speech_prob': 0.20860248804092407}, {'id': 110, 'seek': 82300, 'start': 831.0, 'end': 842.0, 'text': ' but they need to be chaperoned and exported in a safe manner to the cytoplasm where they will meet ribosome such that they can synthesize the protein.', 'tokens': [50763, 475, 484, 761, 284, 307, 442, 2136, 12004, 290, 29050, 287, 257, 3338, 5642, 284, 262, 3075, 4852, 75, 8597, 810, 484, 481, 1826, 12183, 418, 462, 884, 326, 484, 460, 24983, 1096, 262, 7532, 13, 51313], 'temperature': 0.0, 'avg_logprob': -0.13216607466987942, 'compression_ratio': 1.5086705202312138, 'no_speech_prob': 0.20860248804092407}, {'id': 111, 'seek': 84200, 'start': 843.0, 'end': 850.0, 'text': ' If you have mRNA produced in the nucleus that are not stable and protected by nucleases degradation,', 'tokens': [50413, 1002, 345, 423, 47227, 4635, 287, 262, 29984, 326, 389, 407, 8245, 290, 6861, 416, 17751, 1386, 27486, 11, 50763], 'temperature': 0.0, 'avg_logprob': -0.16692736377454784, 'compression_ratio': 1.5693069306930694, 'no_speech_prob': 0.7356226444244385}, {'id': 112, 'seek': 84200, 'start': 850.0, 'end': 856.0, 'text': ' we talked a lot about RNA nucleases last lecture as a means of therapeutic mechanism.', 'tokens': [50763, 356, 6619, 257, 1256, 546, 25897, 17751, 1386, 938, 19143, 355, 257, 1724, 286, 21546, 9030, 13, 51063], 'temperature': 0.0, 'avg_logprob': -0.16692736377454784, 'compression_ratio': 1.5693069306930694, 'no_speech_prob': 0.7356226444244385}, {'id': 113, 'seek': 84200, 'start': 856.0, 'end': 865.0, 'text': ' Both nucleases could essentially chop up the RNA that is newly synthesized in the nucleus before it can even get to the cytoplasm.', 'tokens': [51063, 5747, 17751, 1386, 714, 6986, 30506, 510, 262, 25897, 326, 318, 8308, 24983, 1143, 287, 262, 29984, 878, 340, 460, 772, 651, 284, 262, 3075, 4852, 75, 8597, 13, 51513], 'temperature': 0.0, 'avg_logprob': -0.16692736377454784, 'compression_ratio': 1.5693069306930694, 'no_speech_prob': 0.7356226444244385}, {'id': 114, 'seek': 86500, 'start': 865.0, 'end': 891.0, 'text': ' So these accessories, 5 prime capping and poly-A tail tagging at the mRNA level is critical to stabilize and protect the eukaryotic mRNA such that they can be recognized by proteins that will bind to these units and actually chaperone them, escort them if you like, for mRNA export from the nucleus to the cytoplasm for protein translation.', 'tokens': [50363, 1406, 777, 18199, 11, 642, 6994, 1275, 2105, 290, 7514, 12, 32, 7894, 49620, 379, 262, 47227, 1241, 318, 4688, 284, 32411, 290, 1805, 262, 304, 2724, 560, 6210, 47227, 884, 326, 484, 460, 307, 8018, 416, 15568, 326, 481, 11007, 284, 777, 4991, 290, 1682, 442, 2136, 505, 606, 11, 27675, 606, 611, 345, 588, 11, 329, 47227, 10784, 422, 262, 29984, 284, 262, 3075, 4852, 75, 8597, 329, 7532, 11059, 13, 51663], 'temperature': 0.0, 'avg_logprob': -0.13183381959989474, 'compression_ratio': 1.596244131455399, 'no_speech_prob': 0.26294776797294617}, {'id': 115, 'seek': 89100, 'start': 891.0, 'end': 898.0, 'text': ' Keep in mind of this kind of segregation of compartments that makes the system more complex.', 'tokens': [50363, 9175, 287, 2000, 286, 428, 1611, 286, 26718, 286, 552, 32514, 326, 1838, 262, 1080, 517, 3716, 13, 50713], 'temperature': 0.0, 'avg_logprob': -0.20717738373110992, 'compression_ratio': 1.5393700787401574, 'no_speech_prob': 0.40725141763687134}, {'id': 116, 'seek': 89100, 'start': 898.0, 'end': 909.0, 'text': ' That means you require more factors to stabilize these mRNAs, protect these mRNA, chaperone and export them safely to the cytoplasm.', 'tokens': [50713, 1320, 1724, 345, 2421, 517, 5087, 284, 32411, 777, 285, 42336, 1722, 11, 1805, 777, 47227, 11, 442, 2136, 505, 290, 10784, 606, 11512, 284, 262, 3075, 4852, 75, 8597, 13, 51263], 'temperature': 0.0, 'avg_logprob': -0.20717738373110992, 'compression_ratio': 1.5393700787401574, 'no_speech_prob': 0.40725141763687134}, {'id': 117, 'seek': 89100, 'start': 909.0, 'end': 912.0, 'text': \" So that's what you're seeing here.\", 'tokens': [51263, 1406, 326, 338, 644, 345, 821, 4379, 994, 13, 51413], 'temperature': 0.0, 'avg_logprob': -0.20717738373110992, 'compression_ratio': 1.5393700787401574, 'no_speech_prob': 0.40725141763687134}, {'id': 118, 'seek': 89100, 'start': 912.0, 'end': 920.0, 'text': \" Sub-cellular loca will already talk about it. What I've just discussed, you can also read further in your own time in these texts.\", 'tokens': [51413, 3834, 12, 3846, 934, 1179, 64, 481, 1541, 1561, 546, 340, 13, 1867, 314, 1053, 655, 6693, 11, 345, 460, 635, 1100, 2252, 287, 534, 898, 640, 287, 777, 13399, 13, 51813], 'temperature': 0.0, 'avg_logprob': -0.20717738373110992, 'compression_ratio': 1.5393700787401574, 'no_speech_prob': 0.40725141763687134}, {'id': 119, 'seek': 92000, 'start': 920.0, 'end': 932.0, 'text': \" But let's go through a little bit more of what allows a ribosome to know you have a long piece of mRNA starting from the genome that's coming along.\", 'tokens': [50363, 887, 1309, 338, 467, 832, 257, 1310, 1643, 517, 286, 644, 3578, 257, 12183, 418, 462, 284, 760, 345, 423, 257, 890, 3704, 286, 47227, 3599, 422, 262, 19270, 326, 338, 2406, 1863, 13, 50963], 'temperature': 0.0, 'avg_logprob': -0.16349092343958413, 'compression_ratio': 1.6650717703349283, 'no_speech_prob': 0.32910171151161194}, {'id': 120, 'seek': 92000, 'start': 932.0, 'end': 944.0, 'text': ' How does ribosome know when to start translating, translating and synthesizing the amino acid, if you make this polymeric chain for a protein, when to start this process and when to end this process?', 'tokens': [50963, 1374, 857, 12183, 418, 462, 760, 618, 284, 923, 34665, 11, 34665, 290, 24983, 2890, 262, 23206, 7408, 11, 611, 345, 787, 428, 7514, 946, 6333, 329, 257, 7532, 11, 618, 284, 923, 428, 1429, 290, 618, 284, 886, 428, 1429, 30, 51563], 'temperature': 0.0, 'avg_logprob': -0.16349092343958413, 'compression_ratio': 1.6650717703349283, 'no_speech_prob': 0.32910171151161194}, {'id': 121, 'seek': 94400, 'start': 944.0, 'end': 955.0, 'text': ' So there are specific codes that you will be implemented in your mRNA in nature that allows a ribosome to signal that this must be a star site. Star codon is always AUG.', 'tokens': [50363, 1406, 612, 389, 2176, 12416, 326, 345, 481, 307, 9177, 287, 534, 47227, 287, 3450, 326, 3578, 257, 12183, 418, 462, 284, 6737, 326, 428, 1276, 307, 257, 3491, 2524, 13, 2907, 14873, 261, 318, 1464, 317, 7340, 13, 50913], 'temperature': 0.0, 'avg_logprob': -0.1386324988471137, 'compression_ratio': 1.7718446601941749, 'no_speech_prob': 0.7562346458435059}, {'id': 122, 'seek': 94400, 'start': 955.0, 'end': 967.0, 'text': ' And then there will be the end site, there will be a star codon in nature, there are three star codons and these three nucleotides will signal for where the star site and the end point should be.', 'tokens': [50913, 843, 788, 612, 481, 307, 262, 886, 2524, 11, 612, 481, 307, 257, 3491, 14873, 261, 287, 3450, 11, 612, 389, 1115, 3491, 14873, 684, 290, 777, 1115, 17751, 313, 1460, 481, 6737, 329, 810, 262, 3491, 2524, 290, 262, 886, 966, 815, 307, 13, 51513], 'temperature': 0.0, 'avg_logprob': -0.1386324988471137, 'compression_ratio': 1.7718446601941749, 'no_speech_prob': 0.7562346458435059}, {'id': 123, 'seek': 96700, 'start': 967.0, 'end': 981.0, 'text': \" Because if you think about it, these are, you know, if they are altered in any way, you'll have a protein that are either truncated, i.e. too short, shorter than you desire, or maybe start in the long side.\", 'tokens': [50363, 4362, 611, 345, 892, 546, 340, 11, 777, 389, 11, 345, 760, 11, 611, 484, 389, 14294, 287, 597, 835, 11, 345, 1183, 423, 257, 7532, 326, 389, 2035, 40122, 515, 11, 1312, 13, 68, 13, 1165, 1790, 11, 12238, 621, 345, 6227, 11, 393, 3863, 923, 287, 262, 890, 1735, 13, 51063], 'temperature': 0.0, 'avg_logprob': -0.11341676046681959, 'compression_ratio': 1.647887323943662, 'no_speech_prob': 0.42656153440475464}, {'id': 124, 'seek': 96700, 'start': 981.0, 'end': 990.0, 'text': ' So these stop and start codons and signals are critical for ribosomes to understand where the translation should be started and should be ended.', 'tokens': [51063, 1406, 777, 2245, 290, 923, 14873, 684, 290, 10425, 389, 4688, 329, 12183, 418, 2586, 284, 1833, 810, 262, 11059, 815, 307, 2067, 290, 815, 307, 4444, 13, 51513], 'temperature': 0.0, 'avg_logprob': -0.11341676046681959, 'compression_ratio': 1.647887323943662, 'no_speech_prob': 0.42656153440475464}, {'id': 125, 'seek': 99000, 'start': 990.0, 'end': 998.0, 'text': \" So let's take a look at this in a sort of an overview level from the whole process with this illustration.\", 'tokens': [50363, 1406, 1309, 338, 1011, 257, 804, 379, 428, 287, 257, 3297, 286, 281, 16700, 1241, 422, 262, 2187, 1429, 351, 428, 20936, 13, 50763], 'temperature': 0.0, 'avg_logprob': -0.12582741086445157, 'compression_ratio': 1.6145251396648044, 'no_speech_prob': 0.3834340274333954}, {'id': 126, 'seek': 99000, 'start': 998.0, 'end': 1009.0, 'text': ' Many of you probably have come across chemical polymerization, perhaps in high school, and you know, a lot of polymerization processes in chemistry and industry, chemical processing.', 'tokens': [50763, 4650, 286, 345, 2192, 423, 1282, 1973, 5931, 30528, 1634, 11, 3737, 287, 1029, 1524, 11, 290, 345, 760, 11, 257, 1256, 286, 30528, 1634, 7767, 287, 16585, 290, 2831, 11, 5931, 7587, 13, 51313], 'temperature': 0.0, 'avg_logprob': -0.12582741086445157, 'compression_ratio': 1.6145251396648044, 'no_speech_prob': 0.3834340274333954}, {'id': 127, 'seek': 100900, 'start': 1009.0, 'end': 1019.0, 'text': ' Biology in many cases work the same way. In any polymerization process, what do you have? There are typically three steps in any kind of polymerization.', 'tokens': [50363, 24698, 287, 867, 2663, 670, 262, 976, 835, 13, 554, 597, 30528, 1634, 1429, 11, 644, 466, 345, 423, 30, 1318, 389, 6032, 1115, 4831, 287, 597, 1611, 286, 30528, 1634, 13, 50863], 'temperature': 0.0, 'avg_logprob': -0.2726984715116197, 'compression_ratio': 1.6051282051282052, 'no_speech_prob': 0.8261743187904358}, {'id': 128, 'seek': 100900, 'start': 1019.0, 'end': 1030.0, 'text': \" Sometimes they change the activities, that's how we would encounter in USA is we encounter chemical polymerization. There are always three steps. What are they?\", 'tokens': [50863, 8975, 484, 1487, 262, 4568, 11, 326, 338, 703, 356, 561, 8791, 287, 4916, 318, 356, 8791, 5931, 30528, 1634, 13, 1318, 389, 1464, 1115, 4831, 13, 1867, 389, 484, 30, 51413], 'temperature': 0.0, 'avg_logprob': -0.2726984715116197, 'compression_ratio': 1.6051282051282052, 'no_speech_prob': 0.8261743187904358}, {'id': 129, 'seek': 103000, 'start': 1030.0, 'end': 1042.0, 'text': \" You have to initiate, you have to elongate, process, and then you have to terminate. Biology does the same thing, nature does the same thing. There's initiation step, elongation step, and termination step.\", 'tokens': [50363, 921, 423, 284, 22118, 11, 345, 423, 284, 39696, 378, 11, 1429, 11, 290, 788, 345, 423, 284, 23654, 13, 24698, 857, 262, 976, 1517, 11, 3450, 857, 262, 976, 1517, 13, 1318, 338, 30122, 2239, 11, 39696, 341, 2239, 11, 290, 19883, 2239, 13, 50963], 'temperature': 0.0, 'avg_logprob': -0.15458863258361816, 'compression_ratio': 1.7962962962962963, 'no_speech_prob': 0.2026246190071106}, {'id': 130, 'seek': 103000, 'start': 1042.0, 'end': 1057.0, 'text': \" So let's take a walkthrough of these individual steps. So what I'm showing you here is mRNA, 5' and 3', and you can see the stop codon, and stop codon is one of the stop codons here.\", 'tokens': [50963, 1406, 1309, 338, 1011, 257, 2513, 9579, 286, 777, 1981, 4831, 13, 1406, 644, 314, 1101, 4478, 345, 994, 318, 47227, 11, 642, 6, 290, 513, 3256, 290, 345, 460, 766, 262, 2245, 14873, 261, 11, 290, 2245, 14873, 261, 318, 530, 286, 262, 2245, 14873, 684, 994, 13, 51713], 'temperature': 0.0, 'avg_logprob': -0.15458863258361816, 'compression_ratio': 1.7962962962962963, 'no_speech_prob': 0.2026246190071106}, {'id': 131, 'seek': 105700, 'start': 1058.0, 'end': 1072.0, 'text': ' And ribosome essentially is made up of sort of two sub-major subunits, one large and one small. So not just having an mRNA and this ribosome machinery, obviously something has to bring in your individual amino acid.', 'tokens': [50413, 843, 12183, 418, 462, 6986, 318, 925, 510, 286, 3297, 286, 734, 850, 12, 22478, 850, 41667, 11, 530, 1588, 290, 530, 1402, 13, 1406, 407, 655, 1719, 281, 47227, 290, 428, 12183, 418, 462, 20230, 11, 6189, 1223, 468, 284, 2222, 287, 534, 1981, 23206, 7408, 13, 51113], 'temperature': 0.0, 'avg_logprob': -0.18110293608445388, 'compression_ratio': 1.396103896103896, 'no_speech_prob': 0.18400055170059204}, {'id': 132, 'seek': 107200, 'start': 1073.0, 'end': 1096.0, 'text': ' So how does nature do that in all living systems? And it turns out that there is this additional type of RNA, still nucleic acid, called transfer RNA, tRNA, which serves as an adapter to bring in the precise amino acid shown here in the purple color, to actually allow this polymerization process to happen in a coded manner.', 'tokens': [50413, 1406, 703, 857, 3450, 466, 326, 287, 477, 2877, 3341, 30, 843, 340, 4962, 503, 326, 612, 318, 428, 3224, 2099, 286, 25897, 11, 991, 17751, 291, 7408, 11, 1444, 4351, 25897, 11, 256, 27204, 11, 543, 9179, 355, 281, 21302, 284, 2222, 287, 262, 7141, 23206, 7408, 3402, 994, 287, 262, 14032, 3124, 11, 284, 1682, 1249, 428, 30528, 1634, 1429, 284, 1645, 287, 257, 30817, 5642, 13, 51563], 'temperature': 0.0, 'avg_logprob': -0.0989885395520354, 'compression_ratio': 1.5330188679245282, 'no_speech_prob': 0.4177020192146301}, {'id': 133, 'seek': 109600, 'start': 1097.0, 'end': 1106.0, 'text': ' I said coded manner because which amino acid to be brought in will be defined by the actual nucleotide sequence that is present along your mRNA.', 'tokens': [50413, 314, 531, 30817, 5642, 780, 543, 23206, 7408, 284, 307, 3181, 287, 481, 307, 5447, 416, 262, 4036, 17751, 45608, 8379, 326, 318, 1944, 1863, 534, 47227, 13, 50863], 'temperature': 0.0, 'avg_logprob': -0.15413781368371213, 'compression_ratio': 1.6602316602316602, 'no_speech_prob': 0.23855355381965637}, {'id': 134, 'seek': 109600, 'start': 1108.0, 'end': 1125.0, 'text': ' How do you think this position happens? So mRNA will have different nucleosides here, and I could have tRNA bringing in one amino acid at a time. How do I know how this system, how can you encode for the fact that if I hit this particular codon, I will bring in the correct amino acid?', 'tokens': [50963, 1374, 466, 345, 892, 428, 2292, 4325, 30, 1406, 47227, 481, 423, 1180, 17751, 418, 1460, 994, 11, 290, 314, 714, 423, 256, 27204, 6079, 287, 530, 23206, 7408, 379, 257, 640, 13, 1374, 466, 314, 760, 703, 428, 1080, 11, 703, 460, 345, 37773, 329, 262, 1109, 326, 611, 314, 2277, 428, 1948, 14873, 261, 11, 314, 481, 2222, 287, 262, 3376, 23206, 7408, 30, 51813], 'temperature': 0.0, 'avg_logprob': -0.15413781368371213, 'compression_ratio': 1.6602316602316602, 'no_speech_prob': 0.23855355381965637}]\n"
     ]
    }
   ],
   "source": [
    "transcription = joblib.load(\"output/transcription_google_colab.jl\")\n",
    "print(transcription[0].keys())\n",
    "print(transcription[0][\"segments\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   id  seek  start  end                                            text   \n",
      "0   0     0    0.0  1.0                                  Thank you all.  \\\n",
      "1   1     0    1.0  2.0                                 Good afternoon.   \n",
      "2   2     0    2.0  3.0                         How are you guys doing?   \n",
      "3   3     0    3.0  4.0                                   It looks fun.   \n",
      "4   4     0    4.0  7.0   How many fans have done themselves last week?   \n",
      "\n",
      "                                              tokens  temperature   \n",
      "0                 [50363, 6952, 345, 477, 13, 50413]          0.0  \\\n",
      "1                     [50413, 4599, 6672, 13, 50463]          0.0   \n",
      "2     [50463, 1374, 389, 345, 3730, 1804, 30, 50513]          0.0   \n",
      "3                [50513, 632, 3073, 1257, 13, 50563]          0.0   \n",
      "4  [50563, 1374, 867, 3296, 423, 1760, 2405, 938,...          0.0   \n",
      "\n",
      "   avg_logprob  compression_ratio  no_speech_prob  \n",
      "0    -0.454801           1.611354        0.274894  \n",
      "1    -0.454801           1.611354        0.274894  \n",
      "2    -0.454801           1.611354        0.274894  \n",
      "3    -0.454801           1.611354        0.274894  \n",
      "4    -0.454801           1.611354        0.274894  \n",
      " Many of you probably have come across chemical polymerization, perhaps in high school, and you know, a lot of polymerization processes in chemistry and industry, chemical processing.\n",
      " Biology in many cases work the same way. In any polymerization process, what do you have? There are typically three steps in any kind of polymerization.\n",
      " Sometimes they change the activities, that's how we would encounter in USA is we encounter chemical polymerization. There are always three steps. What are they?\n",
      " You have to initiate, you have to elongate, process, and then you have to terminate. Biology does the same thing, nature does the same thing. There's initiation step, elongation step, and termination step.\n",
      " So let's take a walkthrough of these individual steps. So what I'm showing you here is mRNA, 5' and 3', and you can see the stop codon, and stop codon is one of the stop codons here.\n",
      " And ribosome essentially is made up of sort of two sub-major subunits, one large and one small. So not just having an mRNA and this ribosome machinery, obviously something has to bring in your individual amino acid.\n",
      " So how does nature do that in all living systems? And it turns out that there is this additional type of RNA, still nucleic acid, called transfer RNA, tRNA, which serves as an adapter to bring in the precise amino acid shown here in the purple color, to actually allow this polymerization process to happen in a coded manner.\n",
      " I said coded manner because which amino acid to be brought in will be defined by the actual nucleotide sequence that is present along your mRNA.\n",
      " How do you think this position happens? So mRNA will have different nucleosides here, and I could have tRNA bringing in one amino acid at a time. How do I know how this system, how can you encode for the fact that if I hit this particular codon, I will bring in the correct amino acid?\n"
     ]
    }
   ],
   "source": [
    "subclip_0_df = pd.DataFrame(transcription[0][\"segments\"])\n",
    "print(subclip_0_df.head())\n",
    "for i in range(-9, 0):\n",
    "    print(subclip_0_df[\"text\"].iloc[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " How do you think this position happens?\n",
      " So mRNA will have different nucleoside fields\n",
      " and I could have tRNA bringing in one amino acid at a time.\n",
      " How do I know how does the system,\n",
      " how can you encode for the fact that,\n",
      " if I hit this particular code off,\n",
      " I will bring in the correct amino acid.\n",
      " There's chemistry behind it.\n",
      " What is this position mechanism?\n",
      " Yeah.\n"
     ]
    }
   ],
   "source": [
    "subclip_1_df = pd.DataFrame(transcription[1][\"segments\"])\n",
    "for i in range(10):\n",
    "    print(subclip_1_df[\"text\"].iloc[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1128</td>\n",
       "      <td>output/subclip_0_0_1128.mp3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1108</td>\n",
       "      <td>2246</td>\n",
       "      <td>output/subclip_1_1108_2246.mp3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2226</td>\n",
       "      <td>3364</td>\n",
       "      <td>output/subclip_2_2226_3364.mp3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3344</td>\n",
       "      <td>4482</td>\n",
       "      <td>output/subclip_3_3344_4482.mp3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4462</td>\n",
       "      <td>5587</td>\n",
       "      <td>output/subclip_4_4462_5587.mp3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  index  start   end                            path\n",
       "0     0      0  1128     output/subclip_0_0_1128.mp3\n",
       "1     1   1108  2246  output/subclip_1_1108_2246.mp3\n",
       "2     2   2226  3364  output/subclip_2_2226_3364.mp3\n",
       "3     3   3344  4482  output/subclip_3_3344_4482.mp3\n",
       "4     4   4462  5587  output/subclip_4_4462_5587.mp3"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subclips_list = joblib.load(\"output/subclips_list.jl\")\n",
    "subclip_info_df = pd.DataFrame(subclips_list)\n",
    "subclip_info_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'numpy.int64'>\n"
     ]
    }
   ],
   "source": [
    "print(type(subclip_info_df[\"start\"].iloc[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>seek</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>text</th>\n",
       "      <th>tokens</th>\n",
       "      <th>temperature</th>\n",
       "      <th>avg_logprob</th>\n",
       "      <th>compression_ratio</th>\n",
       "      <th>no_speech_prob</th>\n",
       "      <th>subclip_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>Thank you all.</td>\n",
       "      <td>[50363, 6952, 345, 477, 13, 50413]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.454801</td>\n",
       "      <td>1.611354</td>\n",
       "      <td>0.274894</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>Good afternoon.</td>\n",
       "      <td>[50413, 4599, 6672, 13, 50463]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.454801</td>\n",
       "      <td>1.611354</td>\n",
       "      <td>0.274894</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>3.00</td>\n",
       "      <td>How are you guys doing?</td>\n",
       "      <td>[50463, 1374, 389, 345, 3730, 1804, 30, 50513]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.454801</td>\n",
       "      <td>1.611354</td>\n",
       "      <td>0.274894</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3.00</td>\n",
       "      <td>4.00</td>\n",
       "      <td>It looks fun.</td>\n",
       "      <td>[50513, 632, 3073, 1257, 13, 50563]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.454801</td>\n",
       "      <td>1.611354</td>\n",
       "      <td>0.274894</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4.00</td>\n",
       "      <td>7.00</td>\n",
       "      <td>How many fans have done themselves last week?</td>\n",
       "      <td>[50563, 1374, 867, 3296, 423, 1760, 2405, 938,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.454801</td>\n",
       "      <td>1.611354</td>\n",
       "      <td>0.274894</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1480</th>\n",
       "      <td>162</td>\n",
       "      <td>108472</td>\n",
       "      <td>5553.12</td>\n",
       "      <td>5558.56</td>\n",
       "      <td>one more class to see you guys, but there wil...</td>\n",
       "      <td>[50683, 530, 517, 1398, 284, 766, 345, 3730, 1...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.208055</td>\n",
       "      <td>1.514894</td>\n",
       "      <td>0.001780</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1481</th>\n",
       "      <td>163</td>\n",
       "      <td>108472</td>\n",
       "      <td>5558.56</td>\n",
       "      <td>5567.44</td>\n",
       "      <td>game-changing systems that you'll learn next ...</td>\n",
       "      <td>[50955, 983, 12, 22954, 3341, 326, 345, 1183, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.208055</td>\n",
       "      <td>1.514894</td>\n",
       "      <td>0.001780</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1482</th>\n",
       "      <td>164</td>\n",
       "      <td>108472</td>\n",
       "      <td>5567.44</td>\n",
       "      <td>5571.68</td>\n",
       "      <td>If you have questions, bring to us. And if yo...</td>\n",
       "      <td>[51399, 1002, 345, 423, 2683, 11, 2222, 284, 5...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.208055</td>\n",
       "      <td>1.514894</td>\n",
       "      <td>0.001780</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1483</th>\n",
       "      <td>165</td>\n",
       "      <td>110968</td>\n",
       "      <td>5571.68</td>\n",
       "      <td>5576.64</td>\n",
       "      <td>you will find problems as really quite straig...</td>\n",
       "      <td>[50363, 345, 481, 1064, 2761, 355, 1107, 2407,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.245308</td>\n",
       "      <td>1.240310</td>\n",
       "      <td>0.044770</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1484</th>\n",
       "      <td>166</td>\n",
       "      <td>110968</td>\n",
       "      <td>5576.64</td>\n",
       "      <td>5584.00</td>\n",
       "      <td>you guys again next Monday. Okay. Good. Have ...</td>\n",
       "      <td>[50611, 345, 3730, 757, 1306, 3321, 13, 16805,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.245308</td>\n",
       "      <td>1.240310</td>\n",
       "      <td>0.044770</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1485 rows  11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       id    seek    start      end   \n",
       "0       0       0     0.00     1.00  \\\n",
       "1       1       0     1.00     2.00   \n",
       "2       2       0     2.00     3.00   \n",
       "3       3       0     3.00     4.00   \n",
       "4       4       0     4.00     7.00   \n",
       "...   ...     ...      ...      ...   \n",
       "1480  162  108472  5553.12  5558.56   \n",
       "1481  163  108472  5558.56  5567.44   \n",
       "1482  164  108472  5567.44  5571.68   \n",
       "1483  165  110968  5571.68  5576.64   \n",
       "1484  166  110968  5576.64  5584.00   \n",
       "\n",
       "                                                   text   \n",
       "0                                        Thank you all.  \\\n",
       "1                                       Good afternoon.   \n",
       "2                               How are you guys doing?   \n",
       "3                                         It looks fun.   \n",
       "4         How many fans have done themselves last week?   \n",
       "...                                                 ...   \n",
       "1480   one more class to see you guys, but there wil...   \n",
       "1481   game-changing systems that you'll learn next ...   \n",
       "1482   If you have questions, bring to us. And if yo...   \n",
       "1483   you will find problems as really quite straig...   \n",
       "1484   you guys again next Monday. Okay. Good. Have ...   \n",
       "\n",
       "                                                 tokens  temperature   \n",
       "0                    [50363, 6952, 345, 477, 13, 50413]          0.0  \\\n",
       "1                        [50413, 4599, 6672, 13, 50463]          0.0   \n",
       "2        [50463, 1374, 389, 345, 3730, 1804, 30, 50513]          0.0   \n",
       "3                   [50513, 632, 3073, 1257, 13, 50563]          0.0   \n",
       "4     [50563, 1374, 867, 3296, 423, 1760, 2405, 938,...          0.0   \n",
       "...                                                 ...          ...   \n",
       "1480  [50683, 530, 517, 1398, 284, 766, 345, 3730, 1...          0.0   \n",
       "1481  [50955, 983, 12, 22954, 3341, 326, 345, 1183, ...          0.0   \n",
       "1482  [51399, 1002, 345, 423, 2683, 11, 2222, 284, 5...          0.0   \n",
       "1483  [50363, 345, 481, 1064, 2761, 355, 1107, 2407,...          0.0   \n",
       "1484  [50611, 345, 3730, 757, 1306, 3321, 13, 16805,...          0.0   \n",
       "\n",
       "      avg_logprob  compression_ratio  no_speech_prob  subclip_index  \n",
       "0       -0.454801           1.611354        0.274894              0  \n",
       "1       -0.454801           1.611354        0.274894              0  \n",
       "2       -0.454801           1.611354        0.274894              0  \n",
       "3       -0.454801           1.611354        0.274894              0  \n",
       "4       -0.454801           1.611354        0.274894              0  \n",
       "...           ...                ...             ...            ...  \n",
       "1480    -0.208055           1.514894        0.001780              4  \n",
       "1481    -0.208055           1.514894        0.001780              4  \n",
       "1482    -0.208055           1.514894        0.001780              4  \n",
       "1483    -0.245308           1.240310        0.044770              4  \n",
       "1484    -0.245308           1.240310        0.044770              4  \n",
       "\n",
       "[1485 rows x 11 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subclip_0_df[\"subclip_index\"] = int(subclip_info_df[\"index\"].iloc[0])\n",
    "subclip_0_df[\"start\"] += subclip_info_df[\"start\"].iloc[0]\n",
    "subclip_0_df[\"end\"] += subclip_info_df[\"start\"].iloc[0]\n",
    "\n",
    "for i in range(1, len(transcription)):\n",
    "    subclip_i_df = pd.DataFrame(transcription[i][\"segments\"])\n",
    "    subclip_i_df[\"subclip_index\"] = int(subclip_info_df[\"index\"].iloc[i])\n",
    "    subclip_i_df[\"start\"] += subclip_info_df[\"start\"].iloc[i]\n",
    "    subclip_i_df[\"end\"] += subclip_info_df[\"start\"].iloc[i]\n",
    "    subclip_0_df = pd.concat([subclip_0_df, subclip_i_df], ignore_index=True)\n",
    "subclip_0_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>seek</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>text</th>\n",
       "      <th>tokens</th>\n",
       "      <th>temperature</th>\n",
       "      <th>avg_logprob</th>\n",
       "      <th>compression_ratio</th>\n",
       "      <th>no_speech_prob</th>\n",
       "      <th>subclip_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>133</td>\n",
       "      <td>109600</td>\n",
       "      <td>1097.00</td>\n",
       "      <td>1106.00</td>\n",
       "      <td>I said coded manner because which amino acid ...</td>\n",
       "      <td>[50413, 314, 531, 30817, 5642, 780, 543, 23206...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.154138</td>\n",
       "      <td>1.660232</td>\n",
       "      <td>0.238554</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>134</td>\n",
       "      <td>109600</td>\n",
       "      <td>1108.00</td>\n",
       "      <td>1125.00</td>\n",
       "      <td>How do you think this position happens? So mR...</td>\n",
       "      <td>[50963, 1374, 466, 345, 892, 428, 2292, 4325, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.154138</td>\n",
       "      <td>1.660232</td>\n",
       "      <td>0.238554</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1108.00</td>\n",
       "      <td>1110.24</td>\n",
       "      <td>How do you think this position happens?</td>\n",
       "      <td>[50363, 1374, 466, 345, 892, 428, 2292, 4325, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1110.24</td>\n",
       "      <td>1112.96</td>\n",
       "      <td>So mRNA will have different nucleoside fields</td>\n",
       "      <td>[50475, 1406, 47227, 481, 423, 1180, 17751, 41...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1112.96</td>\n",
       "      <td>1116.92</td>\n",
       "      <td>and I could have tRNA bringing in one amino a...</td>\n",
       "      <td>[50611, 290, 314, 714, 423, 256, 27204, 6079, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1116.92</td>\n",
       "      <td>1119.02</td>\n",
       "      <td>How do I know how does the system,</td>\n",
       "      <td>[50809, 1374, 466, 314, 760, 703, 857, 262, 10...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1119.02</td>\n",
       "      <td>1121.20</td>\n",
       "      <td>how can you encode for the fact that,</td>\n",
       "      <td>[50914, 703, 460, 345, 37773, 329, 262, 1109, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1122.42</td>\n",
       "      <td>1124.52</td>\n",
       "      <td>if I hit this particular code off,</td>\n",
       "      <td>[51084, 611, 314, 2277, 428, 1948, 2438, 572, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1124.52</td>\n",
       "      <td>1126.74</td>\n",
       "      <td>I will bring in the correct amino acid.</td>\n",
       "      <td>[51189, 314, 481, 2222, 287, 262, 3376, 23206,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>1128.36</td>\n",
       "      <td>1129.60</td>\n",
       "      <td>There's chemistry behind it.</td>\n",
       "      <td>[51381, 1318, 338, 16585, 2157, 340, 13, 51443]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>1129.60</td>\n",
       "      <td>1132.48</td>\n",
       "      <td>What is this position mechanism?</td>\n",
       "      <td>[51443, 1867, 318, 428, 2292, 9030, 30, 51587]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>1132.48</td>\n",
       "      <td>1133.30</td>\n",
       "      <td>Yeah.</td>\n",
       "      <td>[51587, 9425, 13, 51628]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id    seek    start      end   \n",
       "133  133  109600  1097.00  1106.00  \\\n",
       "134  134  109600  1108.00  1125.00   \n",
       "135    0       0  1108.00  1110.24   \n",
       "136    1       0  1110.24  1112.96   \n",
       "137    2       0  1112.96  1116.92   \n",
       "138    3       0  1116.92  1119.02   \n",
       "139    4       0  1119.02  1121.20   \n",
       "140    5       0  1122.42  1124.52   \n",
       "141    6       0  1124.52  1126.74   \n",
       "142    7       0  1128.36  1129.60   \n",
       "143    8       0  1129.60  1132.48   \n",
       "144    9       0  1132.48  1133.30   \n",
       "\n",
       "                                                  text   \n",
       "133   I said coded manner because which amino acid ...  \\\n",
       "134   How do you think this position happens? So mR...   \n",
       "135            How do you think this position happens?   \n",
       "136      So mRNA will have different nucleoside fields   \n",
       "137   and I could have tRNA bringing in one amino a...   \n",
       "138                 How do I know how does the system,   \n",
       "139              how can you encode for the fact that,   \n",
       "140                 if I hit this particular code off,   \n",
       "141            I will bring in the correct amino acid.   \n",
       "142                       There's chemistry behind it.   \n",
       "143                   What is this position mechanism?   \n",
       "144                                              Yeah.   \n",
       "\n",
       "                                                tokens  temperature   \n",
       "133  [50413, 314, 531, 30817, 5642, 780, 543, 23206...          0.0  \\\n",
       "134  [50963, 1374, 466, 345, 892, 428, 2292, 4325, ...          0.0   \n",
       "135  [50363, 1374, 466, 345, 892, 428, 2292, 4325, ...          0.0   \n",
       "136  [50475, 1406, 47227, 481, 423, 1180, 17751, 41...          0.0   \n",
       "137  [50611, 290, 314, 714, 423, 256, 27204, 6079, ...          0.0   \n",
       "138  [50809, 1374, 466, 314, 760, 703, 857, 262, 10...          0.0   \n",
       "139  [50914, 703, 460, 345, 37773, 329, 262, 1109, ...          0.0   \n",
       "140  [51084, 611, 314, 2277, 428, 1948, 2438, 572, ...          0.0   \n",
       "141  [51189, 314, 481, 2222, 287, 262, 3376, 23206,...          0.0   \n",
       "142    [51381, 1318, 338, 16585, 2157, 340, 13, 51443]          0.0   \n",
       "143     [51443, 1867, 318, 428, 2292, 9030, 30, 51587]          0.0   \n",
       "144                           [51587, 9425, 13, 51628]          0.0   \n",
       "\n",
       "     avg_logprob  compression_ratio  no_speech_prob  subclip_index  \n",
       "133    -0.154138           1.660232        0.238554              0  \n",
       "134    -0.154138           1.660232        0.238554              0  \n",
       "135    -0.299951           1.581395        0.019813              1  \n",
       "136    -0.299951           1.581395        0.019813              1  \n",
       "137    -0.299951           1.581395        0.019813              1  \n",
       "138    -0.299951           1.581395        0.019813              1  \n",
       "139    -0.299951           1.581395        0.019813              1  \n",
       "140    -0.299951           1.581395        0.019813              1  \n",
       "141    -0.299951           1.581395        0.019813              1  \n",
       "142    -0.299951           1.581395        0.019813              1  \n",
       "143    -0.299951           1.581395        0.019813              1  \n",
       "144    -0.299951           1.581395        0.019813              1  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subclip_0_df[133:145]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>seek</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>text</th>\n",
       "      <th>tokens</th>\n",
       "      <th>temperature</th>\n",
       "      <th>avg_logprob</th>\n",
       "      <th>compression_ratio</th>\n",
       "      <th>no_speech_prob</th>\n",
       "      <th>subclip_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>395</td>\n",
       "      <td>111420</td>\n",
       "      <td>2226.72</td>\n",
       "      <td>2230.44</td>\n",
       "      <td>this five prime capping, this sort of a squar...</td>\n",
       "      <td>[50589, 428, 1936, 6994, 1275, 2105, 11, 428, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>396</td>\n",
       "      <td>111420</td>\n",
       "      <td>2230.44</td>\n",
       "      <td>2232.52</td>\n",
       "      <td>and then poly-A tail tagging.</td>\n",
       "      <td>[50775, 290, 788, 7514, 12, 32, 7894, 49620, 1...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>532</th>\n",
       "      <td>397</td>\n",
       "      <td>111420</td>\n",
       "      <td>2232.52</td>\n",
       "      <td>2233.72</td>\n",
       "      <td>So maybe on the footnote,</td>\n",
       "      <td>[50879, 1406, 3863, 319, 262, 38512, 11, 50939]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>398</td>\n",
       "      <td>111420</td>\n",
       "      <td>2233.72</td>\n",
       "      <td>2236.32</td>\n",
       "      <td>I mentioned five prime cap is a purple square,</td>\n",
       "      <td>[50939, 314, 4750, 1936, 6994, 1451, 318, 257,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534</th>\n",
       "      <td>399</td>\n",
       "      <td>111420</td>\n",
       "      <td>2236.32</td>\n",
       "      <td>2238.68</td>\n",
       "      <td>and then A is, if you see multiple A's,</td>\n",
       "      <td>[51069, 290, 788, 317, 318, 11, 611, 345, 766,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>400</td>\n",
       "      <td>111420</td>\n",
       "      <td>2238.68</td>\n",
       "      <td>2240.36</td>\n",
       "      <td>that's essentially poly-A tail.</td>\n",
       "      <td>[51187, 326, 338, 6986, 7514, 12, 32, 7894, 13...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>401</td>\n",
       "      <td>111420</td>\n",
       "      <td>2240.36</td>\n",
       "      <td>2241.52</td>\n",
       "      <td>So that already tells you,</td>\n",
       "      <td>[51271, 1406, 326, 1541, 4952, 345, 11, 51329]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537</th>\n",
       "      <td>402</td>\n",
       "      <td>111420</td>\n",
       "      <td>2241.52</td>\n",
       "      <td>2245.16</td>\n",
       "      <td>even if these labels are missing,</td>\n",
       "      <td>[51329, 772, 611, 777, 14722, 389, 4814, 11, 5...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>538</th>\n",
       "      <td>403</td>\n",
       "      <td>111420</td>\n",
       "      <td>2245.16</td>\n",
       "      <td>2246.32</td>\n",
       "      <td>you know that the...</td>\n",
       "      <td>[51511, 345, 760, 326, 262, 986, 51569]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2226.00</td>\n",
       "      <td>2230.32</td>\n",
       "      <td>this five prime capping, this sort of a squar...</td>\n",
       "      <td>[50363, 428, 1936, 6994, 1275, 2105, 11, 428, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2230.32</td>\n",
       "      <td>2232.40</td>\n",
       "      <td>and then poly a tail tagging.</td>\n",
       "      <td>[50579, 290, 788, 7514, 257, 7894, 49620, 13, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2232.40</td>\n",
       "      <td>2233.60</td>\n",
       "      <td>So maybe on the footnote,</td>\n",
       "      <td>[50683, 1406, 3863, 319, 262, 38512, 11, 50743]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2233.60</td>\n",
       "      <td>2236.20</td>\n",
       "      <td>I mentioned five prime cap is a purple square.</td>\n",
       "      <td>[50743, 314, 4750, 1936, 6994, 1451, 318, 257,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>2236.20</td>\n",
       "      <td>2238.56</td>\n",
       "      <td>And then a is, if you see multiple a's,</td>\n",
       "      <td>[50873, 843, 788, 257, 318, 11, 611, 345, 766,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>2238.56</td>\n",
       "      <td>2240.24</td>\n",
       "      <td>that's essentially poly a tail.</td>\n",
       "      <td>[50991, 326, 338, 6986, 7514, 257, 7894, 13, 5...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id    seek    start      end   \n",
       "530  395  111420  2226.72  2230.44  \\\n",
       "531  396  111420  2230.44  2232.52   \n",
       "532  397  111420  2232.52  2233.72   \n",
       "533  398  111420  2233.72  2236.32   \n",
       "534  399  111420  2236.32  2238.68   \n",
       "535  400  111420  2238.68  2240.36   \n",
       "536  401  111420  2240.36  2241.52   \n",
       "537  402  111420  2241.52  2245.16   \n",
       "538  403  111420  2245.16  2246.32   \n",
       "539    0       0  2226.00  2230.32   \n",
       "540    1       0  2230.32  2232.40   \n",
       "541    2       0  2232.40  2233.60   \n",
       "542    3       0  2233.60  2236.20   \n",
       "543    4       0  2236.20  2238.56   \n",
       "544    5       0  2238.56  2240.24   \n",
       "\n",
       "                                                  text   \n",
       "530   this five prime capping, this sort of a squar...  \\\n",
       "531                      and then poly-A tail tagging.   \n",
       "532                          So maybe on the footnote,   \n",
       "533     I mentioned five prime cap is a purple square,   \n",
       "534            and then A is, if you see multiple A's,   \n",
       "535                    that's essentially poly-A tail.   \n",
       "536                         So that already tells you,   \n",
       "537                  even if these labels are missing,   \n",
       "538                               you know that the...   \n",
       "539   this five prime capping, this sort of a squar...   \n",
       "540                      and then poly a tail tagging.   \n",
       "541                          So maybe on the footnote,   \n",
       "542     I mentioned five prime cap is a purple square.   \n",
       "543            And then a is, if you see multiple a's,   \n",
       "544                    that's essentially poly a tail.   \n",
       "\n",
       "                                                tokens  temperature   \n",
       "530  [50589, 428, 1936, 6994, 1275, 2105, 11, 428, ...          0.0  \\\n",
       "531  [50775, 290, 788, 7514, 12, 32, 7894, 49620, 1...          0.0   \n",
       "532    [50879, 1406, 3863, 319, 262, 38512, 11, 50939]          0.0   \n",
       "533  [50939, 314, 4750, 1936, 6994, 1451, 318, 257,...          0.0   \n",
       "534  [51069, 290, 788, 317, 318, 11, 611, 345, 766,...          0.0   \n",
       "535  [51187, 326, 338, 6986, 7514, 12, 32, 7894, 13...          0.0   \n",
       "536     [51271, 1406, 326, 1541, 4952, 345, 11, 51329]          0.0   \n",
       "537  [51329, 772, 611, 777, 14722, 389, 4814, 11, 5...          0.0   \n",
       "538            [51511, 345, 760, 326, 262, 986, 51569]          0.0   \n",
       "539  [50363, 428, 1936, 6994, 1275, 2105, 11, 428, ...          0.0   \n",
       "540  [50579, 290, 788, 7514, 257, 7894, 49620, 13, ...          0.0   \n",
       "541    [50683, 1406, 3863, 319, 262, 38512, 11, 50743]          0.0   \n",
       "542  [50743, 314, 4750, 1936, 6994, 1451, 318, 257,...          0.0   \n",
       "543  [50873, 843, 788, 257, 318, 11, 611, 345, 766,...          0.0   \n",
       "544  [50991, 326, 338, 6986, 7514, 257, 7894, 13, 5...          0.0   \n",
       "\n",
       "     avg_logprob  compression_ratio  no_speech_prob  subclip_index  \n",
       "530    -0.220976           1.672489        0.004981              1  \n",
       "531    -0.220976           1.672489        0.004981              1  \n",
       "532    -0.220976           1.672489        0.004981              1  \n",
       "533    -0.220976           1.672489        0.004981              1  \n",
       "534    -0.220976           1.672489        0.004981              1  \n",
       "535    -0.220976           1.672489        0.004981              1  \n",
       "536    -0.220976           1.672489        0.004981              1  \n",
       "537    -0.220976           1.672489        0.004981              1  \n",
       "538    -0.220976           1.672489        0.004981              1  \n",
       "539    -0.349091           1.596639        0.241382              2  \n",
       "540    -0.349091           1.596639        0.241382              2  \n",
       "541    -0.349091           1.596639        0.241382              2  \n",
       "542    -0.349091           1.596639        0.241382              2  \n",
       "543    -0.349091           1.596639        0.241382              2  \n",
       "544    -0.349091           1.596639        0.241382              2  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subclip_0_df[530:545]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>seek</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>text</th>\n",
       "      <th>tokens</th>\n",
       "      <th>temperature</th>\n",
       "      <th>avg_logprob</th>\n",
       "      <th>compression_ratio</th>\n",
       "      <th>no_speech_prob</th>\n",
       "      <th>subclip_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>528</th>\n",
       "      <td>393</td>\n",
       "      <td>111420</td>\n",
       "      <td>2222.24</td>\n",
       "      <td>2224.44</td>\n",
       "      <td>they need to be processed and mature.</td>\n",
       "      <td>[50365, 484, 761, 284, 307, 13686, 290, 15345,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>394</td>\n",
       "      <td>111420</td>\n",
       "      <td>2224.44</td>\n",
       "      <td>2226.72</td>\n",
       "      <td>And this maturation process requires</td>\n",
       "      <td>[50475, 843, 428, 285, 36921, 1429, 4433, 50589]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2226.00</td>\n",
       "      <td>2230.32</td>\n",
       "      <td>this five prime capping, this sort of a squar...</td>\n",
       "      <td>[50363, 428, 1936, 6994, 1275, 2105, 11, 428, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>395</td>\n",
       "      <td>111420</td>\n",
       "      <td>2226.72</td>\n",
       "      <td>2230.44</td>\n",
       "      <td>this five prime capping, this sort of a squar...</td>\n",
       "      <td>[50589, 428, 1936, 6994, 1275, 2105, 11, 428, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2230.32</td>\n",
       "      <td>2232.40</td>\n",
       "      <td>and then poly a tail tagging.</td>\n",
       "      <td>[50579, 290, 788, 7514, 257, 7894, 49620, 13, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>396</td>\n",
       "      <td>111420</td>\n",
       "      <td>2230.44</td>\n",
       "      <td>2232.52</td>\n",
       "      <td>and then poly-A tail tagging.</td>\n",
       "      <td>[50775, 290, 788, 7514, 12, 32, 7894, 49620, 1...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2232.40</td>\n",
       "      <td>2233.60</td>\n",
       "      <td>So maybe on the footnote,</td>\n",
       "      <td>[50683, 1406, 3863, 319, 262, 38512, 11, 50743]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>532</th>\n",
       "      <td>397</td>\n",
       "      <td>111420</td>\n",
       "      <td>2232.52</td>\n",
       "      <td>2233.72</td>\n",
       "      <td>So maybe on the footnote,</td>\n",
       "      <td>[50879, 1406, 3863, 319, 262, 38512, 11, 50939]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2233.60</td>\n",
       "      <td>2236.20</td>\n",
       "      <td>I mentioned five prime cap is a purple square.</td>\n",
       "      <td>[50743, 314, 4750, 1936, 6994, 1451, 318, 257,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>398</td>\n",
       "      <td>111420</td>\n",
       "      <td>2233.72</td>\n",
       "      <td>2236.32</td>\n",
       "      <td>I mentioned five prime cap is a purple square,</td>\n",
       "      <td>[50939, 314, 4750, 1936, 6994, 1451, 318, 257,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>2236.20</td>\n",
       "      <td>2238.56</td>\n",
       "      <td>And then a is, if you see multiple a's,</td>\n",
       "      <td>[50873, 843, 788, 257, 318, 11, 611, 345, 766,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534</th>\n",
       "      <td>399</td>\n",
       "      <td>111420</td>\n",
       "      <td>2236.32</td>\n",
       "      <td>2238.68</td>\n",
       "      <td>and then A is, if you see multiple A's,</td>\n",
       "      <td>[51069, 290, 788, 317, 318, 11, 611, 345, 766,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>2238.56</td>\n",
       "      <td>2240.24</td>\n",
       "      <td>that's essentially poly a tail.</td>\n",
       "      <td>[50991, 326, 338, 6986, 7514, 257, 7894, 13, 5...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>400</td>\n",
       "      <td>111420</td>\n",
       "      <td>2238.68</td>\n",
       "      <td>2240.36</td>\n",
       "      <td>that's essentially poly-A tail.</td>\n",
       "      <td>[51187, 326, 338, 6986, 7514, 12, 32, 7894, 13...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>2240.24</td>\n",
       "      <td>2241.44</td>\n",
       "      <td>So that already tells you,</td>\n",
       "      <td>[51075, 1406, 326, 1541, 4952, 345, 11, 51135]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>401</td>\n",
       "      <td>111420</td>\n",
       "      <td>2240.36</td>\n",
       "      <td>2241.52</td>\n",
       "      <td>So that already tells you,</td>\n",
       "      <td>[51271, 1406, 326, 1541, 4952, 345, 11, 51329]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>546</th>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>2241.44</td>\n",
       "      <td>2245.08</td>\n",
       "      <td>even if these labels are missing,</td>\n",
       "      <td>[51135, 772, 611, 777, 14722, 389, 4814, 11, 5...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id    seek    start      end   \n",
       "528  393  111420  2222.24  2224.44  \\\n",
       "529  394  111420  2224.44  2226.72   \n",
       "539    0       0  2226.00  2230.32   \n",
       "530  395  111420  2226.72  2230.44   \n",
       "540    1       0  2230.32  2232.40   \n",
       "531  396  111420  2230.44  2232.52   \n",
       "541    2       0  2232.40  2233.60   \n",
       "532  397  111420  2232.52  2233.72   \n",
       "542    3       0  2233.60  2236.20   \n",
       "533  398  111420  2233.72  2236.32   \n",
       "543    4       0  2236.20  2238.56   \n",
       "534  399  111420  2236.32  2238.68   \n",
       "544    5       0  2238.56  2240.24   \n",
       "535  400  111420  2238.68  2240.36   \n",
       "545    6       0  2240.24  2241.44   \n",
       "536  401  111420  2240.36  2241.52   \n",
       "546    7       0  2241.44  2245.08   \n",
       "\n",
       "                                                  text   \n",
       "528              they need to be processed and mature.  \\\n",
       "529               And this maturation process requires   \n",
       "539   this five prime capping, this sort of a squar...   \n",
       "530   this five prime capping, this sort of a squar...   \n",
       "540                      and then poly a tail tagging.   \n",
       "531                      and then poly-A tail tagging.   \n",
       "541                          So maybe on the footnote,   \n",
       "532                          So maybe on the footnote,   \n",
       "542     I mentioned five prime cap is a purple square.   \n",
       "533     I mentioned five prime cap is a purple square,   \n",
       "543            And then a is, if you see multiple a's,   \n",
       "534            and then A is, if you see multiple A's,   \n",
       "544                    that's essentially poly a tail.   \n",
       "535                    that's essentially poly-A tail.   \n",
       "545                         So that already tells you,   \n",
       "536                         So that already tells you,   \n",
       "546                  even if these labels are missing,   \n",
       "\n",
       "                                                tokens  temperature   \n",
       "528  [50365, 484, 761, 284, 307, 13686, 290, 15345,...          0.0  \\\n",
       "529   [50475, 843, 428, 285, 36921, 1429, 4433, 50589]          0.0   \n",
       "539  [50363, 428, 1936, 6994, 1275, 2105, 11, 428, ...          0.0   \n",
       "530  [50589, 428, 1936, 6994, 1275, 2105, 11, 428, ...          0.0   \n",
       "540  [50579, 290, 788, 7514, 257, 7894, 49620, 13, ...          0.0   \n",
       "531  [50775, 290, 788, 7514, 12, 32, 7894, 49620, 1...          0.0   \n",
       "541    [50683, 1406, 3863, 319, 262, 38512, 11, 50743]          0.0   \n",
       "532    [50879, 1406, 3863, 319, 262, 38512, 11, 50939]          0.0   \n",
       "542  [50743, 314, 4750, 1936, 6994, 1451, 318, 257,...          0.0   \n",
       "533  [50939, 314, 4750, 1936, 6994, 1451, 318, 257,...          0.0   \n",
       "543  [50873, 843, 788, 257, 318, 11, 611, 345, 766,...          0.0   \n",
       "534  [51069, 290, 788, 317, 318, 11, 611, 345, 766,...          0.0   \n",
       "544  [50991, 326, 338, 6986, 7514, 257, 7894, 13, 5...          0.0   \n",
       "535  [51187, 326, 338, 6986, 7514, 12, 32, 7894, 13...          0.0   \n",
       "545     [51075, 1406, 326, 1541, 4952, 345, 11, 51135]          0.0   \n",
       "536     [51271, 1406, 326, 1541, 4952, 345, 11, 51329]          0.0   \n",
       "546  [51135, 772, 611, 777, 14722, 389, 4814, 11, 5...          0.0   \n",
       "\n",
       "     avg_logprob  compression_ratio  no_speech_prob  subclip_index  \n",
       "528    -0.220976           1.672489        0.004981              1  \n",
       "529    -0.220976           1.672489        0.004981              1  \n",
       "539    -0.349091           1.596639        0.241382              2  \n",
       "530    -0.220976           1.672489        0.004981              1  \n",
       "540    -0.349091           1.596639        0.241382              2  \n",
       "531    -0.220976           1.672489        0.004981              1  \n",
       "541    -0.349091           1.596639        0.241382              2  \n",
       "532    -0.220976           1.672489        0.004981              1  \n",
       "542    -0.349091           1.596639        0.241382              2  \n",
       "533    -0.220976           1.672489        0.004981              1  \n",
       "543    -0.349091           1.596639        0.241382              2  \n",
       "534    -0.220976           1.672489        0.004981              1  \n",
       "544    -0.349091           1.596639        0.241382              2  \n",
       "535    -0.220976           1.672489        0.004981              1  \n",
       "545    -0.349091           1.596639        0.241382              2  \n",
       "536    -0.220976           1.672489        0.004981              1  \n",
       "546    -0.349091           1.596639        0.241382              2  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subclip_0_df.sort_values(by=[\"start\"]).iloc[528:545]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/xd/10nyn_j13qs58rww11sjkgy80000gn/T/ipykernel_98810/1052025233.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  subclip_0_df[\"start_diff\"].iloc[0] = 0\n"
     ]
    }
   ],
   "source": [
    "subclip_0_df = subclip_0_df.sort_values(by=[\"start\"])\n",
    "subclip_0_df = subclip_0_df.reset_index(drop=True)\n",
    "\n",
    "subclip_0_df[\"start_diff\"] = subclip_0_df[\"start\"].diff()\n",
    "subclip_0_df[\"start_diff\"].iloc[0] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>seek</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>text</th>\n",
       "      <th>tokens</th>\n",
       "      <th>temperature</th>\n",
       "      <th>avg_logprob</th>\n",
       "      <th>compression_ratio</th>\n",
       "      <th>no_speech_prob</th>\n",
       "      <th>subclip_index</th>\n",
       "      <th>start_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>528</th>\n",
       "      <td>393</td>\n",
       "      <td>111420</td>\n",
       "      <td>2222.24</td>\n",
       "      <td>2224.44</td>\n",
       "      <td>they need to be processed and mature.</td>\n",
       "      <td>[50365, 484, 761, 284, 307, 13686, 290, 15345,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "      <td>2.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>394</td>\n",
       "      <td>111420</td>\n",
       "      <td>2224.44</td>\n",
       "      <td>2226.72</td>\n",
       "      <td>And this maturation process requires</td>\n",
       "      <td>[50475, 843, 428, 285, 36921, 1429, 4433, 50589]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "      <td>2.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2226.00</td>\n",
       "      <td>2230.32</td>\n",
       "      <td>this five prime capping, this sort of a squar...</td>\n",
       "      <td>[50363, 428, 1936, 6994, 1275, 2105, 11, 428, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "      <td>1.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>395</td>\n",
       "      <td>111420</td>\n",
       "      <td>2226.72</td>\n",
       "      <td>2230.44</td>\n",
       "      <td>this five prime capping, this sort of a squar...</td>\n",
       "      <td>[50589, 428, 1936, 6994, 1275, 2105, 11, 428, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "      <td>0.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>532</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2230.32</td>\n",
       "      <td>2232.40</td>\n",
       "      <td>and then poly a tail tagging.</td>\n",
       "      <td>[50579, 290, 788, 7514, 257, 7894, 49620, 13, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "      <td>3.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>396</td>\n",
       "      <td>111420</td>\n",
       "      <td>2230.44</td>\n",
       "      <td>2232.52</td>\n",
       "      <td>and then poly-A tail tagging.</td>\n",
       "      <td>[50775, 290, 788, 7514, 12, 32, 7894, 49620, 1...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2232.40</td>\n",
       "      <td>2233.60</td>\n",
       "      <td>So maybe on the footnote,</td>\n",
       "      <td>[50683, 1406, 3863, 319, 262, 38512, 11, 50743]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "      <td>1.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>397</td>\n",
       "      <td>111420</td>\n",
       "      <td>2232.52</td>\n",
       "      <td>2233.72</td>\n",
       "      <td>So maybe on the footnote,</td>\n",
       "      <td>[50879, 1406, 3863, 319, 262, 38512, 11, 50939]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2233.60</td>\n",
       "      <td>2236.20</td>\n",
       "      <td>I mentioned five prime cap is a purple square.</td>\n",
       "      <td>[50743, 314, 4750, 1936, 6994, 1451, 318, 257,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "      <td>1.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537</th>\n",
       "      <td>398</td>\n",
       "      <td>111420</td>\n",
       "      <td>2233.72</td>\n",
       "      <td>2236.32</td>\n",
       "      <td>I mentioned five prime cap is a purple square,</td>\n",
       "      <td>[50939, 314, 4750, 1936, 6994, 1451, 318, 257,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>538</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>2236.20</td>\n",
       "      <td>2238.56</td>\n",
       "      <td>And then a is, if you see multiple a's,</td>\n",
       "      <td>[50873, 843, 788, 257, 318, 11, 611, 345, 766,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "      <td>2.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>399</td>\n",
       "      <td>111420</td>\n",
       "      <td>2236.32</td>\n",
       "      <td>2238.68</td>\n",
       "      <td>and then A is, if you see multiple A's,</td>\n",
       "      <td>[51069, 290, 788, 317, 318, 11, 611, 345, 766,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>2238.56</td>\n",
       "      <td>2240.24</td>\n",
       "      <td>that's essentially poly a tail.</td>\n",
       "      <td>[50991, 326, 338, 6986, 7514, 257, 7894, 13, 5...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "      <td>2.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>400</td>\n",
       "      <td>111420</td>\n",
       "      <td>2238.68</td>\n",
       "      <td>2240.36</td>\n",
       "      <td>that's essentially poly-A tail.</td>\n",
       "      <td>[51187, 326, 338, 6986, 7514, 12, 32, 7894, 13...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>2240.24</td>\n",
       "      <td>2241.44</td>\n",
       "      <td>So that already tells you,</td>\n",
       "      <td>[51075, 1406, 326, 1541, 4952, 345, 11, 51135]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "      <td>1.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>401</td>\n",
       "      <td>111420</td>\n",
       "      <td>2240.36</td>\n",
       "      <td>2241.52</td>\n",
       "      <td>So that already tells you,</td>\n",
       "      <td>[51271, 1406, 326, 1541, 4952, 345, 11, 51329]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.220976</td>\n",
       "      <td>1.672489</td>\n",
       "      <td>0.004981</td>\n",
       "      <td>1</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>2241.44</td>\n",
       "      <td>2245.08</td>\n",
       "      <td>even if these labels are missing,</td>\n",
       "      <td>[51135, 772, 611, 777, 14722, 389, 4814, 11, 5...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.349091</td>\n",
       "      <td>1.596639</td>\n",
       "      <td>0.241382</td>\n",
       "      <td>2</td>\n",
       "      <td>1.08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id    seek    start      end   \n",
       "528  393  111420  2222.24  2224.44  \\\n",
       "529  394  111420  2224.44  2226.72   \n",
       "530    0       0  2226.00  2230.32   \n",
       "531  395  111420  2226.72  2230.44   \n",
       "532    1       0  2230.32  2232.40   \n",
       "533  396  111420  2230.44  2232.52   \n",
       "534    2       0  2232.40  2233.60   \n",
       "535  397  111420  2232.52  2233.72   \n",
       "536    3       0  2233.60  2236.20   \n",
       "537  398  111420  2233.72  2236.32   \n",
       "538    4       0  2236.20  2238.56   \n",
       "539  399  111420  2236.32  2238.68   \n",
       "540    5       0  2238.56  2240.24   \n",
       "541  400  111420  2238.68  2240.36   \n",
       "542    6       0  2240.24  2241.44   \n",
       "543  401  111420  2240.36  2241.52   \n",
       "544    7       0  2241.44  2245.08   \n",
       "\n",
       "                                                  text   \n",
       "528              they need to be processed and mature.  \\\n",
       "529               And this maturation process requires   \n",
       "530   this five prime capping, this sort of a squar...   \n",
       "531   this five prime capping, this sort of a squar...   \n",
       "532                      and then poly a tail tagging.   \n",
       "533                      and then poly-A tail tagging.   \n",
       "534                          So maybe on the footnote,   \n",
       "535                          So maybe on the footnote,   \n",
       "536     I mentioned five prime cap is a purple square.   \n",
       "537     I mentioned five prime cap is a purple square,   \n",
       "538            And then a is, if you see multiple a's,   \n",
       "539            and then A is, if you see multiple A's,   \n",
       "540                    that's essentially poly a tail.   \n",
       "541                    that's essentially poly-A tail.   \n",
       "542                         So that already tells you,   \n",
       "543                         So that already tells you,   \n",
       "544                  even if these labels are missing,   \n",
       "\n",
       "                                                tokens  temperature   \n",
       "528  [50365, 484, 761, 284, 307, 13686, 290, 15345,...          0.0  \\\n",
       "529   [50475, 843, 428, 285, 36921, 1429, 4433, 50589]          0.0   \n",
       "530  [50363, 428, 1936, 6994, 1275, 2105, 11, 428, ...          0.0   \n",
       "531  [50589, 428, 1936, 6994, 1275, 2105, 11, 428, ...          0.0   \n",
       "532  [50579, 290, 788, 7514, 257, 7894, 49620, 13, ...          0.0   \n",
       "533  [50775, 290, 788, 7514, 12, 32, 7894, 49620, 1...          0.0   \n",
       "534    [50683, 1406, 3863, 319, 262, 38512, 11, 50743]          0.0   \n",
       "535    [50879, 1406, 3863, 319, 262, 38512, 11, 50939]          0.0   \n",
       "536  [50743, 314, 4750, 1936, 6994, 1451, 318, 257,...          0.0   \n",
       "537  [50939, 314, 4750, 1936, 6994, 1451, 318, 257,...          0.0   \n",
       "538  [50873, 843, 788, 257, 318, 11, 611, 345, 766,...          0.0   \n",
       "539  [51069, 290, 788, 317, 318, 11, 611, 345, 766,...          0.0   \n",
       "540  [50991, 326, 338, 6986, 7514, 257, 7894, 13, 5...          0.0   \n",
       "541  [51187, 326, 338, 6986, 7514, 12, 32, 7894, 13...          0.0   \n",
       "542     [51075, 1406, 326, 1541, 4952, 345, 11, 51135]          0.0   \n",
       "543     [51271, 1406, 326, 1541, 4952, 345, 11, 51329]          0.0   \n",
       "544  [51135, 772, 611, 777, 14722, 389, 4814, 11, 5...          0.0   \n",
       "\n",
       "     avg_logprob  compression_ratio  no_speech_prob  subclip_index  start_diff  \n",
       "528    -0.220976           1.672489        0.004981              1        2.56  \n",
       "529    -0.220976           1.672489        0.004981              1        2.20  \n",
       "530    -0.349091           1.596639        0.241382              2        1.56  \n",
       "531    -0.220976           1.672489        0.004981              1        0.72  \n",
       "532    -0.349091           1.596639        0.241382              2        3.60  \n",
       "533    -0.220976           1.672489        0.004981              1        0.12  \n",
       "534    -0.349091           1.596639        0.241382              2        1.96  \n",
       "535    -0.220976           1.672489        0.004981              1        0.12  \n",
       "536    -0.349091           1.596639        0.241382              2        1.08  \n",
       "537    -0.220976           1.672489        0.004981              1        0.12  \n",
       "538    -0.349091           1.596639        0.241382              2        2.48  \n",
       "539    -0.220976           1.672489        0.004981              1        0.12  \n",
       "540    -0.349091           1.596639        0.241382              2        2.24  \n",
       "541    -0.220976           1.672489        0.004981              1        0.12  \n",
       "542    -0.349091           1.596639        0.241382              2        1.56  \n",
       "543    -0.220976           1.672489        0.004981              1        0.12  \n",
       "544    -0.349091           1.596639        0.241382              2        1.08  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subclip_0_df.iloc[528:545]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([531, 533, 535, 537, 539, 541, 543, 545, 547], dtype='int64')"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subclip_0_df[\n",
    "    (subclip_0_df[\"start\"] > 2226.00) & (subclip_0_df[\"subclip_index\"] < 2)\n",
    "].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2226.0"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subclip_0_df[(subclip_0_df[\"subclip_index\"] == 2)][\"start\"].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop every row where start > 2226.00 and subclip_index < 1\n",
    "\n",
    "\n",
    "for i in range(1, len(transcription)):\n",
    "    min_start_i = subclip_0_df[(subclip_0_df[\"subclip_index\"] == i)][\"start\"].min()\n",
    "\n",
    "    subclip_0_df = subclip_0_df.drop(\n",
    "        subclip_0_df[\n",
    "            (subclip_0_df[\"start\"] > min_start_i) & (subclip_0_df[\"subclip_index\"] < i)\n",
    "        ].index\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>seek</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>text</th>\n",
       "      <th>tokens</th>\n",
       "      <th>temperature</th>\n",
       "      <th>avg_logprob</th>\n",
       "      <th>compression_ratio</th>\n",
       "      <th>no_speech_prob</th>\n",
       "      <th>subclip_index</th>\n",
       "      <th>start_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>133</td>\n",
       "      <td>109600</td>\n",
       "      <td>1097.00</td>\n",
       "      <td>1106.00</td>\n",
       "      <td>I said coded manner because which amino acid ...</td>\n",
       "      <td>[50413, 314, 531, 30817, 5642, 780, 543, 23206...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.154138</td>\n",
       "      <td>1.660232</td>\n",
       "      <td>0.238554</td>\n",
       "      <td>0</td>\n",
       "      <td>24.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>134</td>\n",
       "      <td>109600</td>\n",
       "      <td>1108.00</td>\n",
       "      <td>1125.00</td>\n",
       "      <td>How do you think this position happens? So mR...</td>\n",
       "      <td>[50963, 1374, 466, 345, 892, 428, 2292, 4325, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.154138</td>\n",
       "      <td>1.660232</td>\n",
       "      <td>0.238554</td>\n",
       "      <td>0</td>\n",
       "      <td>11.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1108.00</td>\n",
       "      <td>1110.24</td>\n",
       "      <td>How do you think this position happens?</td>\n",
       "      <td>[50363, 1374, 466, 345, 892, 428, 2292, 4325, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1110.24</td>\n",
       "      <td>1112.96</td>\n",
       "      <td>So mRNA will have different nucleoside fields</td>\n",
       "      <td>[50475, 1406, 47227, 481, 423, 1180, 17751, 41...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "      <td>2.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1112.96</td>\n",
       "      <td>1116.92</td>\n",
       "      <td>and I could have tRNA bringing in one amino a...</td>\n",
       "      <td>[50611, 290, 314, 714, 423, 256, 27204, 6079, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "      <td>2.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1116.92</td>\n",
       "      <td>1119.02</td>\n",
       "      <td>How do I know how does the system,</td>\n",
       "      <td>[50809, 1374, 466, 314, 760, 703, 857, 262, 10...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "      <td>3.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1119.02</td>\n",
       "      <td>1121.20</td>\n",
       "      <td>how can you encode for the fact that,</td>\n",
       "      <td>[50914, 703, 460, 345, 37773, 329, 262, 1109, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "      <td>2.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1122.42</td>\n",
       "      <td>1124.52</td>\n",
       "      <td>if I hit this particular code off,</td>\n",
       "      <td>[51084, 611, 314, 2277, 428, 1948, 2438, 572, ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "      <td>3.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1124.52</td>\n",
       "      <td>1126.74</td>\n",
       "      <td>I will bring in the correct amino acid.</td>\n",
       "      <td>[51189, 314, 481, 2222, 287, 262, 3376, 23206,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "      <td>2.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>1128.36</td>\n",
       "      <td>1129.60</td>\n",
       "      <td>There's chemistry behind it.</td>\n",
       "      <td>[51381, 1318, 338, 16585, 2157, 340, 13, 51443]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "      <td>3.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>1129.60</td>\n",
       "      <td>1132.48</td>\n",
       "      <td>What is this position mechanism?</td>\n",
       "      <td>[51443, 1867, 318, 428, 2292, 9030, 30, 51587]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "      <td>1.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>1132.48</td>\n",
       "      <td>1133.30</td>\n",
       "      <td>Yeah.</td>\n",
       "      <td>[51587, 9425, 13, 51628]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.299951</td>\n",
       "      <td>1.581395</td>\n",
       "      <td>0.019813</td>\n",
       "      <td>1</td>\n",
       "      <td>2.88</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id    seek    start      end   \n",
       "133  133  109600  1097.00  1106.00  \\\n",
       "134  134  109600  1108.00  1125.00   \n",
       "135    0       0  1108.00  1110.24   \n",
       "136    1       0  1110.24  1112.96   \n",
       "137    2       0  1112.96  1116.92   \n",
       "138    3       0  1116.92  1119.02   \n",
       "139    4       0  1119.02  1121.20   \n",
       "140    5       0  1122.42  1124.52   \n",
       "141    6       0  1124.52  1126.74   \n",
       "142    7       0  1128.36  1129.60   \n",
       "143    8       0  1129.60  1132.48   \n",
       "144    9       0  1132.48  1133.30   \n",
       "\n",
       "                                                  text   \n",
       "133   I said coded manner because which amino acid ...  \\\n",
       "134   How do you think this position happens? So mR...   \n",
       "135            How do you think this position happens?   \n",
       "136      So mRNA will have different nucleoside fields   \n",
       "137   and I could have tRNA bringing in one amino a...   \n",
       "138                 How do I know how does the system,   \n",
       "139              how can you encode for the fact that,   \n",
       "140                 if I hit this particular code off,   \n",
       "141            I will bring in the correct amino acid.   \n",
       "142                       There's chemistry behind it.   \n",
       "143                   What is this position mechanism?   \n",
       "144                                              Yeah.   \n",
       "\n",
       "                                                tokens  temperature   \n",
       "133  [50413, 314, 531, 30817, 5642, 780, 543, 23206...          0.0  \\\n",
       "134  [50963, 1374, 466, 345, 892, 428, 2292, 4325, ...          0.0   \n",
       "135  [50363, 1374, 466, 345, 892, 428, 2292, 4325, ...          0.0   \n",
       "136  [50475, 1406, 47227, 481, 423, 1180, 17751, 41...          0.0   \n",
       "137  [50611, 290, 314, 714, 423, 256, 27204, 6079, ...          0.0   \n",
       "138  [50809, 1374, 466, 314, 760, 703, 857, 262, 10...          0.0   \n",
       "139  [50914, 703, 460, 345, 37773, 329, 262, 1109, ...          0.0   \n",
       "140  [51084, 611, 314, 2277, 428, 1948, 2438, 572, ...          0.0   \n",
       "141  [51189, 314, 481, 2222, 287, 262, 3376, 23206,...          0.0   \n",
       "142    [51381, 1318, 338, 16585, 2157, 340, 13, 51443]          0.0   \n",
       "143     [51443, 1867, 318, 428, 2292, 9030, 30, 51587]          0.0   \n",
       "144                           [51587, 9425, 13, 51628]          0.0   \n",
       "\n",
       "     avg_logprob  compression_ratio  no_speech_prob  subclip_index  start_diff  \n",
       "133    -0.154138           1.660232        0.238554              0       24.00  \n",
       "134    -0.154138           1.660232        0.238554              0       11.00  \n",
       "135    -0.299951           1.581395        0.019813              1        0.00  \n",
       "136    -0.299951           1.581395        0.019813              1        2.24  \n",
       "137    -0.299951           1.581395        0.019813              1        2.72  \n",
       "138    -0.299951           1.581395        0.019813              1        3.96  \n",
       "139    -0.299951           1.581395        0.019813              1        2.10  \n",
       "140    -0.299951           1.581395        0.019813              1        3.40  \n",
       "141    -0.299951           1.581395        0.019813              1        2.10  \n",
       "142    -0.299951           1.581395        0.019813              1        3.84  \n",
       "143    -0.299951           1.581395        0.019813              1        1.24  \n",
       "144    -0.299951           1.581395        0.019813              1        2.88  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subclip_0_df[133:145]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[' Thank you all.',\n",
       " ' Good afternoon.',\n",
       " ' How are you guys doing?',\n",
       " ' It looks fun.',\n",
       " ' How many fans have done themselves last week?',\n",
       " ' Too easy?',\n",
       " ' At least, maybe you think, well, applied the concepts that we learned in the past.',\n",
       " ' I hope that we all have relevant problems, for modern students like you guys.',\n",
       " ' At least, I hope you had some fun from doing it and not finding it too challenging.',\n",
       " ' The final exam questions, they will be largely in the context of problem solving like this.',\n",
       " ' And the level of challenge will be suddenly a little bit less, perhaps,',\n",
       " ' because in the final exam you are kind of limited on it.',\n",
       " ' So the highest level of challenge you will see, and at least my point in the course,',\n",
       " ' will be represented by the problem sets questions.',\n",
       " ' So if you feel that we understand them and you can tackle them, especially after the session,',\n",
       " ' and you are engaged in the class and you really should not need to worry suddenly from the',\n",
       " \" three lecture series that I'm very happy to provide here for this course.\",\n",
       " \" So let's continue to today's topic too.\",\n",
       " \" We're going to learn about designing and applying these RNA switches.\",\n",
       " ' You probably know synthetic biologists are very much inspired by nature,',\n",
       " ' by how nature designs these circuits and controls cell decision making and cell responses,',\n",
       " ' all the way at the animal level.',\n",
       " ' And when we use these ideas, we really be able to control back cells and living culture models.',\n",
       " ' So we will focus on RNA level switches.',\n",
       " ' Last week it was therapeutics, as you guys recall.',\n",
       " ' So the key aspects that we will target will be first going through the principles of natural regulatory',\n",
       " \" RNA. That's kind of important fundamentals because we need to know how nature does these mechanisms,\",\n",
       " ' and how we can hijack these principles to really control the cell decision making.',\n",
       " ' And there will be three courses that we will concern with for our course level.',\n",
       " ' This will be regulators, ribozymes, RNA-based enzymes, and riboswitches.',\n",
       " ' And lastly, connected to the second topic, it will be small molecule ligand or drug-guided RNA conformational gating',\n",
       " ' that will allow you to achieve this reversible regulatory control at the level of these RNA switches.',\n",
       " \" And first, to cover everyone's kind of basic fundamental principles on ribosome protein translation.\",\n",
       " ' Pretty much last week, we talked about this central dogma, right, starting from transcription and translation,',\n",
       " \" and of course there's also this reverse transcription method that's happened in nature in certain contexts.\",\n",
       " ' But what about this second step to the I mentioned here, protein translation.',\n",
       " ' You guys know what machinery in any organism that does this translation, what is it called?',\n",
       " \" Machinery, biomolecular machine that does this translation. It's actually known as ribosome.\",\n",
       " \" So we'll talk a little bit about ribosome. So as you can see in the footnote, it's a protein synthesis machine in the cell.\",\n",
       " ' It is a molecular, macromolecular complex made up of protein subunits together with certain RNAs that you will see is a catalytic RNA',\n",
       " ' that essentially together, this machine allows you to convert the mRNA sequence to your protein sequence.',\n",
       " ' Proteins are made up of amino acids that essentially is a long polymer of individual 20-coding amino acids.',\n",
       " \" So you can imagine this machine, it doesn't have in all cells, including bacteria and bacteria, those bacteria like us,\",\n",
       " \" essentially it's a polymerization machine if you think about it, because you're going to allow amino acids, individual amino acids to be covalently linked.\",\n",
       " ' In peptide bonds, so-called A-migons, you have these long chain proteins that you know that proteins are absolutely essential macromolecules that actually drive the cellular engine, just like in a car engine.',\n",
       " ' So that is a very essential aspect of protein synthesis is a ribosome.',\n",
       " ' So before I go there, just give you an idea of, you know, you may wonder, where does protein synthesis happen?',\n",
       " \" What I'm showing you here is the architecture of a eukaryotic cell.\",\n",
       " ' Why do you think protein synthesis happens?',\n",
       " ' Which compartment?',\n",
       " ' In a very specific location, the majority is on the ER surface, but largely speaking, does it happen in the nucleus of the cytoplasm?',\n",
       " ' Which compartment?',\n",
       " ' Cytoplasm, right?',\n",
       " ' Whereas mRNA synthesis, where does it happen in this eukaryotic cell?',\n",
       " ' In the nucleus, right?',\n",
       " ' So there is this really compartmentalized segregation in eukaryotic system where mRNA synthesis happens in the nucleus,',\n",
       " ' and this mRNA undergoes maturation processing, ultimately get exported outside the nucleus into the cytoplasm,',\n",
       " ' where it will meet this machinery that I mentioned, ribosome, that will catalyze the polypeptide, i.e. amino acids polymerization, to give you the protein.',\n",
       " ' So these individual black dots, you may see better in your slides, these are actually ribosome, these sort of a complex of RNAs and proteins.',\n",
       " ' They are essentially type complex. They are individual ribosome complex molecules that are diffusing throughout the cytoplasm.',\n",
       " \" So protein synthesis actually happens in the cytoplasm. It is true, the student over there said, it's also concentrated on the ER, so these tubules, tubular green structures you see are endoplasmic reticulum.\",\n",
       " \" So we won't worry so much in our cost level, but essentially the key point is that protein synthesis happens outside the nucleus, broadly speaking in the cytoplasm, which is very different from bacteria system like E. coli.\",\n",
       " \" In E. coli, do we worry about cytoplasm versus nucleus? The answer is no, right? There's essentially translation and transcription couple process in E. coli, as we talked about last week.\",\n",
       " ' So you may wonder, right, how fast are these processes? If you think about it, you know, one second is very fast, right? And actually ribosome can polymerize 10, on average, 10 amino acids in a second, which is really very rapid, right?',\n",
       " \" As I'm talking about, even before I finish the sentence, it will be several seconds, right, perhaps. And ribosome has already made 10 amino acids linked together.\",\n",
       " ' And average protein is about 300 amino acids linked. So, if you compare ribosome rate that I mentioned, protein translation rate.',\n",
       " \" So what I'm showing you here, you don't need to memorize these numbers, but at least good to know at the level of, you know, back of the envelope number, what kind of rate that are we talking about, when you think about RNA polymerization versus amino acid polymerization.\",\n",
       " ' RNA polymerization gives you mRNA from DNA, amino acid polymerization gives you protein from mRNA. So what types of enzyme that we saw last week, what does the mRNA synthesis?',\n",
       " \" So a machinery that produces mRNA, what we call these types of enzyme as, are polymerases and the RNA polymerases, right, machinery that makes DNA will be DNA polymerases, right, because it's all polymerizing the biomolecular polymers.\",\n",
       " ' If you think about the rates in terms of actual DNA replication rate is in the bacteria world on the E. coli, for instance, versus mammalian cells such as, you know, cancer cell type of cancer cell line, HeLa cells, the replication timescale are quite different, right.',\n",
       " \" But at the fundamental level, at the nucleotide level, the actual rate of transcription is, largely speaking, identical. You can make per second, if I'm taking 10 to 100 nucleotides.\",\n",
       " ' But why then? Why is it that our cell division in human cells is almost a day, whereas cell division in E. coli and many bacteria are less than an hour?',\n",
       " ' Why is it that DNA replication overall process takes much longer in our cells compared to E. coli?',\n",
       " \" Yes, certainly it's more complicated. You have to do the group checking, editing, etc. But still, the key point also is the fact that in the gene size, on average gene size, human gene is a large, much larger than bacterial gene.\",\n",
       " ' Can you guys think about the fundamental reason why that is the case? We talked a bit about that last week. What is it that is intrinsically, as a stark difference in the genome level between E. coli genome and human genome?',\n",
       " \" Yeah, presence of non-coding genetic regions, so-called introns. Intons and exons, you don't have to worry about that in prokaryotic systems like E. coli or bacteria.\",\n",
       " \" But we don't have introns in E. coli, whereas we have introns. So that means that for a given gene, even if ultimately the protein sequence is identical between E. coli protein and human protein, the actual starting genomes will be quite different.\",\n",
       " ' Because in origin of the DNA in the human cell, we have introns and exons. Ultimately, introns are spliced out inside the nucleus, right, we talked about that last week, before you ultimately get to translated protein.',\n",
       " ' So even if the amino acid level sequence is identical, at the DNA level, there will be introns in the human cells and eukaryotic system that will make the whole trans DNA synthesis step taking much longer.',\n",
       " ' So replication process will be a lot longer. So that is certainly one key fundamental aspect of the eukaryotic cell, having a presence of introns.',\n",
       " \" Now, protein synthesis, which is the focus of today's discussion, as I said, the average translation rate is 10 amino acids given take per second.\",\n",
       " \" So by that time, why is it that we don't have to care about these introns and exons anymore by the time we're talking about ribosome translation?\",\n",
       " \" So the blue unit that I'm pointing out here is essentially a classical way to illustrate the ribosome because it has two subunits, a large one and a small one.\",\n",
       " ' So the green piece is the polypeptides of proteins that are being translated, are being synthesized.',\n",
       " ' And then the red piece is the incoming mRNA that is coding for what amino acids should be in the protein sequence.',\n",
       " \" So given this, why do you think that at the point that we talk about translation, we don't have this difference anymore?\",\n",
       " ' Yes, it is splicing. Absolutely.',\n",
       " ' So mRNA that we talk about here is already a mature messenger RNA.',\n",
       " ' Whereas the newly synthesized, the nascently synthesized, if you like, the mRNA up there at the point of the RNA polymerase doing the transcription will be the premature mRNA.',\n",
       " ' You also tackled that question in the problem set last week, where you need to go through splicing and export, you know, decoration with poly-A tail, etc.',\n",
       " ' before you can actually get to the mature mRNA that is ready for ribosome translation.',\n",
       " \" So remember, this has to be mature mRNA. At that point, it's identical.\",\n",
       " ' Introns no longer matter because they have been spliced out.',\n",
       " \" OK. So the take home message is there's differences in time scale, broadly speaking, and also in terms of the translation, efficiency is the same between the two systems.\",\n",
       " \" So let's think about bacterial mRNA and eukaryotic mRNA.\",\n",
       " \" What I'm showing you here is a domain structure, very simplified, but it gives you a generic feature.\",\n",
       " \" In terms of coding sequences, don't do the same for a given protein.\",\n",
       " \" Because we don't need to worry about introns anymore by the time you get to the mature mRNA.\",\n",
       " ' But what other striking differences do you see between the two systems?',\n",
       " \" Because we'll hijack quite a bit of these mRNA regulation and their individual domains and complements when we think about ribosome switch design and application,\",\n",
       " ' the later part of the topic. So what do you see beyond the coding sequence?',\n",
       " ' Coding sequence is the mRNA that will code for your amino acid sequence on the protein level.',\n",
       " ' You see untranslated region. So this is upstream untranslated region.',\n",
       " \" So 5' UTR, so untranslated region and downstream UTR.\",\n",
       " ' And those are present in both cases.',\n",
       " \" Those components are essential for gene regulation and stability of mRNA. We'll come to that.\",\n",
       " ' And that essentially happens in both bacteria and eukaryotic world.',\n",
       " \" But you see additional units such as poly-A tail and 5' prime capping.\",\n",
       " ' That is absolutely essential to introduce this poly-A tail tagging and then cap modification on the 5 prime.',\n",
       " ' Fourth, eukaryotic mRNA, because when you think about eukaryotic mRNA, they are first produced in the nucleus,',\n",
       " ' but they need to be chaperoned and exported in a safe manner to the cytoplasm where they will meet ribosome such that they can synthesize the protein.',\n",
       " ' If you have mRNA produced in the nucleus that are not stable and protected by nucleases degradation,',\n",
       " ' we talked a lot about RNA nucleases last lecture as a means of therapeutic mechanism.',\n",
       " ' Both nucleases could essentially chop up the RNA that is newly synthesized in the nucleus before it can even get to the cytoplasm.',\n",
       " ' So these accessories, 5 prime capping and poly-A tail tagging at the mRNA level is critical to stabilize and protect the eukaryotic mRNA such that they can be recognized by proteins that will bind to these units and actually chaperone them, escort them if you like, for mRNA export from the nucleus to the cytoplasm for protein translation.',\n",
       " ' Keep in mind of this kind of segregation of compartments that makes the system more complex.',\n",
       " ' That means you require more factors to stabilize these mRNAs, protect these mRNA, chaperone and export them safely to the cytoplasm.',\n",
       " \" So that's what you're seeing here.\",\n",
       " \" Sub-cellular loca will already talk about it. What I've just discussed, you can also read further in your own time in these texts.\",\n",
       " \" But let's go through a little bit more of what allows a ribosome to know you have a long piece of mRNA starting from the genome that's coming along.\",\n",
       " ' How does ribosome know when to start translating, translating and synthesizing the amino acid, if you make this polymeric chain for a protein, when to start this process and when to end this process?',\n",
       " ' So there are specific codes that you will be implemented in your mRNA in nature that allows a ribosome to signal that this must be a star site. Star codon is always AUG.',\n",
       " ' And then there will be the end site, there will be a star codon in nature, there are three star codons and these three nucleotides will signal for where the star site and the end point should be.',\n",
       " \" Because if you think about it, these are, you know, if they are altered in any way, you'll have a protein that are either truncated, i.e. too short, shorter than you desire, or maybe start in the long side.\",\n",
       " ' So these stop and start codons and signals are critical for ribosomes to understand where the translation should be started and should be ended.',\n",
       " \" So let's take a look at this in a sort of an overview level from the whole process with this illustration.\",\n",
       " ' Many of you probably have come across chemical polymerization, perhaps in high school, and you know, a lot of polymerization processes in chemistry and industry, chemical processing.',\n",
       " ' Biology in many cases work the same way. In any polymerization process, what do you have? There are typically three steps in any kind of polymerization.',\n",
       " \" Sometimes they change the activities, that's how we would encounter in USA is we encounter chemical polymerization. There are always three steps. What are they?\",\n",
       " \" You have to initiate, you have to elongate, process, and then you have to terminate. Biology does the same thing, nature does the same thing. There's initiation step, elongation step, and termination step.\",\n",
       " \" So let's take a walkthrough of these individual steps. So what I'm showing you here is mRNA, 5' and 3', and you can see the stop codon, and stop codon is one of the stop codons here.\",\n",
       " ' And ribosome essentially is made up of sort of two sub-major subunits, one large and one small. So not just having an mRNA and this ribosome machinery, obviously something has to bring in your individual amino acid.',\n",
       " ' So how does nature do that in all living systems? And it turns out that there is this additional type of RNA, still nucleic acid, called transfer RNA, tRNA, which serves as an adapter to bring in the precise amino acid shown here in the purple color, to actually allow this polymerization process to happen in a coded manner.',\n",
       " ' I said coded manner because which amino acid to be brought in will be defined by the actual nucleotide sequence that is present along your mRNA.',\n",
       " ' How do you think this position happens? So mRNA will have different nucleosides here, and I could have tRNA bringing in one amino acid at a time. How do I know how this system, how can you encode for the fact that if I hit this particular codon, I will bring in the correct amino acid?',\n",
       " ' How do you think this position happens?',\n",
       " ' So mRNA will have different nucleoside fields',\n",
       " ' and I could have tRNA bringing in one amino acid at a time.',\n",
       " ' How do I know how does the system,',\n",
       " ' how can you encode for the fact that,',\n",
       " ' if I hit this particular code off,',\n",
       " ' I will bring in the correct amino acid.',\n",
       " \" There's chemistry behind it.\",\n",
       " ' What is this position mechanism?',\n",
       " ' Yeah.',\n",
       " ' Base pairing, right?',\n",
       " ' So if you have very good,',\n",
       " ' if you have the nucleotide sequence,',\n",
       " ' you essentially have these codons on the mRNA',\n",
       " ' that will allow precise anti-codons through base pairing.',\n",
       " ' So you have several tRNA molecules already charged',\n",
       " ' and loaded with several amino acids',\n",
       " ' available in the cellular pool,',\n",
       " ' but they will keep coming and finding and dividing',\n",
       " ' to search for the right anti-codon, codon base pairing.',\n",
       " ' When the base pairing is correct,',\n",
       " ' these tRNAs will come to the specific site of this ribosome',\n",
       " ' and allow this amino acid, a new amino acid,',\n",
       " ' to come in right next to the polymerizing amino acid,',\n",
       " \" if you like, and you'll have this amino acid\",\n",
       " ' additional adding to extend this chain',\n",
       " ' that is being polymerized by the ribosome.',\n",
       " ' So ultimately, when ribosomes see the stop codon',\n",
       " ' on the mRNA sequence, such as UAG,',\n",
       " ' instead of uptaking the tRNA on this site,',\n",
       " \" you'll have the so-called proteins release factors\",\n",
       " ' that will come in and eject this particular polypeptide',\n",
       " ' out of the ribosome site.',\n",
       " ' In that way, your polymerization event has ended',\n",
       " ' and you have produced successfully one completed protein.',\n",
       " ' Okay, I hope that makes sense.',\n",
       " ' All right, very good.',\n",
       " \" So let's take a look at a little bit further details.\",\n",
       " ' So these are essentially different codons that you have.',\n",
       " ' If you want mnemonic, when I was a student,',\n",
       " ' I used to kind of, you are gone, you go away,',\n",
       " \" and you are away if you're worried about these stop codons,\",\n",
       " \" but I don't expect you guys to memorize\",\n",
       " ' more for problem solving in science.',\n",
       " \" All right, so let's take a bit of a dive\",\n",
       " \" into this cartoon here, what's going on here\",\n",
       " ' in this amino acid transfer.',\n",
       " ' I mentioned that ribosomes are essentially',\n",
       " ' a cellular biomolecular complex',\n",
       " ' that contain both proteins and RNA.',\n",
       " ' So you may wonder, obviously they need to do this catalysis',\n",
       " ' of amino acids polymerizing to form',\n",
       " ' a elongated protein chain, right?',\n",
       " ' So you may wonder what does this catalysis, right?',\n",
       " ' So it turns out that, let me start from here.',\n",
       " \" It turns out that what I'm showing you here is, I guess,\",\n",
       " \" let's imagine you are the smallest subunit of ribosome.\",\n",
       " ' I mentioned all ribosomes, both in E. coli ribosomes',\n",
       " ' and human ribosomes, we have two subunits,',\n",
       " ' one big and small.',\n",
       " \" Let's imagine you are a small subunit\",\n",
       " \" and you're looking out for the big subunit\",\n",
       " ' when they form the complex, this is what you see, right?',\n",
       " ' So this is a big complex from the perspective',\n",
       " ' of the small subunit looking at that, right?',\n",
       " ' So these purple units are proteins,',\n",
       " ' and individual proteins that make up this large subunit.',\n",
       " ' And you can see other colors,',\n",
       " ' they all essentially components that are essentially RNA',\n",
       " ' based on ribonutrient acid.',\n",
       " \" R-RNAs, there's little r, stands for ribosomal RNA.\",\n",
       " ' So you can see, for instance, this burgundy and white color',\n",
       " ' are ribosomal RNA, green is also,',\n",
       " ' darker green is also 23S,',\n",
       " ' these are just the terminologies of ribosomal RNA',\n",
       " ' in orange and white.',\n",
       " ' But you can see the tRNA that needs to bring in',\n",
       " ' your desired amino acids, essentially in red.',\n",
       " ' And obviously there will be two tRNAs',\n",
       " ' at the point that the chain elongation is happening.',\n",
       " ' If you allow me to go back, you can see two tRNAs',\n",
       " ' on the, so ribosome actually has three sites',\n",
       " ' in the large subunit if you follow here.',\n",
       " ' And there will be two tRNAs,',\n",
       " ' at least two on each of these sites.',\n",
       " ' During the process, you are doing this amino acid',\n",
       " ' conjugation polymerization,',\n",
       " ' because one has to be the tRNA that is carrying',\n",
       " ' the elongating chain of your polypeptide or protein.',\n",
       " ' And the other has to be the additional amino acid',\n",
       " ' that will come in and extend the chain.',\n",
       " ' I hope that makes sense.',\n",
       " \" And that is what you're seeing, the two tRNAs.\",\n",
       " ' But at the molecular level, if you zoom in further',\n",
       " ' at the amino acids that are making this peptide bond,',\n",
       " ' just remember proteins are linked to peptide bonds',\n",
       " ' or AMI bonds.',\n",
       " ' So this is essentially the peptides that you have formed.',\n",
       " ' So there are two tRNAs coded by red and green color',\n",
       " ' in two different sites known as B site and A site.',\n",
       " ' B site is essentially polymerizing site,',\n",
       " ' that is tRNA, the center site that I showed you',\n",
       " ' in the previous cartoon,',\n",
       " ' this tRNA carrying the already elongating polypeptide.',\n",
       " \" And essentially it's connected through this ester bond,\",\n",
       " ' the oxygen with a carbonate here.',\n",
       " ' At the end of the tRNA,',\n",
       " ' remember tRNAs are nucleic acid, right?',\n",
       " ' So a CCA is always the last three base pairs',\n",
       " ' that will exist on the three prime end of any tRNA',\n",
       " ' that will be carrying the elongating polypeptide.',\n",
       " ' And the other tRNA in the site next door known as A site,',\n",
       " ' essentially A site is essentially starting site,',\n",
       " ' essentially is carrying a single amino acid',\n",
       " ' again at the three prime end of the tRNA sugar, right?',\n",
       " ' So look at the peptide bond formation based on this,',\n",
       " ' you start with ester bond,',\n",
       " ' you have the amino group,',\n",
       " ' this is a free amino group from your amino acid',\n",
       " ' that is conjugated to the tRNA.',\n",
       " ' You ultimately lead to the peptide bond here.',\n",
       " ' What do you think the thermodynamic drive here is?',\n",
       " ' Ester versus AMI bond, proteins are made of AMI bonds.',\n",
       " \" Why do you think that's the case?\",\n",
       " ' Protein needs to be stable, right?',\n",
       " ' So why do you think proteins, amino acid linkages',\n",
       " ' are linked through AMI bonds, so-called peptide bonds,',\n",
       " \" instead of let's say a bonds\",\n",
       " ' like this kind of oxygen-based ester bond?',\n",
       " \" It might say, I don't know,\",\n",
       " ' if you guys in high school learn about resonance structure,',\n",
       " ' balancing weight, no?',\n",
       " ' Yes, essentially these bonds are more stable,',\n",
       " \" don't worry about it if you are not familiar with.\",\n",
       " ' The peptide bonds are essentially one of most stable bonds',\n",
       " \" you'll find in the chemical world\",\n",
       " ' and also in nature as well.',\n",
       " ' So essentially you gain energy,',\n",
       " ' essentially thermodynamically is downhill.',\n",
       " \" I should say, yeah, it's more stable,\",\n",
       " \" so it's lower energy, not gain energy,\",\n",
       " ' and you break the weaker bond.',\n",
       " \" So it's a thermodynamically downhill process,\",\n",
       " ' but still to make this synthesis happen',\n",
       " ' at 10 amino acid,',\n",
       " ' some of the combination per second,',\n",
       " ' the process has to be highly efficient.',\n",
       " ' Why?',\n",
       " ' So there is a catalysis there',\n",
       " ' that lowers the process,',\n",
       " ' conditions take energy to speed up the reaction',\n",
       " ' and that is the job of the ribosome.',\n",
       " ' So I mentioned in the title,',\n",
       " ' the ribosome is actually a ribozyme.',\n",
       " ' Zymes sounds like enzyme, right?',\n",
       " ' So it was discovered now several decades ago',\n",
       " ' that the way the ribosome catalyzed',\n",
       " ' this peptide bond formation process',\n",
       " ' to make the proteins is not actually by protein catalysis,',\n",
       " ' but it is RNA catalysis.',\n",
       " ' Can you guys point to me',\n",
       " ' based on what I have discussed so far',\n",
       " ' of the ribosome architecture,',\n",
       " ' where is the RNA catalysis coming in?',\n",
       " ' Or where maybe the origin of RNA catalysis',\n",
       " ' instead of protein catalysis?',\n",
       " ' Most enzymes are proteins, right?',\n",
       " ' You guys know.',\n",
       " ' So ribosome being an enzyme',\n",
       " ' where catalysis is coming from RNA',\n",
       " ' instead of protein based mechanism,',\n",
       " ' it was a major discovery,',\n",
       " ' one of the most parts,',\n",
       " ' one of the areas.',\n",
       " ' So what do you think',\n",
       " ' what I have just discussed with ribosome architecture?',\n",
       " ' Maybe this gives you a hint as well.',\n",
       " ' I think I have an arrow here.',\n",
       " ' So what do you think?',\n",
       " ' What is going on here?',\n",
       " ' So what do you think this A is?',\n",
       " ' Where is it coming from?',\n",
       " ' So this, whatever lone pair is deprotonating',\n",
       " ' to extract the proton',\n",
       " ' to essentially activate this nitrogen',\n",
       " ' to be more active to actually go and attack this ester bond.',\n",
       " ' So where do you think this A2451 is coming from?',\n",
       " ' Essentially from one of the nitrogen on the adenine',\n",
       " ' based off the one of the ribosomal RNA',\n",
       " ' that is associated with this ribosome.',\n",
       " \" That's why ribosome is actually the ribosome.\",\n",
       " ' And after that, not just ribosome,',\n",
       " ' the additional RNAs that people have found in nature,',\n",
       " ' especially in bacterial world',\n",
       " ' that actually serve as an enzyme',\n",
       " ' by RNA based enzyme',\n",
       " ' when this was discovered in a more broadly manner.',\n",
       " ' It was a big sort of hoo-ha back in those days',\n",
       " ' and inspired by these kinds of discoveries',\n",
       " ' synthetic biologists later on engineer',\n",
       " ' different RNA scaffold and structures',\n",
       " ' that will have enzymatic properties.',\n",
       " \" And we'll see that later\",\n",
       " \" but this is a nature's example of RNA catalysis.\",\n",
       " ' Right, ribosome we have it in the cell',\n",
       " ' and is doing it at the rate of 10 amino acid per second.',\n",
       " ' Okay.',\n",
       " ' All right.',\n",
       " ' So at the level of mRNA, right,',\n",
       " ' we have various components.',\n",
       " ' I mentioned about coding sequence and others,',\n",
       " ' but, and also there are five prime UTR',\n",
       " ' and three prime UTR and translated region.',\n",
       " \" I'd like to highlight the fact that both five prime\",\n",
       " ' and three prime UTR are essential components',\n",
       " ' that really required for regulating the translation.',\n",
       " ' So where do you think that',\n",
       " ' if I have essentially structured like this',\n",
       " ' the coding sequences here,',\n",
       " \" but there's also, you see the UORF,\",\n",
       " ' ORF stands for open reading frame,',\n",
       " ' but obviously coding sequence itself',\n",
       " ' is already one open reading frame, right?',\n",
       " ' Open reading frame is the sequence',\n",
       " ' that will allow you to code',\n",
       " ' for your protein sequence ultimately, right?',\n",
       " ' It is a sequence that is bracketed',\n",
       " ' between star codon and stop codon.',\n",
       " ' That is a definition of ORF.',\n",
       " ' People call us just ORF, right?',\n",
       " ' Not quite a lot of the rings,',\n",
       " ' but this is essentially ORF.',\n",
       " ' Do you see another ORF in this illustration?',\n",
       " ' Yes, we do, right?',\n",
       " ' There is another upstream ORF.',\n",
       " ' Given this, how many translation star size',\n",
       " ' that you could potentially have?',\n",
       " ' Yeah, it will be two very good.',\n",
       " ' And the star codon is always AUG, right?',\n",
       " ' That is how recognition happens.',\n",
       " ' So there should be AUG.',\n",
       " ' If I have a structure like that,',\n",
       " ' I should immediately know that',\n",
       " ' if I have upstream URORF',\n",
       " ' in addition to the canonical coding sequence ORF,',\n",
       " ' then you will have AUG here',\n",
       " ' and there has to be AUG here, right?',\n",
       " ' Without AUG, you just cannot start the translation',\n",
       " ' because AUG is a part of the ribosome docking site.',\n",
       " ' Ribosome needs to formally and physically interact',\n",
       " ' with the mRNA to initiate the translation.',\n",
       " ' And that sequence is known as COSAC sequence.',\n",
       " ' So if you have an mRNA that is missing COSAC sequence,',\n",
       " ' forget about protein translation',\n",
       " ' because ribosome will not be able to dock',\n",
       " ' on your eukaryotic mRNA',\n",
       " ' to initiate the protein translation, okay?',\n",
       " ' So AUG, which is a star codon,',\n",
       " ' I stress again,',\n",
       " ' it is actually a part of the COSAC sequence.',\n",
       " \" What do you think there's,\",\n",
       " ' you see another thing, zip code, right?',\n",
       " ' Do you think zip code is part of the coding sequence,',\n",
       " ' part of the amino acid,',\n",
       " ' ultimate amino acid sequence coding',\n",
       " ' based on the structure,',\n",
       " ' based on the architecture?',\n",
       " ' Yes, no?',\n",
       " ' No, very good.',\n",
       " ' Why do you think that people call the sequence',\n",
       " ' that is somewhere within the three prime',\n",
       " ' untranslated region, UTR for untranslated region,',\n",
       " \" i.e. you don't make proteins from this sequence, right?\",\n",
       " ' Why do you think people designate this as a zip code?',\n",
       " ' Yeah, absolutely.',\n",
       " ' Yeah, exactly.',\n",
       " ' Absolutely a great, fantastic answer.',\n",
       " ' So essentially the name tells the same, right?',\n",
       " ' I mean, essentially three prime UTR plays a big role',\n",
       " ' to control where should this mRNA be sent to, right?',\n",
       " ' So many times in the protein coding mRNAs,',\n",
       " ' you will see in the UTR region,',\n",
       " \" there's a zip code,\",\n",
       " ' a sequence that will signal',\n",
       " ' for a particular mRNA binding protein,',\n",
       " \" hey, I need to go to, let's say ER,\",\n",
       " \" I need to go to, I don't know,\",\n",
       " ' my lysosome or whatever, right?',\n",
       " ' In the cellular framework, take me there.',\n",
       " \" So that's what the zip code sequence is for, right?\",\n",
       " ' You can also see other features',\n",
       " ' we already talk about stabilization factors,',\n",
       " ' such as poly-A tail and then the five prime capping.',\n",
       " ' But I think the take home here is that the five prime UTR',\n",
       " ' that essentially in front of the coding sequence,',\n",
       " ' this is a stop codon, stop codon,',\n",
       " ' and then three prime UTR',\n",
       " ' at the downstream of the stop sequence,',\n",
       " ' they are very essential,',\n",
       " \" even though they don't lead to the protein translation,\",\n",
       " ' because these sequences are not for coding,',\n",
       " \" but they're still essential\",\n",
       " ' because you will have several signals',\n",
       " ' that are important for stability,',\n",
       " ' localization in the cell,',\n",
       " ' and also modulating the translation.',\n",
       " ' So five prime UTR does more for the translation modulation',\n",
       " ' in a large part because there are additional loop',\n",
       " ' and stem structure or stem loop structure of RNAs',\n",
       " ' that will interfere and regulate',\n",
       " ' whether ribosome can access to this architecture or not.',\n",
       " ' Imagine ribosome is pretty big,',\n",
       " ' it needs to go and find this COSAC sequence.',\n",
       " ' If these structures, 3D structures are in the way,',\n",
       " ' it will prevent your ribosome to do the translation.',\n",
       " \" Okay, so that's certainly a takeaway message from here.\",\n",
       " ' Okay, so far so good for us.',\n",
       " ' Okay, very good.',\n",
       " \" So let's go through this table exercise, checking my time.\",\n",
       " ' So as I said,',\n",
       " ' synth biologists are inspired by nature, right?',\n",
       " \" So it's a little small, I do apologize,\",\n",
       " ' but to engage with you guys better,',\n",
       " \" it's better that all of these examples\",\n",
       " ' are staying in a single slide.',\n",
       " \" That's why it's individual diagrams, a little small.\",\n",
       " ' So maybe refer to your slides or print out.',\n",
       " \" What we're going to go through is\",\n",
       " ' several natural systems',\n",
       " ' of so-called endogenous regulated allergies.',\n",
       " ' And then in the subsequent discussion,',\n",
       " \" we'll make use of these nature idea and nature's mechanism\",\n",
       " ' to control cellular response output and decision-making.',\n",
       " \" So let's go through as an exercise,\",\n",
       " \" while we're also learning the concept.\",\n",
       " ' So there is a classic terminology',\n",
       " ' that synthetic biologists and bioengineers use,',\n",
       " ' transacting versus cis-acting.',\n",
       " ' Have you guys been discussed,',\n",
       " ' or have you guys been told about transacting and cis-acting?',\n",
       " \" Okay, so it's a simple thing.\",\n",
       " ' Trans-means opposite cis-means on the same side, right?',\n",
       " ' So if something is cis-acting,',\n",
       " ' the ingenious system, either natural or ingenious design,',\n",
       " ' it is acting on a different strand of the,',\n",
       " ' you know, the target ion, or you know,',\n",
       " ' nucleic acid of sequence, that would be trans-acting.',\n",
       " ' If it is on the same strand,',\n",
       " ' without any, most of the nucleic acid would be double-trended.',\n",
       " ' If it is acting on the same strand,',\n",
       " ' it would be cis-acting.',\n",
       " \" So we're going to have to choose,\",\n",
       " ' based on the illustration,',\n",
       " ' whether the natural system is cis-acting or trans-acting.',\n",
       " ' Okay, and of course, many times,',\n",
       " ' depending on what the context is,',\n",
       " ' typically the figure will tell you a lot.',\n",
       " ' The binding effect of your, you know,',\n",
       " ' of the small RNAs, or regulated RNAs,',\n",
       " ' to your target RNA,',\n",
       " ' could lead to either translation of regulation,',\n",
       " ' or translation inhibition, right?',\n",
       " ' Or either suppression or activation.',\n",
       " ' You also have to be able to consider those forms.',\n",
       " ' Because only with these concepts together,',\n",
       " ' you can actually get the actual output, right?',\n",
       " \" Because you don't want to end up designing something,\",\n",
       " ' you want it to inhibit the translation,',\n",
       " ' and ultimately what you design ended up activating',\n",
       " ' or promoting the translation, that would be bad.',\n",
       " \" So there are two points we'll discuss\",\n",
       " ' for each of the examples here.',\n",
       " ' First is trans-acting versus cis-acting.',\n",
       " ' Second is whether it leads to activation of translation,',\n",
       " ' or suppression of the translation, okay?',\n",
       " \" All right, so let's first start with the figure eight here.\",\n",
       " ' So comment on differences between prokaryotes',\n",
       " ' and eukaryotes.',\n",
       " \" Well, there'll be some side questions\",\n",
       " ' I want you guys to engage,',\n",
       " ' but beyond the two generic concepts I mentioned.',\n",
       " ' So the figure says the top one is prokaryotic,',\n",
       " ' and the second one is eukaryotic.',\n",
       " \" So the piece, the long piece that you're seeing here\",\n",
       " ' obviously mRNA, just by looking at the figure,',\n",
       " ' many of you know that one is prokaryotic mRNA,',\n",
       " ' one is eukaryotic mRNA.',\n",
       " ' Can you guys tell me what each is happening',\n",
       " ' in this visualization?',\n",
       " \" Yeah, let's see.\",\n",
       " ' Absolutely, so we learned last week as well as this week',\n",
       " ' that if eukaryotic mRNA,',\n",
       " ' before they can go to ribodome and translate,',\n",
       " ' they need to be processed and mature.',\n",
       " ' And this maturation process requires',\n",
       " ' this five prime capping, this sort of a square box,',\n",
       " ' and then poly a tail tagging.',\n",
       " ' So maybe on the footnote,',\n",
       " ' I mentioned five prime cap is a purple square.',\n",
       " \" And then a is, if you see multiple a's,\",\n",
       " \" that's essentially poly a tail.\",\n",
       " ' So that already tells you,',\n",
       " ' even if these labels are missing,',\n",
       " ' you know that the mRNA,',\n",
       " \" the one that I'm pointing,\",\n",
       " ' must be d-caryotic mRNA.',\n",
       " ' Okay.',\n",
       " ' All right.',\n",
       " ' And then,',\n",
       " ' all right, what do you think?',\n",
       " ' I exposed the answer.',\n",
       " ' So my regulatory mRNA,',\n",
       " ' it is the red piece, what?',\n",
       " ' The small regulatory mRNA.',\n",
       " ' And that is these regulatory mRNAs',\n",
       " ' are present in nature,',\n",
       " ' bacteria as well as in the mammalian system.',\n",
       " \" So here it's essentially,\",\n",
       " ' it operates in both prokaryotic and proprietary system.',\n",
       " ' They are called antisense RNA.',\n",
       " ' Antisense regulatory RNA,',\n",
       " ' is similar to antisense on the nuclear drugs',\n",
       " ' that you guys learned last week, right?',\n",
       " ' So they evolved the similar size',\n",
       " ' to the drugable RNAs, right?',\n",
       " ' So what do you say,',\n",
       " \" looking at the way that it's drawn,\",\n",
       " ' is that transacting or cisacting?',\n",
       " ' Maybe you saw that, yeah.',\n",
       " ' Transacting, very good,',\n",
       " ' because my regulatory RNA is the red,',\n",
       " \" and then the target that I'm binding and affecting\",\n",
       " ' is opposite strand, right?',\n",
       " ' So essentially it will be transacting.',\n",
       " ' And obviously you can perhaps predict',\n",
       " ' that this binding should repress translation.',\n",
       " ' Can you guys explain why that is?',\n",
       " ' Why is it that in both system binding,',\n",
       " \" as I've drawn here,\",\n",
       " ' to be expected to suppress translation',\n",
       " ' rather than promoting?',\n",
       " ' Yeah.',\n",
       " \" Wouldn't it just stop the red?\",\n",
       " ' Very good answer.',\n",
       " ' Yeah, exactly.',\n",
       " \" So you don't need to slice through,\",\n",
       " ' through a stretch of the mRNA sequence,',\n",
       " ' and it processes,',\n",
       " ' it has a processivity in catalysis',\n",
       " ' and polymerization mechanisms,',\n",
       " ' simply slice it through,',\n",
       " ' a bit like Combiabelt,',\n",
       " ' into like many ribosome molecules line up on the mRNA,',\n",
       " ' and they slice through,',\n",
       " ' and that will be blocked.',\n",
       " ' So remember the green part is actually coding region, right?',\n",
       " ' So the coding region is blocked',\n",
       " ' because the regulatory RNA is bound on that green unit,',\n",
       " ' which is a ORF or coding sequence.',\n",
       " ' Okay, good.',\n",
       " \" So let's continue.\",\n",
       " ' All right, so the second example',\n",
       " ' is still called small sRNA in prokaryotes.',\n",
       " ' So rival regulators are exclusively found in prokaryotes,',\n",
       " ' not in human,',\n",
       " ' but not in eukaryotic system.',\n",
       " ' They actually have a system where,',\n",
       " ' again, all these RNAs are relatively small units, right?',\n",
       " ' So you can see this additional piece here,',\n",
       " ' and also additional piece here.',\n",
       " ' They have this base pairing capability',\n",
       " ' that will act on the target mRNA.',\n",
       " ' One is acting outside of the coding region, right?',\n",
       " ' The other is adding',\n",
       " ' in the inner side of the coding region in green, right?',\n",
       " ' Are this acting or transacting for rival regulators?',\n",
       " ' Very good, excellent answer.',\n",
       " \" Transacting because still I'm acting on an opposite\",\n",
       " ' or different strand, right?',\n",
       " ' Of what my regulatory RNA is.',\n",
       " ' And the statement here says that binding of the regulators',\n",
       " ' to their target may repress or activate translation.',\n",
       " ' Why is it that my cartoon here',\n",
       " ' is consistent with this statement?',\n",
       " ' Can someone tell me why?',\n",
       " ' So unlike, so the top mechanism,',\n",
       " ' antisense mechanism always suppress translation, right?',\n",
       " ' But when you get to rival regulators,',\n",
       " ' depending on their structure and base pairing signs,',\n",
       " ' some of them may suppress',\n",
       " ' and some of them may not suppress or activate, right?',\n",
       " ' So what do you think?',\n",
       " \" So let's look at this, the one that I'm pointing here.\",\n",
       " ' So would this one be a suppressing translation or?',\n",
       " ' Yeah.',\n",
       " ' So that would be suppression for sure.',\n",
       " ' It will interfere with ribosome slightly, very good.',\n",
       " ' What about the top?',\n",
       " ' Confirmation?',\n",
       " ' Yeah.',\n",
       " ' Can access, yeah.',\n",
       " ' It looks like a very good answer.',\n",
       " ' It looks like maybe that, you know,',\n",
       " ' when this is in the way ribosome cannot perhaps, you know,',\n",
       " ' or not just ribosome, right?',\n",
       " ' We need to have other factors to come in',\n",
       " ' and initiate the translation that may not be as efficient',\n",
       " ' when this structure of the RNA is folded in this system,',\n",
       " ' right?',\n",
       " ' And it is getting in the way, perhaps sterically blocking',\n",
       " ' and that may lead to, you know,',\n",
       " ' if you essentially have this activation',\n",
       " ' or maybe enhancing the ribosome efficiency of translation.',\n",
       " ' Okay.',\n",
       " ' Very good.',\n",
       " \" So after we have gone through this while we're here,\",\n",
       " \" we can take a break, but let's keep going.\",\n",
       " ' Now we see the ribozyme.',\n",
       " ' Ribozyme are special name for catalytic RNA, right?',\n",
       " ' What you have seen so far,',\n",
       " \" antisense RNA regulators, they're not doing chemistry, right?\",\n",
       " \" They're not catalysis, they're just regulating.\",\n",
       " \" They're modulators, they're not catalysts.\",\n",
       " ' Whereas when you get to ribozyme, they are enzyme.',\n",
       " ' RNA based enzyme, they are catalysts, right?',\n",
       " ' Biological catalysts or nucleic acid-based catalysts.',\n",
       " ' So you can see in the coding region',\n",
       " ' and a lot of structure in the five prime UTR, right?',\n",
       " ' So ribozymes typically adopt this 3D shape',\n",
       " ' to give them a fold that would have this catalytic activity.',\n",
       " ' And many of the ribozymes can do self-catalysis,',\n",
       " ' i.e. self-cleavage.',\n",
       " \" What you're seeing here is the mechanism where,\",\n",
       " ' or illustration where cleavage actually happens',\n",
       " ' in this particular point.',\n",
       " ' What would you say in this particular illustration?',\n",
       " ' Is it acting in cis or trans?',\n",
       " ' It will be cis again, right?',\n",
       " ' Because the strands, sorry,',\n",
       " ' that breaks essentially the coding strand, right?',\n",
       " ' The strand that is carrying the green unit.',\n",
       " ' Again, apologize a bit small,',\n",
       " ' but hopefully you can view better in your slides.',\n",
       " ' So that will be the acting in.',\n",
       " ' All right.',\n",
       " ' So how about the activity in terms of the translation?',\n",
       " ' Maybe you can imagine that if you are actively cleaved',\n",
       " ' in the cis acting manner,',\n",
       " ' it will actually repress the translation.',\n",
       " ' Why?',\n",
       " \" Because you're going to lose some parts\",\n",
       " ' of the five prime UTR.',\n",
       " ' As I mentioned, it is important for recruiting ribosome',\n",
       " ' and initiating a translation that will be lost',\n",
       " ' and that will lead to repression of translation.',\n",
       " ' Okay.',\n",
       " ' But if you do a cleavage on the trans acting manner,',\n",
       " ' i.e opposite strand, we cannot predict a priori.',\n",
       " ' What will be the effect on the translation, right?',\n",
       " ' It depends on the structure, additional context, et cetera.',\n",
       " ' Is that clear for you guys?',\n",
       " ' In the case of ribosome,',\n",
       " \" where whether it's going to be repressive or stimulatory.\",\n",
       " ' Yeah, you have a question?',\n",
       " ' Okay.',\n",
       " ' Ribosome is a common name for RNAs',\n",
       " ' that have enzymatic properties.',\n",
       " ' So catalytic RNAs,',\n",
       " ' if you see any RNAs that have catalytic property,',\n",
       " ' you can call them as ribosome.',\n",
       " ' Ribosome is an example in nature that is a catalytic RNA',\n",
       " ' because it has the RNA components',\n",
       " ' that actually catalyzes the amino acid polymerization.',\n",
       " ' But ribosome itself is a machinery',\n",
       " ' that allows you to translate proteins,',\n",
       " ' i.e doing this polymerization.',\n",
       " ' I hope that makes sense.',\n",
       " ' Okay, excellent and good question.',\n",
       " ' All right.',\n",
       " ' So you guys get this cis-trans',\n",
       " \" because these kinds of things that you'll see there\",\n",
       " \" and problem set, you'll see the design and application\",\n",
       " ' in the final exam as well.',\n",
       " ' Okay.',\n",
       " \" So let's go through almost to the end before the break.\",\n",
       " ' So in D, you see a class called riboswitches,',\n",
       " ' which already means you need to be able to turn on and off.',\n",
       " ' And typically what kind of system allows you to turn on',\n",
       " ' and off very rapid manner?',\n",
       " ' Are you talking about genetic system',\n",
       " ' or small molecule pharmacological system?',\n",
       " ' Rapid on and off typically is a small molecule system',\n",
       " ' because link binding, de-binding can be very rapid,',\n",
       " ' both in vivo and in the cell culture or in vitro system.',\n",
       " ' So indeed riboswitches are structured RNAs',\n",
       " ' that can regulate protein translation in the level',\n",
       " ' that they can be either activated or deactivated',\n",
       " ' by small molecule binding.',\n",
       " ' So if you look at this particular structure',\n",
       " \" I'm pointing here, they can also be ingenious\",\n",
       " ' such that their regulation could be transcription',\n",
       " ' or translation.',\n",
       " ' But if you look at perhaps these example,',\n",
       " ' you can see that when the ligand is bound,',\n",
       " ' they have this structure on the RNA,',\n",
       " ' five prime UTR if you like, that is different',\n",
       " \" from the system when the ligand binding doesn't exist.\",\n",
       " ' So when the ligand dissociates due to drug concentration',\n",
       " ' drops in vivo or whatnot, et cetera,',\n",
       " ' you will have a different structure of the resulting RNA',\n",
       " ' that will allow the transcription',\n",
       " ' or translation to take place.',\n",
       " ' So that allows you to modulate at the on off level',\n",
       " ' for ribosome accessibility on the coding strand.',\n",
       " ' So given this, are they cis-acting, trans-acting?',\n",
       " ' Typically, yeah.',\n",
       " ' This is very good.',\n",
       " ' So you can see from the figure as well,',\n",
       " ' the coding region is on the same strand',\n",
       " ' as where the structural changes are happening',\n",
       " ' by ligand binding, de-binding.',\n",
       " ' Okay, the last two, the first E, small interfering siRNA.',\n",
       " ' What do you think, cis-acting, trans-acting?',\n",
       " ' Trans-acting, excellent.',\n",
       " ' And suppression and deactivation, again,',\n",
       " ' which one would you expect?',\n",
       " ' This is happening in the coding region again,',\n",
       " ' most likely suppression, right, interfere with translation.',\n",
       " ' What about the last category, micro mRNA',\n",
       " ' that are very abundant in human cells as well,',\n",
       " ' so-called miRNA.',\n",
       " ' Trans-acting, excellent.',\n",
       " ' And binding represses translation.',\n",
       " ' Majorly, it represses translation,',\n",
       " ' but there are also systems these days being discovered',\n",
       " ' that may go the other way as well.',\n",
       " ' All right, so do you guys have any questions?',\n",
       " ' So these individual questions are essentially answering',\n",
       " ' what we have just discussed.',\n",
       " \" Maybe the last question I'll ask from my side is that\",\n",
       " \" out of all of these various forms of nature's\",\n",
       " ' regulatory RNA, so these are endogenous in organisms,',\n",
       " ' including humans, which specific type or types',\n",
       " ' is all obligatory and they get dated properly.',\n",
       " ' Yeah.',\n",
       " ' Ribosucious one, very good.',\n",
       " \" That's quite obvious, very good,\",\n",
       " ' still very good, well done.',\n",
       " ' What else?',\n",
       " ' What other example that you see',\n",
       " ' that could be legit dated as well?',\n",
       " \" It's not so obvious.\",\n",
       " ' So one obvious example is ribosucious.',\n",
       " ' On an off switch, I mean, you allow biasing data.',\n",
       " ' Although ABC and BNF, so very many of my software',\n",
       " ' that may be convertible on top by many enzymes.',\n",
       " ' It turns out that ribozymes,',\n",
       " ' they are essentially catalytic RNAs,',\n",
       " ' just like in enzyme, they have binding pockets and structures',\n",
       " ' are usually structures and shape.',\n",
       " ' Most of the time you can engineer small molecules',\n",
       " ' that will bind to ribozymes and allow the catalysis',\n",
       " ' to be either suppressed or activated.',\n",
       " ' So indeed ribozymes are also another category',\n",
       " ' of RNA-based regulators or RNA regulators',\n",
       " ' that allow ligand-gated control',\n",
       " ' that gives you an additional level of on-off control',\n",
       " ' that is otherwise not possible.',\n",
       " \" I hope you're enjoying so far.\",\n",
       " \" So we'll take a break as usual for 10 minutes\",\n",
       " ' and then come back and go with case studies in the next.',\n",
       " ' Thanks.',\n",
       " ' So, you know, friends versus cis-actin.',\n",
       " ' Actually, one of the students asked me a question',\n",
       " \" related to what we've learned last week.\",\n",
       " \" It's good questions about these particular microRNAs.\",\n",
       " ' Last, sorry, therapeutic strategies of ASOs,',\n",
       " ' if you recall last lecture and problems,',\n",
       " ' they can also target, and essentially the outcome',\n",
       " ' is typically downregulation of the protein level.',\n",
       " ' But remember, this interference,',\n",
       " ' here we said the binding repressed translation,',\n",
       " ' even if the binding is not on the coding region,',\n",
       " ' you could still interfere with a poly-A tail sequence',\n",
       " ' that I mentioned that is a recognition sequence',\n",
       " ' required for mRNA binding proteins to protect a given mRNA',\n",
       " ' and export them outside the nucleus to the cytoplasm',\n",
       " ' where they will meet ribosome, right?',\n",
       " ' So if this protection fails',\n",
       " ' because the regulatory mRNA in red gets in the way,',\n",
       " ' your mRNA will be degraded',\n",
       " ' before it reaches the cytoplasm point to do the translation.',\n",
       " ' So this kind of interference mechanism',\n",
       " ' could really be at the upstream',\n",
       " ' rather than active ribosome translation event, right?',\n",
       " ' Okay, so I just wanted to touch base',\n",
       " ' and make sure I clarified this.',\n",
       " \" Okay, so we're going to do a lot of interesting case examples.\",\n",
       " ' Now, coming to the point where we get this understanding',\n",
       " ' from nature, how nature systems are operating',\n",
       " ' at the RNA regulatory level or RNA regulators,',\n",
       " ' and how we can leverage these insights',\n",
       " ' to better design and apply sort of engineer RNAs',\n",
       " ' that will allow us to control the protein translation',\n",
       " ' or gene regulation overall.',\n",
       " \" And there are two classes we'll focus on.\",\n",
       " ' One is engineer RNAs,',\n",
       " ' and second or added a layer of complexity',\n",
       " ' of additional layer of control is ligand-gated system, right?',\n",
       " ' They will allow you to on-off modulate',\n",
       " ' in an on-off manner, more dynamic way, right?',\n",
       " ' So which type of natural regulatory RNAs',\n",
       " ' should this other class mimic?',\n",
       " ' So which, yeah, this one.',\n",
       " ' Yeah, the switch is going to be nice.',\n",
       " ' Very good answer.',\n",
       " ' So we know that ribosuches from the last slide',\n",
       " \" is ligand-gated, actually, I'm sorry.\",\n",
       " ' And then indeed, sorry about this, yeah.',\n",
       " ' Indeed, ribosuches will be the classic category',\n",
       " ' that will allow ligand binding,',\n",
       " ' D binding to regulate the RNA structure.',\n",
       " ' And then of course, their interaction with their target RNA',\n",
       " ' and ribozymes, i.e. catalytic RNAs,',\n",
       " ' will also be a type of a class of RNA',\n",
       " ' that have 3D structure that will create ligand binding site.',\n",
       " ' So there will be two categories',\n",
       " ' under this particular class, right?',\n",
       " ' So in the first class, just with engineer RNAs,',\n",
       " ' no small molecule bits, there are three types,',\n",
       " \" but we won't cover the last type.\",\n",
       " \" We'll cover the first type as two examples.\",\n",
       " ' So engineer RNA transcriptional activators, right?',\n",
       " ' So let me throw in a question.',\n",
       " ' Why do you think that this is engineer?',\n",
       " \" I know we say biologists, we're always engineering,\",\n",
       " ' but why do you think when we talk about',\n",
       " ' RNA transcriptional activators',\n",
       " ' or transcriptional factors that are RNA-based,',\n",
       " ' they need to be synthetic, right?',\n",
       " ' So what are transcriptional factors in general?',\n",
       " ' We talked about it quite a bit last week.',\n",
       " ' Transcriptional factors, yeah?',\n",
       " ' Yeah, they are proteins that bind to DNA',\n",
       " ' and regulate gene transcription, right?',\n",
       " ' And we want to achieve the same output, same mechanism,',\n",
       " ' but rather than engineering protein,',\n",
       " \" we're going to engineer RNA\",\n",
       " ' to be able to achieve the same final output, right?',\n",
       " ' So saying that this is RNA-based,',\n",
       " ' transcriptional modulator or transcriptional factors,',\n",
       " ' they have to be engineered because in nature,',\n",
       " ' all transcriptional factors are proteins.',\n",
       " ' And I think that you already answered.',\n",
       " ' So this part is not as part of the course,',\n",
       " \" but there's advanced student who wants to learn.\",\n",
       " \" I put the slide up, but we won't discuss this.\",\n",
       " ' Let me kill these things, yeah, the floating thing.',\n",
       " ' Okay, all right.',\n",
       " \" So let's start with RNA-based transcriptional modulators.\",\n",
       " ' Okay, fine.',\n",
       " ' The goal is to design RNA-based transcription factors, right?',\n",
       " ' To make you guys refresh from last week,',\n",
       " ' we learned about transcription factors in general.',\n",
       " ' They are proteins.',\n",
       " ' They always have two units, DNA binding unit, a DBD in blue,',\n",
       " ' and then transactivating unit, the unit in red.',\n",
       " ' And together,',\n",
       " ' they make up these protein-based transcription factors.',\n",
       " ' They bind to the target DNA',\n",
       " ' and regulate the gene activation, right?',\n",
       " ' Gene transcription.',\n",
       " ' This is very important types of proteins.',\n",
       " \" We're now going to achieve through the design of the RNA\",\n",
       " ' to achieve the same level of output,',\n",
       " ' i.e. the RNAs that can mimic the property',\n",
       " ' of transcription factor proteins.',\n",
       " ' Okay, so the example I am showing you here',\n",
       " ' is a system design that is applicable in yeast,',\n",
       " ' So you can see the downstream output is kind of a platform',\n",
       " ' where if the gene transcription is successful,',\n",
       " \" you'll either express this either history gene\",\n",
       " ' or the Lexi gene.',\n",
       " ' This is just a, you know, example.',\n",
       " \" It's not fixed to that system only.\",\n",
       " ' Same Lexi, right, essentially is one of the downstream gene',\n",
       " ' that you can read out ultimately by a color,',\n",
       " ' a fluorometric or colorimetric assay.',\n",
       " ' History is again, a different type of gene',\n",
       " ' that allows you to screen for live-death output, right?',\n",
       " \" It's just allowing you to see whether\",\n",
       " ' this particular gene activation happens or not, right?',\n",
       " ' The goal is to actually achieve this gene activation.',\n",
       " \" And there's always operator side that will be bound\",\n",
       " ' by particular protein-based transcription factor.',\n",
       " ' In this case, LexA, this LexA bind to LexA operator side.',\n",
       " ' And the way that they come in a unit at the RNA level,',\n",
       " ' why so this is RNA structure,',\n",
       " ' that will allow you to sort of control',\n",
       " ' this downstream gene activation is through this design.',\n",
       " ' What people know is that in nature,',\n",
       " ' there is a protein RNA-based interaction',\n",
       " ' for a protein that is known as MS2 coat protein.',\n",
       " ' And protein, by definition, protein RNA interaction',\n",
       " \" means that if there's a protein,\",\n",
       " \" there's a specific RNA unit that will bind\",\n",
       " ' to that MS2 coat protein.',\n",
       " ' And those are so-called MS2 RNA hairpins.',\n",
       " ' And you can see at the level of engineered DNA construct,',\n",
       " ' those two units are incorporated, right?',\n",
       " ' So after transcription happened in the yeast,',\n",
       " ' you will get this RNA module, right?',\n",
       " ' And then you can see that because you install',\n",
       " ' this MS2 hairpin units in your coding DNA,',\n",
       " ' in your transcribed RNA,',\n",
       " \" you're going to carry the same units.\",\n",
       " \" And there's a five prime, three prime, the usual thing,\",\n",
       " ' and some sort of stem loop structure,',\n",
       " ' which is typical for 3D shape RNA.',\n",
       " ' And there is this variable region that you can play around',\n",
       " ' with various nucleotide sequence.',\n",
       " ' And then to screen for a specific sequence',\n",
       " ' that will make the system effective',\n",
       " ' at gene regulation downstream.',\n",
       " ' So the researchers vary these units.',\n",
       " \" And of course it's a selection round screening\",\n",
       " \" because there's phenotypic screening, life and death.\",\n",
       " ' So you can really clearly tell',\n",
       " ' if the gene activation happens,',\n",
       " \" if you're growing the yeast\",\n",
       " ' in the histidine dropout media,',\n",
       " \" if this particular yeast strain doesn't function\",\n",
       " ' without this metabolite,',\n",
       " ' you know that if the yeast cell survive,',\n",
       " ' that is because this gene activation is happening.',\n",
       " ' So this is a very high throughput phenotypic screening',\n",
       " ' because you drop out a particular metabolite',\n",
       " ' in your yeast growth media,',\n",
       " ' and you incorporate a genetic system',\n",
       " ' that upon activation of that system,',\n",
       " ' you will produce and not just lead that metabolite.',\n",
       " ' So you can rescue the dead phenotype.',\n",
       " ' So if you see a yeast cell activating from your screen,',\n",
       " ' you know that whatever variable sequence,',\n",
       " ' region that you have introduced,',\n",
       " ' must be a good RNA sequence',\n",
       " ' that will allow you to activate the histidine gene,',\n",
       " ' the histidine metabolite synthesis gene.',\n",
       " ' I hope that makes sense.',\n",
       " \" So that's the screening setup.\",\n",
       " ' And they actually show that',\n",
       " ' by introducing this particular structure of the RNA,',\n",
       " ' they can modulate the LexA,',\n",
       " ' the transcription factor-based protein',\n",
       " ' together with this RNA system',\n",
       " ' to modulate this LexA operator downstream gene activation.',\n",
       " ' So let me ask you a question.',\n",
       " ' You see this two unit, right?',\n",
       " ' And we also saw in the natural system,',\n",
       " ' DNA binding and transactivating.',\n",
       " ' So what do you think?',\n",
       " ' Your RNA engineer system is here, right?',\n",
       " ' Together with the protein bit over there.',\n",
       " ' So what is this whole unit representing?',\n",
       " ' Is that representing DBD or TAD?',\n",
       " ...]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subclip_0_df[\"text\"].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_text = \" \".join(subclip_0_df[\"text\"].to_list())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\" Thank you all.  Good afternoon.  How are you guys doing?  It looks fun.  How many fans have done themselves last week?  Too easy?  At least, maybe you think, well, applied the concepts that we learned in the past.  I hope that we all have relevant problems, for modern students like you guys.  At least, I hope you had some fun from doing it and not finding it too challenging.  The final exam questions, they will be largely in the context of problem solving like this.  And the level of challenge will be suddenly a little bit less, perhaps,  because in the final exam you are kind of limited on it.  So the highest level of challenge you will see, and at least my point in the course,  will be represented by the problem sets questions.  So if you feel that we understand them and you can tackle them, especially after the session,  and you are engaged in the class and you really should not need to worry suddenly from the  three lecture series that I'm very happy to provide here for this course.  So let's continue to today's topic too.  We're going to learn about designing and applying these RNA switches.  You probably know synthetic biologists are very much inspired by nature,  by how nature designs these circuits and controls cell decision making and cell responses,  all the way at the animal level.  And when we use these ideas, we really be able to control back cells and living culture models.  So we will focus on RNA level switches.  Last week it was therapeutics, as you guys recall.  So the key aspects that we will target will be first going through the principles of natural regulatory  RNA. That's kind of important fundamentals because we need to know how nature does these mechanisms,  and how we can hijack these principles to really control the cell decision making.  And there will be three courses that we will concern with for our course level.  This will be regulators, ribozymes, RNA-based enzymes, and riboswitches.  And lastly, connected to the second topic, it will be small molecule ligand or drug-guided RNA conformational gating  that will allow you to achieve this reversible regulatory control at the level of these RNA switches.  And first, to cover everyone's kind of basic fundamental principles on ribosome protein translation.  Pretty much last week, we talked about this central dogma, right, starting from transcription and translation,  and of course there's also this reverse transcription method that's happened in nature in certain contexts.  But what about this second step to the I mentioned here, protein translation.  You guys know what machinery in any organism that does this translation, what is it called?  Machinery, biomolecular machine that does this translation. It's actually known as ribosome.  So we'll talk a little bit about ribosome. So as you can see in the footnote, it's a protein synthesis machine in the cell.  It is a molecular, macromolecular complex made up of protein subunits together with certain RNAs that you will see is a catalytic RNA  that essentially together, this machine allows you to convert the mRNA sequence to your protein sequence.  Proteins are made up of amino acids that essentially is a long polymer of individual 20-coding amino acids.  So you can imagine this machine, it doesn't have in all cells, including bacteria and bacteria, those bacteria like us,  essentially it's a polymerization machine if you think about it, because you're going to allow amino acids, individual amino acids to be covalently linked.  In peptide bonds, so-called A-migons, you have these long chain proteins that you know that proteins are absolutely essential macromolecules that actually drive the cellular engine, just like in a car engine.  So that is a very essential aspect of protein synthesis is a ribosome.  So before I go there, just give you an idea of, you know, you may wonder, where does protein synthesis happen?  What I'm showing you here is the architecture of a eukaryotic cell.  Why do you think protein synthesis happens?  Which compartment?  In a very specific location, the majority is on the ER surface, but largely speaking, does it happen in the nucleus of the cytoplasm?  Which compartment?  Cytoplasm, right?  Whereas mRNA synthesis, where does it happen in this eukaryotic cell?  In the nucleus, right?  So there is this really compartmentalized segregation in eukaryotic system where mRNA synthesis happens in the nucleus,  and this mRNA undergoes maturation processing, ultimately get exported outside the nucleus into the cytoplasm,  where it will meet this machinery that I mentioned, ribosome, that will catalyze the polypeptide, i.e. amino acids polymerization, to give you the protein.  So these individual black dots, you may see better in your slides, these are actually ribosome, these sort of a complex of RNAs and proteins.  They are essentially type complex. They are individual ribosome complex molecules that are diffusing throughout the cytoplasm.  So protein synthesis actually happens in the cytoplasm. It is true, the student over there said, it's also concentrated on the ER, so these tubules, tubular green structures you see are endoplasmic reticulum.  So we won't worry so much in our cost level, but essentially the key point is that protein synthesis happens outside the nucleus, broadly speaking in the cytoplasm, which is very different from bacteria system like E. coli.  In E. coli, do we worry about cytoplasm versus nucleus? The answer is no, right? There's essentially translation and transcription couple process in E. coli, as we talked about last week.  So you may wonder, right, how fast are these processes? If you think about it, you know, one second is very fast, right? And actually ribosome can polymerize 10, on average, 10 amino acids in a second, which is really very rapid, right?  As I'm talking about, even before I finish the sentence, it will be several seconds, right, perhaps. And ribosome has already made 10 amino acids linked together.  And average protein is about 300 amino acids linked. So, if you compare ribosome rate that I mentioned, protein translation rate.  So what I'm showing you here, you don't need to memorize these numbers, but at least good to know at the level of, you know, back of the envelope number, what kind of rate that are we talking about, when you think about RNA polymerization versus amino acid polymerization.  RNA polymerization gives you mRNA from DNA, amino acid polymerization gives you protein from mRNA. So what types of enzyme that we saw last week, what does the mRNA synthesis?  So a machinery that produces mRNA, what we call these types of enzyme as, are polymerases and the RNA polymerases, right, machinery that makes DNA will be DNA polymerases, right, because it's all polymerizing the biomolecular polymers.  If you think about the rates in terms of actual DNA replication rate is in the bacteria world on the E. coli, for instance, versus mammalian cells such as, you know, cancer cell type of cancer cell line, HeLa cells, the replication timescale are quite different, right.  But at the fundamental level, at the nucleotide level, the actual rate of transcription is, largely speaking, identical. You can make per second, if I'm taking 10 to 100 nucleotides.  But why then? Why is it that our cell division in human cells is almost a day, whereas cell division in E. coli and many bacteria are less than an hour?  Why is it that DNA replication overall process takes much longer in our cells compared to E. coli?  Yes, certainly it's more complicated. You have to do the group checking, editing, etc. But still, the key point also is the fact that in the gene size, on average gene size, human gene is a large, much larger than bacterial gene.  Can you guys think about the fundamental reason why that is the case? We talked a bit about that last week. What is it that is intrinsically, as a stark difference in the genome level between E. coli genome and human genome?  Yeah, presence of non-coding genetic regions, so-called introns. Intons and exons, you don't have to worry about that in prokaryotic systems like E. coli or bacteria.  But we don't have introns in E. coli, whereas we have introns. So that means that for a given gene, even if ultimately the protein sequence is identical between E. coli protein and human protein, the actual starting genomes will be quite different.  Because in origin of the DNA in the human cell, we have introns and exons. Ultimately, introns are spliced out inside the nucleus, right, we talked about that last week, before you ultimately get to translated protein.  So even if the amino acid level sequence is identical, at the DNA level, there will be introns in the human cells and eukaryotic system that will make the whole trans DNA synthesis step taking much longer.  So replication process will be a lot longer. So that is certainly one key fundamental aspect of the eukaryotic cell, having a presence of introns.  Now, protein synthesis, which is the focus of today's discussion, as I said, the average translation rate is 10 amino acids given take per second.  So by that time, why is it that we don't have to care about these introns and exons anymore by the time we're talking about ribosome translation?  So the blue unit that I'm pointing out here is essentially a classical way to illustrate the ribosome because it has two subunits, a large one and a small one.  So the green piece is the polypeptides of proteins that are being translated, are being synthesized.  And then the red piece is the incoming mRNA that is coding for what amino acids should be in the protein sequence.  So given this, why do you think that at the point that we talk about translation, we don't have this difference anymore?  Yes, it is splicing. Absolutely.  So mRNA that we talk about here is already a mature messenger RNA.  Whereas the newly synthesized, the nascently synthesized, if you like, the mRNA up there at the point of the RNA polymerase doing the transcription will be the premature mRNA.  You also tackled that question in the problem set last week, where you need to go through splicing and export, you know, decoration with poly-A tail, etc.  before you can actually get to the mature mRNA that is ready for ribosome translation.  So remember, this has to be mature mRNA. At that point, it's identical.  Introns no longer matter because they have been spliced out.  OK. So the take home message is there's differences in time scale, broadly speaking, and also in terms of the translation, efficiency is the same between the two systems.  So let's think about bacterial mRNA and eukaryotic mRNA.  What I'm showing you here is a domain structure, very simplified, but it gives you a generic feature.  In terms of coding sequences, don't do the same for a given protein.  Because we don't need to worry about introns anymore by the time you get to the mature mRNA.  But what other striking differences do you see between the two systems?  Because we'll hijack quite a bit of these mRNA regulation and their individual domains and complements when we think about ribosome switch design and application,  the later part of the topic. So what do you see beyond the coding sequence?  Coding sequence is the mRNA that will code for your amino acid sequence on the protein level.  You see untranslated region. So this is upstream untranslated region.  So 5' UTR, so untranslated region and downstream UTR.  And those are present in both cases.  Those components are essential for gene regulation and stability of mRNA. We'll come to that.  And that essentially happens in both bacteria and eukaryotic world.  But you see additional units such as poly-A tail and 5' prime capping.  That is absolutely essential to introduce this poly-A tail tagging and then cap modification on the 5 prime.  Fourth, eukaryotic mRNA, because when you think about eukaryotic mRNA, they are first produced in the nucleus,  but they need to be chaperoned and exported in a safe manner to the cytoplasm where they will meet ribosome such that they can synthesize the protein.  If you have mRNA produced in the nucleus that are not stable and protected by nucleases degradation,  we talked a lot about RNA nucleases last lecture as a means of therapeutic mechanism.  Both nucleases could essentially chop up the RNA that is newly synthesized in the nucleus before it can even get to the cytoplasm.  So these accessories, 5 prime capping and poly-A tail tagging at the mRNA level is critical to stabilize and protect the eukaryotic mRNA such that they can be recognized by proteins that will bind to these units and actually chaperone them, escort them if you like, for mRNA export from the nucleus to the cytoplasm for protein translation.  Keep in mind of this kind of segregation of compartments that makes the system more complex.  That means you require more factors to stabilize these mRNAs, protect these mRNA, chaperone and export them safely to the cytoplasm.  So that's what you're seeing here.  Sub-cellular loca will already talk about it. What I've just discussed, you can also read further in your own time in these texts.  But let's go through a little bit more of what allows a ribosome to know you have a long piece of mRNA starting from the genome that's coming along.  How does ribosome know when to start translating, translating and synthesizing the amino acid, if you make this polymeric chain for a protein, when to start this process and when to end this process?  So there are specific codes that you will be implemented in your mRNA in nature that allows a ribosome to signal that this must be a star site. Star codon is always AUG.  And then there will be the end site, there will be a star codon in nature, there are three star codons and these three nucleotides will signal for where the star site and the end point should be.  Because if you think about it, these are, you know, if they are altered in any way, you'll have a protein that are either truncated, i.e. too short, shorter than you desire, or maybe start in the long side.  So these stop and start codons and signals are critical for ribosomes to understand where the translation should be started and should be ended.  So let's take a look at this in a sort of an overview level from the whole process with this illustration.  Many of you probably have come across chemical polymerization, perhaps in high school, and you know, a lot of polymerization processes in chemistry and industry, chemical processing.  Biology in many cases work the same way. In any polymerization process, what do you have? There are typically three steps in any kind of polymerization.  Sometimes they change the activities, that's how we would encounter in USA is we encounter chemical polymerization. There are always three steps. What are they?  You have to initiate, you have to elongate, process, and then you have to terminate. Biology does the same thing, nature does the same thing. There's initiation step, elongation step, and termination step.  So let's take a walkthrough of these individual steps. So what I'm showing you here is mRNA, 5' and 3', and you can see the stop codon, and stop codon is one of the stop codons here.  And ribosome essentially is made up of sort of two sub-major subunits, one large and one small. So not just having an mRNA and this ribosome machinery, obviously something has to bring in your individual amino acid.  So how does nature do that in all living systems? And it turns out that there is this additional type of RNA, still nucleic acid, called transfer RNA, tRNA, which serves as an adapter to bring in the precise amino acid shown here in the purple color, to actually allow this polymerization process to happen in a coded manner.  I said coded manner because which amino acid to be brought in will be defined by the actual nucleotide sequence that is present along your mRNA.  How do you think this position happens? So mRNA will have different nucleosides here, and I could have tRNA bringing in one amino acid at a time. How do I know how this system, how can you encode for the fact that if I hit this particular codon, I will bring in the correct amino acid?  How do you think this position happens?  So mRNA will have different nucleoside fields  and I could have tRNA bringing in one amino acid at a time.  How do I know how does the system,  how can you encode for the fact that,  if I hit this particular code off,  I will bring in the correct amino acid.  There's chemistry behind it.  What is this position mechanism?  Yeah.  Base pairing, right?  So if you have very good,  if you have the nucleotide sequence,  you essentially have these codons on the mRNA  that will allow precise anti-codons through base pairing.  So you have several tRNA molecules already charged  and loaded with several amino acids  available in the cellular pool,  but they will keep coming and finding and dividing  to search for the right anti-codon, codon base pairing.  When the base pairing is correct,  these tRNAs will come to the specific site of this ribosome  and allow this amino acid, a new amino acid,  to come in right next to the polymerizing amino acid,  if you like, and you'll have this amino acid  additional adding to extend this chain  that is being polymerized by the ribosome.  So ultimately, when ribosomes see the stop codon  on the mRNA sequence, such as UAG,  instead of uptaking the tRNA on this site,  you'll have the so-called proteins release factors  that will come in and eject this particular polypeptide  out of the ribosome site.  In that way, your polymerization event has ended  and you have produced successfully one completed protein.  Okay, I hope that makes sense.  All right, very good.  So let's take a look at a little bit further details.  So these are essentially different codons that you have.  If you want mnemonic, when I was a student,  I used to kind of, you are gone, you go away,  and you are away if you're worried about these stop codons,  but I don't expect you guys to memorize  more for problem solving in science.  All right, so let's take a bit of a dive  into this cartoon here, what's going on here  in this amino acid transfer.  I mentioned that ribosomes are essentially  a cellular biomolecular complex  that contain both proteins and RNA.  So you may wonder, obviously they need to do this catalysis  of amino acids polymerizing to form  a elongated protein chain, right?  So you may wonder what does this catalysis, right?  So it turns out that, let me start from here.  It turns out that what I'm showing you here is, I guess,  let's imagine you are the smallest subunit of ribosome.  I mentioned all ribosomes, both in E. coli ribosomes  and human ribosomes, we have two subunits,  one big and small.  Let's imagine you are a small subunit  and you're looking out for the big subunit  when they form the complex, this is what you see, right?  So this is a big complex from the perspective  of the small subunit looking at that, right?  So these purple units are proteins,  and individual proteins that make up this large subunit.  And you can see other colors,  they all essentially components that are essentially RNA  based on ribonutrient acid.  R-RNAs, there's little r, stands for ribosomal RNA.  So you can see, for instance, this burgundy and white color  are ribosomal RNA, green is also,  darker green is also 23S,  these are just the terminologies of ribosomal RNA  in orange and white.  But you can see the tRNA that needs to bring in  your desired amino acids, essentially in red.  And obviously there will be two tRNAs  at the point that the chain elongation is happening.  If you allow me to go back, you can see two tRNAs  on the, so ribosome actually has three sites  in the large subunit if you follow here.  And there will be two tRNAs,  at least two on each of these sites.  During the process, you are doing this amino acid  conjugation polymerization,  because one has to be the tRNA that is carrying  the elongating chain of your polypeptide or protein.  And the other has to be the additional amino acid  that will come in and extend the chain.  I hope that makes sense.  And that is what you're seeing, the two tRNAs.  But at the molecular level, if you zoom in further  at the amino acids that are making this peptide bond,  just remember proteins are linked to peptide bonds  or AMI bonds.  So this is essentially the peptides that you have formed.  So there are two tRNAs coded by red and green color  in two different sites known as B site and A site.  B site is essentially polymerizing site,  that is tRNA, the center site that I showed you  in the previous cartoon,  this tRNA carrying the already elongating polypeptide.  And essentially it's connected through this ester bond,  the oxygen with a carbonate here.  At the end of the tRNA,  remember tRNAs are nucleic acid, right?  So a CCA is always the last three base pairs  that will exist on the three prime end of any tRNA  that will be carrying the elongating polypeptide.  And the other tRNA in the site next door known as A site,  essentially A site is essentially starting site,  essentially is carrying a single amino acid  again at the three prime end of the tRNA sugar, right?  So look at the peptide bond formation based on this,  you start with ester bond,  you have the amino group,  this is a free amino group from your amino acid  that is conjugated to the tRNA.  You ultimately lead to the peptide bond here.  What do you think the thermodynamic drive here is?  Ester versus AMI bond, proteins are made of AMI bonds.  Why do you think that's the case?  Protein needs to be stable, right?  So why do you think proteins, amino acid linkages  are linked through AMI bonds, so-called peptide bonds,  instead of let's say a bonds  like this kind of oxygen-based ester bond?  It might say, I don't know,  if you guys in high school learn about resonance structure,  balancing weight, no?  Yes, essentially these bonds are more stable,  don't worry about it if you are not familiar with.  The peptide bonds are essentially one of most stable bonds  you'll find in the chemical world  and also in nature as well.  So essentially you gain energy,  essentially thermodynamically is downhill.  I should say, yeah, it's more stable,  so it's lower energy, not gain energy,  and you break the weaker bond.  So it's a thermodynamically downhill process,  but still to make this synthesis happen  at 10 amino acid,  some of the combination per second,  the process has to be highly efficient.  Why?  So there is a catalysis there  that lowers the process,  conditions take energy to speed up the reaction  and that is the job of the ribosome.  So I mentioned in the title,  the ribosome is actually a ribozyme.  Zymes sounds like enzyme, right?  So it was discovered now several decades ago  that the way the ribosome catalyzed  this peptide bond formation process  to make the proteins is not actually by protein catalysis,  but it is RNA catalysis.  Can you guys point to me  based on what I have discussed so far  of the ribosome architecture,  where is the RNA catalysis coming in?  Or where maybe the origin of RNA catalysis  instead of protein catalysis?  Most enzymes are proteins, right?  You guys know.  So ribosome being an enzyme  where catalysis is coming from RNA  instead of protein based mechanism,  it was a major discovery,  one of the most parts,  one of the areas.  So what do you think  what I have just discussed with ribosome architecture?  Maybe this gives you a hint as well.  I think I have an arrow here.  So what do you think?  What is going on here?  So what do you think this A is?  Where is it coming from?  So this, whatever lone pair is deprotonating  to extract the proton  to essentially activate this nitrogen  to be more active to actually go and attack this ester bond.  So where do you think this A2451 is coming from?  Essentially from one of the nitrogen on the adenine  based off the one of the ribosomal RNA  that is associated with this ribosome.  That's why ribosome is actually the ribosome.  And after that, not just ribosome,  the additional RNAs that people have found in nature,  especially in bacterial world  that actually serve as an enzyme  by RNA based enzyme  when this was discovered in a more broadly manner.  It was a big sort of hoo-ha back in those days  and inspired by these kinds of discoveries  synthetic biologists later on engineer  different RNA scaffold and structures  that will have enzymatic properties.  And we'll see that later  but this is a nature's example of RNA catalysis.  Right, ribosome we have it in the cell  and is doing it at the rate of 10 amino acid per second.  Okay.  All right.  So at the level of mRNA, right,  we have various components.  I mentioned about coding sequence and others,  but, and also there are five prime UTR  and three prime UTR and translated region.  I'd like to highlight the fact that both five prime  and three prime UTR are essential components  that really required for regulating the translation.  So where do you think that  if I have essentially structured like this  the coding sequences here,  but there's also, you see the UORF,  ORF stands for open reading frame,  but obviously coding sequence itself  is already one open reading frame, right?  Open reading frame is the sequence  that will allow you to code  for your protein sequence ultimately, right?  It is a sequence that is bracketed  between star codon and stop codon.  That is a definition of ORF.  People call us just ORF, right?  Not quite a lot of the rings,  but this is essentially ORF.  Do you see another ORF in this illustration?  Yes, we do, right?  There is another upstream ORF.  Given this, how many translation star size  that you could potentially have?  Yeah, it will be two very good.  And the star codon is always AUG, right?  That is how recognition happens.  So there should be AUG.  If I have a structure like that,  I should immediately know that  if I have upstream URORF  in addition to the canonical coding sequence ORF,  then you will have AUG here  and there has to be AUG here, right?  Without AUG, you just cannot start the translation  because AUG is a part of the ribosome docking site.  Ribosome needs to formally and physically interact  with the mRNA to initiate the translation.  And that sequence is known as COSAC sequence.  So if you have an mRNA that is missing COSAC sequence,  forget about protein translation  because ribosome will not be able to dock  on your eukaryotic mRNA  to initiate the protein translation, okay?  So AUG, which is a star codon,  I stress again,  it is actually a part of the COSAC sequence.  What do you think there's,  you see another thing, zip code, right?  Do you think zip code is part of the coding sequence,  part of the amino acid,  ultimate amino acid sequence coding  based on the structure,  based on the architecture?  Yes, no?  No, very good.  Why do you think that people call the sequence  that is somewhere within the three prime  untranslated region, UTR for untranslated region,  i.e. you don't make proteins from this sequence, right?  Why do you think people designate this as a zip code?  Yeah, absolutely.  Yeah, exactly.  Absolutely a great, fantastic answer.  So essentially the name tells the same, right?  I mean, essentially three prime UTR plays a big role  to control where should this mRNA be sent to, right?  So many times in the protein coding mRNAs,  you will see in the UTR region,  there's a zip code,  a sequence that will signal  for a particular mRNA binding protein,  hey, I need to go to, let's say ER,  I need to go to, I don't know,  my lysosome or whatever, right?  In the cellular framework, take me there.  So that's what the zip code sequence is for, right?  You can also see other features  we already talk about stabilization factors,  such as poly-A tail and then the five prime capping.  But I think the take home here is that the five prime UTR  that essentially in front of the coding sequence,  this is a stop codon, stop codon,  and then three prime UTR  at the downstream of the stop sequence,  they are very essential,  even though they don't lead to the protein translation,  because these sequences are not for coding,  but they're still essential  because you will have several signals  that are important for stability,  localization in the cell,  and also modulating the translation.  So five prime UTR does more for the translation modulation  in a large part because there are additional loop  and stem structure or stem loop structure of RNAs  that will interfere and regulate  whether ribosome can access to this architecture or not.  Imagine ribosome is pretty big,  it needs to go and find this COSAC sequence.  If these structures, 3D structures are in the way,  it will prevent your ribosome to do the translation.  Okay, so that's certainly a takeaway message from here.  Okay, so far so good for us.  Okay, very good.  So let's go through this table exercise, checking my time.  So as I said,  synth biologists are inspired by nature, right?  So it's a little small, I do apologize,  but to engage with you guys better,  it's better that all of these examples  are staying in a single slide.  That's why it's individual diagrams, a little small.  So maybe refer to your slides or print out.  What we're going to go through is  several natural systems  of so-called endogenous regulated allergies.  And then in the subsequent discussion,  we'll make use of these nature idea and nature's mechanism  to control cellular response output and decision-making.  So let's go through as an exercise,  while we're also learning the concept.  So there is a classic terminology  that synthetic biologists and bioengineers use,  transacting versus cis-acting.  Have you guys been discussed,  or have you guys been told about transacting and cis-acting?  Okay, so it's a simple thing.  Trans-means opposite cis-means on the same side, right?  So if something is cis-acting,  the ingenious system, either natural or ingenious design,  it is acting on a different strand of the,  you know, the target ion, or you know,  nucleic acid of sequence, that would be trans-acting.  If it is on the same strand,  without any, most of the nucleic acid would be double-trended.  If it is acting on the same strand,  it would be cis-acting.  So we're going to have to choose,  based on the illustration,  whether the natural system is cis-acting or trans-acting.  Okay, and of course, many times,  depending on what the context is,  typically the figure will tell you a lot.  The binding effect of your, you know,  of the small RNAs, or regulated RNAs,  to your target RNA,  could lead to either translation of regulation,  or translation inhibition, right?  Or either suppression or activation.  You also have to be able to consider those forms.  Because only with these concepts together,  you can actually get the actual output, right?  Because you don't want to end up designing something,  you want it to inhibit the translation,  and ultimately what you design ended up activating  or promoting the translation, that would be bad.  So there are two points we'll discuss  for each of the examples here.  First is trans-acting versus cis-acting.  Second is whether it leads to activation of translation,  or suppression of the translation, okay?  All right, so let's first start with the figure eight here.  So comment on differences between prokaryotes  and eukaryotes.  Well, there'll be some side questions  I want you guys to engage,  but beyond the two generic concepts I mentioned.  So the figure says the top one is prokaryotic,  and the second one is eukaryotic.  So the piece, the long piece that you're seeing here  obviously mRNA, just by looking at the figure,  many of you know that one is prokaryotic mRNA,  one is eukaryotic mRNA.  Can you guys tell me what each is happening  in this visualization?  Yeah, let's see.  Absolutely, so we learned last week as well as this week  that if eukaryotic mRNA,  before they can go to ribodome and translate,  they need to be processed and mature.  And this maturation process requires  this five prime capping, this sort of a square box,  and then poly a tail tagging.  So maybe on the footnote,  I mentioned five prime cap is a purple square.  And then a is, if you see multiple a's,  that's essentially poly a tail.  So that already tells you,  even if these labels are missing,  you know that the mRNA,  the one that I'm pointing,  must be d-caryotic mRNA.  Okay.  All right.  And then,  all right, what do you think?  I exposed the answer.  So my regulatory mRNA,  it is the red piece, what?  The small regulatory mRNA.  And that is these regulatory mRNAs  are present in nature,  bacteria as well as in the mammalian system.  So here it's essentially,  it operates in both prokaryotic and proprietary system.  They are called antisense RNA.  Antisense regulatory RNA,  is similar to antisense on the nuclear drugs  that you guys learned last week, right?  So they evolved the similar size  to the drugable RNAs, right?  So what do you say,  looking at the way that it's drawn,  is that transacting or cisacting?  Maybe you saw that, yeah.  Transacting, very good,  because my regulatory RNA is the red,  and then the target that I'm binding and affecting  is opposite strand, right?  So essentially it will be transacting.  And obviously you can perhaps predict  that this binding should repress translation.  Can you guys explain why that is?  Why is it that in both system binding,  as I've drawn here,  to be expected to suppress translation  rather than promoting?  Yeah.  Wouldn't it just stop the red?  Very good answer.  Yeah, exactly.  So you don't need to slice through,  through a stretch of the mRNA sequence,  and it processes,  it has a processivity in catalysis  and polymerization mechanisms,  simply slice it through,  a bit like Combiabelt,  into like many ribosome molecules line up on the mRNA,  and they slice through,  and that will be blocked.  So remember the green part is actually coding region, right?  So the coding region is blocked  because the regulatory RNA is bound on that green unit,  which is a ORF or coding sequence.  Okay, good.  So let's continue.  All right, so the second example  is still called small sRNA in prokaryotes.  So rival regulators are exclusively found in prokaryotes,  not in human,  but not in eukaryotic system.  They actually have a system where,  again, all these RNAs are relatively small units, right?  So you can see this additional piece here,  and also additional piece here.  They have this base pairing capability  that will act on the target mRNA.  One is acting outside of the coding region, right?  The other is adding  in the inner side of the coding region in green, right?  Are this acting or transacting for rival regulators?  Very good, excellent answer.  Transacting because still I'm acting on an opposite  or different strand, right?  Of what my regulatory RNA is.  And the statement here says that binding of the regulators  to their target may repress or activate translation.  Why is it that my cartoon here  is consistent with this statement?  Can someone tell me why?  So unlike, so the top mechanism,  antisense mechanism always suppress translation, right?  But when you get to rival regulators,  depending on their structure and base pairing signs,  some of them may suppress  and some of them may not suppress or activate, right?  So what do you think?  So let's look at this, the one that I'm pointing here.  So would this one be a suppressing translation or?  Yeah.  So that would be suppression for sure.  It will interfere with ribosome slightly, very good.  What about the top?  Confirmation?  Yeah.  Can access, yeah.  It looks like a very good answer.  It looks like maybe that, you know,  when this is in the way ribosome cannot perhaps, you know,  or not just ribosome, right?  We need to have other factors to come in  and initiate the translation that may not be as efficient  when this structure of the RNA is folded in this system,  right?  And it is getting in the way, perhaps sterically blocking  and that may lead to, you know,  if you essentially have this activation  or maybe enhancing the ribosome efficiency of translation.  Okay.  Very good.  So after we have gone through this while we're here,  we can take a break, but let's keep going.  Now we see the ribozyme.  Ribozyme are special name for catalytic RNA, right?  What you have seen so far,  antisense RNA regulators, they're not doing chemistry, right?  They're not catalysis, they're just regulating.  They're modulators, they're not catalysts.  Whereas when you get to ribozyme, they are enzyme.  RNA based enzyme, they are catalysts, right?  Biological catalysts or nucleic acid-based catalysts.  So you can see in the coding region  and a lot of structure in the five prime UTR, right?  So ribozymes typically adopt this 3D shape  to give them a fold that would have this catalytic activity.  And many of the ribozymes can do self-catalysis,  i.e. self-cleavage.  What you're seeing here is the mechanism where,  or illustration where cleavage actually happens  in this particular point.  What would you say in this particular illustration?  Is it acting in cis or trans?  It will be cis again, right?  Because the strands, sorry,  that breaks essentially the coding strand, right?  The strand that is carrying the green unit.  Again, apologize a bit small,  but hopefully you can view better in your slides.  So that will be the acting in.  All right.  So how about the activity in terms of the translation?  Maybe you can imagine that if you are actively cleaved  in the cis acting manner,  it will actually repress the translation.  Why?  Because you're going to lose some parts  of the five prime UTR.  As I mentioned, it is important for recruiting ribosome  and initiating a translation that will be lost  and that will lead to repression of translation.  Okay.  But if you do a cleavage on the trans acting manner,  i.e opposite strand, we cannot predict a priori.  What will be the effect on the translation, right?  It depends on the structure, additional context, et cetera.  Is that clear for you guys?  In the case of ribosome,  where whether it's going to be repressive or stimulatory.  Yeah, you have a question?  Okay.  Ribosome is a common name for RNAs  that have enzymatic properties.  So catalytic RNAs,  if you see any RNAs that have catalytic property,  you can call them as ribosome.  Ribosome is an example in nature that is a catalytic RNA  because it has the RNA components  that actually catalyzes the amino acid polymerization.  But ribosome itself is a machinery  that allows you to translate proteins,  i.e doing this polymerization.  I hope that makes sense.  Okay, excellent and good question.  All right.  So you guys get this cis-trans  because these kinds of things that you'll see there  and problem set, you'll see the design and application  in the final exam as well.  Okay.  So let's go through almost to the end before the break.  So in D, you see a class called riboswitches,  which already means you need to be able to turn on and off.  And typically what kind of system allows you to turn on  and off very rapid manner?  Are you talking about genetic system  or small molecule pharmacological system?  Rapid on and off typically is a small molecule system  because link binding, de-binding can be very rapid,  both in vivo and in the cell culture or in vitro system.  So indeed riboswitches are structured RNAs  that can regulate protein translation in the level  that they can be either activated or deactivated  by small molecule binding.  So if you look at this particular structure  I'm pointing here, they can also be ingenious  such that their regulation could be transcription  or translation.  But if you look at perhaps these example,  you can see that when the ligand is bound,  they have this structure on the RNA,  five prime UTR if you like, that is different  from the system when the ligand binding doesn't exist.  So when the ligand dissociates due to drug concentration  drops in vivo or whatnot, et cetera,  you will have a different structure of the resulting RNA  that will allow the transcription  or translation to take place.  So that allows you to modulate at the on off level  for ribosome accessibility on the coding strand.  So given this, are they cis-acting, trans-acting?  Typically, yeah.  This is very good.  So you can see from the figure as well,  the coding region is on the same strand  as where the structural changes are happening  by ligand binding, de-binding.  Okay, the last two, the first E, small interfering siRNA.  What do you think, cis-acting, trans-acting?  Trans-acting, excellent.  And suppression and deactivation, again,  which one would you expect?  This is happening in the coding region again,  most likely suppression, right, interfere with translation.  What about the last category, micro mRNA  that are very abundant in human cells as well,  so-called miRNA.  Trans-acting, excellent.  And binding represses translation.  Majorly, it represses translation,  but there are also systems these days being discovered  that may go the other way as well.  All right, so do you guys have any questions?  So these individual questions are essentially answering  what we have just discussed.  Maybe the last question I'll ask from my side is that  out of all of these various forms of nature's  regulatory RNA, so these are endogenous in organisms,  including humans, which specific type or types  is all obligatory and they get dated properly.  Yeah.  Ribosucious one, very good.  That's quite obvious, very good,  still very good, well done.  What else?  What other example that you see  that could be legit dated as well?  It's not so obvious.  So one obvious example is ribosucious.  On an off switch, I mean, you allow biasing data.  Although ABC and BNF, so very many of my software  that may be convertible on top by many enzymes.  It turns out that ribozymes,  they are essentially catalytic RNAs,  just like in enzyme, they have binding pockets and structures  are usually structures and shape.  Most of the time you can engineer small molecules  that will bind to ribozymes and allow the catalysis  to be either suppressed or activated.  So indeed ribozymes are also another category  of RNA-based regulators or RNA regulators  that allow ligand-gated control  that gives you an additional level of on-off control  that is otherwise not possible.  I hope you're enjoying so far.  So we'll take a break as usual for 10 minutes  and then come back and go with case studies in the next.  Thanks.  So, you know, friends versus cis-actin.  Actually, one of the students asked me a question  related to what we've learned last week.  It's good questions about these particular microRNAs.  Last, sorry, therapeutic strategies of ASOs,  if you recall last lecture and problems,  they can also target, and essentially the outcome  is typically downregulation of the protein level.  But remember, this interference,  here we said the binding repressed translation,  even if the binding is not on the coding region,  you could still interfere with a poly-A tail sequence  that I mentioned that is a recognition sequence  required for mRNA binding proteins to protect a given mRNA  and export them outside the nucleus to the cytoplasm  where they will meet ribosome, right?  So if this protection fails  because the regulatory mRNA in red gets in the way,  your mRNA will be degraded  before it reaches the cytoplasm point to do the translation.  So this kind of interference mechanism  could really be at the upstream  rather than active ribosome translation event, right?  Okay, so I just wanted to touch base  and make sure I clarified this.  Okay, so we're going to do a lot of interesting case examples.  Now, coming to the point where we get this understanding  from nature, how nature systems are operating  at the RNA regulatory level or RNA regulators,  and how we can leverage these insights  to better design and apply sort of engineer RNAs  that will allow us to control the protein translation  or gene regulation overall.  And there are two classes we'll focus on.  One is engineer RNAs,  and second or added a layer of complexity  of additional layer of control is ligand-gated system, right?  They will allow you to on-off modulate  in an on-off manner, more dynamic way, right?  So which type of natural regulatory RNAs  should this other class mimic?  So which, yeah, this one.  Yeah, the switch is going to be nice.  Very good answer.  So we know that ribosuches from the last slide  is ligand-gated, actually, I'm sorry.  And then indeed, sorry about this, yeah.  Indeed, ribosuches will be the classic category  that will allow ligand binding,  D binding to regulate the RNA structure.  And then of course, their interaction with their target RNA  and ribozymes, i.e. catalytic RNAs,  will also be a type of a class of RNA  that have 3D structure that will create ligand binding site.  So there will be two categories  under this particular class, right?  So in the first class, just with engineer RNAs,  no small molecule bits, there are three types,  but we won't cover the last type.  We'll cover the first type as two examples.  So engineer RNA transcriptional activators, right?  So let me throw in a question.  Why do you think that this is engineer?  I know we say biologists, we're always engineering,  but why do you think when we talk about  RNA transcriptional activators  or transcriptional factors that are RNA-based,  they need to be synthetic, right?  So what are transcriptional factors in general?  We talked about it quite a bit last week.  Transcriptional factors, yeah?  Yeah, they are proteins that bind to DNA  and regulate gene transcription, right?  And we want to achieve the same output, same mechanism,  but rather than engineering protein,  we're going to engineer RNA  to be able to achieve the same final output, right?  So saying that this is RNA-based,  transcriptional modulator or transcriptional factors,  they have to be engineered because in nature,  all transcriptional factors are proteins.  And I think that you already answered.  So this part is not as part of the course,  but there's advanced student who wants to learn.  I put the slide up, but we won't discuss this.  Let me kill these things, yeah, the floating thing.  Okay, all right.  So let's start with RNA-based transcriptional modulators.  Okay, fine.  The goal is to design RNA-based transcription factors, right?  To make you guys refresh from last week,  we learned about transcription factors in general.  They are proteins.  They always have two units, DNA binding unit, a DBD in blue,  and then transactivating unit, the unit in red.  And together,  they make up these protein-based transcription factors.  They bind to the target DNA  and regulate the gene activation, right?  Gene transcription.  This is very important types of proteins.  We're now going to achieve through the design of the RNA  to achieve the same level of output,  i.e. the RNAs that can mimic the property  of transcription factor proteins.  Okay, so the example I am showing you here  is a system design that is applicable in yeast,  So you can see the downstream output is kind of a platform  where if the gene transcription is successful,  you'll either express this either history gene  or the Lexi gene.  This is just a, you know, example.  It's not fixed to that system only.  Same Lexi, right, essentially is one of the downstream gene  that you can read out ultimately by a color,  a fluorometric or colorimetric assay.  History is again, a different type of gene  that allows you to screen for live-death output, right?  It's just allowing you to see whether  this particular gene activation happens or not, right?  The goal is to actually achieve this gene activation.  And there's always operator side that will be bound  by particular protein-based transcription factor.  In this case, LexA, this LexA bind to LexA operator side.  And the way that they come in a unit at the RNA level,  why so this is RNA structure,  that will allow you to sort of control  this downstream gene activation is through this design.  What people know is that in nature,  there is a protein RNA-based interaction  for a protein that is known as MS2 coat protein.  And protein, by definition, protein RNA interaction  means that if there's a protein,  there's a specific RNA unit that will bind  to that MS2 coat protein.  And those are so-called MS2 RNA hairpins.  And you can see at the level of engineered DNA construct,  those two units are incorporated, right?  So after transcription happened in the yeast,  you will get this RNA module, right?  And then you can see that because you install  this MS2 hairpin units in your coding DNA,  in your transcribed RNA,  you're going to carry the same units.  And there's a five prime, three prime, the usual thing,  and some sort of stem loop structure,  which is typical for 3D shape RNA.  And there is this variable region that you can play around  with various nucleotide sequence.  And then to screen for a specific sequence  that will make the system effective  at gene regulation downstream.  So the researchers vary these units.  And of course it's a selection round screening  because there's phenotypic screening, life and death.  So you can really clearly tell  if the gene activation happens,  if you're growing the yeast  in the histidine dropout media,  if this particular yeast strain doesn't function  without this metabolite,  you know that if the yeast cell survive,  that is because this gene activation is happening.  So this is a very high throughput phenotypic screening  because you drop out a particular metabolite  in your yeast growth media,  and you incorporate a genetic system  that upon activation of that system,  you will produce and not just lead that metabolite.  So you can rescue the dead phenotype.  So if you see a yeast cell activating from your screen,  you know that whatever variable sequence,  region that you have introduced,  must be a good RNA sequence  that will allow you to activate the histidine gene,  the histidine metabolite synthesis gene.  I hope that makes sense.  So that's the screening setup.  And they actually show that  by introducing this particular structure of the RNA,  they can modulate the LexA,  the transcription factor-based protein  together with this RNA system  to modulate this LexA operator downstream gene activation.  So let me ask you a question.  You see this two unit, right?  And we also saw in the natural system,  DNA binding and transactivating.  So what do you think?  Your RNA engineer system is here, right?  Together with the protein bit over there.  So what is this whole unit representing?  Is that representing DBD or TAD?  So remember, I mentioned that this whole engineer system  will sort of mimic the natural transcription factors  that we know and we learned from last week  that they are always proteins, two units together.  One is DBD, one is TAD.  Looking at this engineer system here,  what do you think your RNA module is representing?  Yeah, about?  Yeah, it will be TAD because it doesn't directly bind  the DNA binding side, right?  Transcription factor classic side,  but essentially it is modulating through the TAD.  And we know that TAD needs to approve other protein factors  that allow the transcription to be controlled, right?  Very good.  So not only they were able to successfully engineer  this whole RNA system,  but they also are able to make this variable region  to adopt a small molecule ligand.  And that's not molecule ligand, it's called TMR,  just the name, right?  So aptamer site you see in green and blue bits.  So aptamers at one point,  it's still very sort of a powerful and attractive system.  So aptamers are the name that is designated  for any nucleic acid that have a 3D shape and structure.  So you will see things like DNA aptamers, RNA aptamers.  You may wonder what you mean  and what they mean by that is simply 3D shape RNA  and DNA structures that are engineered  to bind small molecule ligands, right?  Or metabolites or drugs, et cetera, right?  So they introduced this aptamer unit as shown in blue  that will allow the small molecule ligand to bind  and depending on the ligand bound or unbound state,  the structure or the shape of this whole modular RNA  may be altered and that conformational changes  may affect the gene regulation up there, right?  So that gives you an additional layer of control  to achieve the output if you like, not only on and off,  but you could also vary the magnitude  of a gene upregulation.  Okay, I hope that this example case study,  you guys all understand well, right?  Okay, good.  Yeah, purpose benefit of TMR aptamer site,  obviously, what will be the benefit here?  So if you see this system, let's say it works well  and it activates to transcribe this gene,  would you say this system is dynamic or constitutive?  Which one?  Constitutive mean always, if I say constitutively on  or constitutively off that is permanently on or off, right?  Based on what I've just explained,  this will be a system that is permanently activated, right?  By introducing this additional aptamer unit  that could have a ligand binding, D binding,  you could have this in a more dynamic system, right?  And also you could potentially modulate the level of,  how high is, when you say the system activations happen,  how high is the activation  or what is the magnitude of activation?  So you can also tune perhaps the constitutively  on and off state of the design.  Okay, so I'm gonna go to the second case study,  what is this essentially about post-transcriptional  gene expression control through the use of  rival regulators, right?  I mentioned rival regulators are inspired  by prokaryotic system.  So the goal here is to create a modular post-transcriptional  regulatory system that could be integrated  into biological networks and implemented  with any kind of promoter or gene, right?  So generality is a big thing in synthetic biology, right?  You just don't want to engineer one thing  and only works on one system, right?  So modular approach, general approach  is always a big win here.  So I'm showing you here the system,  where is this equalized system,  prokaryotic gene expression, right?  So you have a DNA construct to start with,  and there's a promoter system,  any gene, every gene needs to have a promoter,  and then ribosome binding site, RBS,  ribosome docking site, and followed by GFP, right?  And DNA undergoes transcription to have this RNA translation,  RNA synthesis, and followed by protein GFP translation.  And the star site AMG you can see here, right?  So the question is how do we modulate  the expression level of GFP using  the nature-inspired rival regulator, right?  And how do we affect that?  Not at the promoter level, at the DNA level  like we saw in the previous example,  but this time at the post-transcriptional level.  What do I mean by post-transcriptional level?  If I say I'm engineering a system  that will perturb my target system  at the post-transcriptional level,  where am I affecting?  Am I affecting at the DNA transcription step  or am I affecting at the translation step?  Exactly in the post-transcriptional step.  So at the mRNA level,  because you have finished the transcription,  that's what it means by post-transcription,  and you will be affecting at the mRNA level.  Excellent answer.  So indeed that is what we're going to design here.  You guys have questions?  Yeah, okay.  Yeah, so then you interfere at the mRNA level.  You already know that's where your design system will be.  And let's take a look at the mRNA  that you will be engineering.  Obviously, any mRNA you introduce to the cell  starts from the DNA level, right?  Typically, because DNA is a more stable  and you can just introduce DNA to the cell or organism  and it stays with a much longer lifetime.  So indeed you start with a DNA system,  but ultimately at the RNA level  is what you want to interfere.  So starting from this dDNA level,  you can see various units, right?  So the ribosome binding site and GFP  is the same as the system on the left.  But what you introduce is the so-called CR  or cis repressed system, right?  And then you are transcribed to this DNA  to get to this RNA with the same pieces.  And what happens here is at the RNA level,  you started to have the base pairing.  So the way that you design your DNA sequence  is such that there will be original base pairing  between the unit in orange  and then the ribosome binding site.  So in the structure RNA, once the transcription happen,  I mean, post-transcriptional level,  you will have this particular structure.  I believe, right?  Because there's an internal base pairing.  But obviously at this stage, what do you think?  Will I get the GFP gene translation or protein translation?  Yes, no?  No, right?  I think I've lost it.  Absolutely, this ribosome binding site  is absolutely essential.  If this site is not accessible by the ribosome,  you're not going to get the gene protein translation.  So if this site is internally blocked by base pairing  through this cis repressor unit,  you guys now know why it is called cis repressor  rather than trans repressor, right?  Because it's acting on the same coding strength.  So it will be the cis repressor.  And you can see essentially you get cis repression,  GFP should not be translated by this blood error.  But what happens?  How do you want to have a modular system  that you can turn it on?  So you bring in additional component  still interfering at the mRNA level,  but still you have to bring in at the DNA level  because DNA is more stable  and you can introduce to the cell.  So you have so-called tRNA.  Here, essentially, why do you guys think that now  I call this as trans?  So if you look at the structure,  ultimately where it is acting,  why do you think that this is essentially trans acting?  Yeah, so essentially it is acting on the opposite strand  compared to this blockages from the same strand right here.  This orange unit that you come along,  you bring along in your tRNA engineer system,  go and essentially bind and interfere the CR unit  and to compete out the binding  such that structurally it is unfolded.  Now the RBS or the blue unit is now exposed.  All right, comment on the design of the sequence,  CR sequence.  So if you're going to design this cis acting repressor,  you know that whatever unit you introduce  must have base pairing capability  to the ribosome binding site.  So that will be the design that you introduce.  You know that this has to be a complimentary sequence  to the ribosome binding sequence  or COSAC sequence on the cis part.  What is the effect of CR RNA on GFP expression?  We know that essentially it should be a suppressor,  likewise it being referred to as cis repressor.  We've already talked about it.  What is the effect of introducing  a second independent construct shown here?  Or is it being referred to as trans activated?  We have discussed that.  Second independent construct allows you  to relieve this blockage,  allows a ribosome to duck through this blue unit  and then translate the EGFP sequence.  Indicate positive and negative regulatory steps.  Comment on the modularity of the system.  Yes, brilliant answer.  Remember, right?  Every engineer system we'd like to achieve generality.  So this is just proof of concept shown with GFP.  So if you think about it,  this AUG and RBS exist in any gene  that we need to translate.  So being able to show that this unit  that you can bring along through the DNA sequence upstream  can ultimately block this translation star site  that could essentially be generally very applicable.  And that will be a negative regulatory step.  And this will be a positive regulatory step, right?  Because you release this blockage of the RBS.  I think I have one last question.  What do you think you're going to apply this system  in the real world, in the cell system?  And let's say you want to do a control  in a repressive way and then followed by  in an activating manner.  What would be the practical aspect you need to consider?  Remember, this is essentially intramolecular-based parrot.  Well, intramolecular systems are always,  they don't have to worry about concentration  or how much you have, right?  And essentially the internal intrinsic system.  And this is intramolecular,  i.e. two things meaning to come together.  So if you think about it,  this should be much easily achievable than this.  So, but we need to achieve this, right?  So what do you think you need to do  if you're going to design this in real-world system?  Very, very high-pitched implementation  that you're in to-  Absolutely, exactly.  So you really need to throw in both bucket loads  of this construct over this construct  so that intramolecular binding can override  to release the intramolecular-based parrot.  So yes, yeah.  Absolutely, a very good thought.  I see what you're saying.  So the student over there is saying  that you could also make this binding less time, right?  So perhaps rather than base pairing at every single base,  you can have less kind of a relaxed base pairing system.  You could have that.  So there's a bit kind of a loose system.  So that's one way to bypass.  You also have a-  I was just checking, are these promoters usable?  I mean, yes, I don't see how the parent can control  this system.  Are these promoters?  I missed the point.  Inducible.  Inducible, yeah, okay.  You could also design the system like TET ON, TET OFF,  the system that we learned earlier on.  But in this particular example,  they are kind of promoted that is kind of a constitutive link.  So that means that they really have to introduce  this particular construct at very high concentration,  which you can do in real world,  to express many more copies of this tRNA  compared to the GFP expressing construct.  That's why we see that.  So what's the difference between all of these risks  and most additional problems that we've done in the-  Yes, exactly.  So essentially to show the proof of concept on the GFP,  but ultimately you may want to do it in another gene  that is important for cell deliberation, for instance.  So this is a proof of concept example  where you could introduce additional units.  So for instance, this kind of gene sequence introduction  could also be done in the native gene level,  which is the blue cell crystal  will run and mix plants and all.  So in that way, this could be an endogenous system as well.  But obviously the synthetic system has to be based.  You wanted expression of the protein at the beginning  and then you constructed the label  to have the same protein as the-  No, so essentially if you essentially have this sequence,  essentially have this sequence, there will not be any gene expression in the beginning.  So this will be completely off. This may be confusing. This is essentially the system that  is completely non-controlled. So you'll have a GFE expression as is shown if you introduce this  construct. But you can essentially put a unit that is a CR, that your cyst repressor unit,  and on the piece of RNA that you want to target. So they'll be self-regulating internally.  So up to this point, you'll have no GFE expression at all because translation is blocked. And then  you can turn on the GFE expression on demand by introducing this synthetic TAR.  So synthetic biology is a broader goal. It's really for activation,  switching it off or modulating a lot of on-hop design. So suddenly this bonus here is to be  able to show that you could use this kind of modular system to control the gene expression.  Here it's suddenly turning on and turning off. All right, any questions or comments?  Good, okay. Very good. So you need to at least be aware that you would need much more of this  because it's an intermolecular binding system to override. So I won't go through the three,  but I did put in answers for those students who are really more into these kind of engineer  system. Feel free to read it in your own time. Come to me if you have questions or when I'm  paying in the TA session. Okay, so we'll go to the ligand-gated design here. And this slide should  be very easy to understand from the illustration here. You can see that now you know what eptamars  are, right? So they can bind target molecules, either DNA or RNA, single-stranded typically.  And when the ligand is bound versus ligand is not bound, you could actually get the confirmation of  the RNA that you're talking about, this epimides, right? So when the ligand is not bound, the  structure is quite relaxed and kind of it can go through a different conformational equilibrium  shown by this arrow. And obviously the ligand can only bind to a portion of this conformational  equilibrium shown here on the top structure. And once the ligand is bound, you actually kill the  conformational equilibrium because ligand binding stabilizes the structure, right? So in general,  ligand binding always stabilizes any biomolecular structure, whether protein or RNA or DNA,  because it's a thermodynamic energy gain, right? So just like in how ligand binding stabilizes  protein and increases the melting temperature, the same for RNA.  Another sort of example here, again, in the conformational equilibrium you could have here  with a stem loop structure of the RNA together with just a non-structured RNA unit. They can  essentially go through internal base pairing like shown here. But as soon as you actually have perhaps  the structure where the particular RNA binding protein can bind, you will only have this system  here. And so essentially this will be a ribosome, for instance, you need to have this structure  where ribosome is accessible. So if your RNA is all in a complex situation, you won't be able to  initiate the translation here. But how you may control this system such that a translation  initiation can happen, so you can do that by so-called transducer module where you would  include so-called connected sequence or so-called transducer. You will see that in the problem set  this week as well. And you could essentially have a system where through this transducer changes in  conformational equilibrium, when the ligand is bound in this loop unit of this stem loop structure  of the RNA, you can see that this self-folding stage or so-called off state, i.e. where ribosome  site cannot be accessible, is no longer achievable when the ligand binding happens. In that case,  this structure is exposed to the ribosome and you can actually initiate translation. That is what  we mean here by on state. So this is to me is a very easy way to understand using this illustration.  Comment on the major added value of ligand-guided RNA switches compared to the previous systems we  showed where there may not be ligand binding possibility. Yes, absolutely. It can be,  so essentially you block this off state. For instance, you will always have ribosome accessible,  so that allows you to have this dynamic system of control on and off and it's more kind of  reversible system rather than permanent system of either permanent on or permanent off.  Imagine, even if I go back to the previous case, when we talk about it, introducing this  is not going to be, you're not talking about five minutes to 30 minutes. Because to introduce  this, you have to start with DNA or RNA, or at least in most cases is DNA, you have to go through  transcription. Systems that are built on nucleic acids always going to take more time because you  need to start from DNA transcription to RNA. Of course, the permeability is also a big issue  in the whole cell or animal context, whereas small molecules, you can make them to be permeable  reasonably easily. You have this dynamic system of typically minute time scale  with small molecule system. That will be a major advantage, dynamic ability to switch on switch off.  We'll go through two more examples here. Let's think about and engineer ribosome mediated RNA  switches. The objective here is to actually engineer ligand responsive ribozymes to regulate  mRNA translation. You know that ribozymes are catalytic RNAs and some of the classes of ribozymes  can self cleave. What we're doing here is the self cleaving ribozymes. Remember, not all ribozymes  are self cleaving, but the class of ribozymes that can do self cleavage and that's exemplified here.  The same here starting from CMD promoter, this constitutive promoter, you have this construct  that express Lexi protein starting from Lexi DNA. If you see the mRNA like this,  you know that that's eukaryotic or prokaryotic RNA. It will be eukaryotic. What they did here  is that you have this poly-A tail, 5 prime cap typical eukaryotic RNA. This is a coding region  with a star site AUG. And of course, you also have in the pre mRNA, you may also have an intron.  And what they did was in these asterisk, sort of a marked sequence region,  shown here, both at the 3 prime UTR and various parts in the 5 prime UTR,  they introduced this structure ribozymes. And in that way, they screen whether the Lexi protein  will be expressed or not. But remember, this particular ribozymes they use can self cleave.  If the ribozymes self cleave, what do you think will happen to the ultimate protein translation?  And in order to achieve that, where should the ribozymes be? Should it be in the 5 prime UTR?  Or maybe that we can also have it in 3 prime UTR. I'd like to hear some feedback.  What do you think? We're engineering a system that will install self-cleaving ribozymes  as a way to perhaps control the protein translation. Where would you place that  structure unit of the ribozymes? Will that be in 5 prime UTR, 3 prime UTR?  Yeah, more chance to see perturbation if you put it in 5 prime UTR. And that is because  you won't be able to guess.  In the mRNA level, that's where the ribozome serving site will be. Right. Very good. So if  cleavage happens, so ribozymes are RME, they still have rejected the RNA state  before they become catalytically active and self cleaved. So that means that they cannot be acting  on the DNA level. Because they are structured RNA and their structure and their activity will  only be achieved when they become RME, when they are not at the DNA level. So that's the first  thing you need to appreciate. And when they're active, they start with self cleaved. Obviously,  if you want to interfere in the maximum possible level, you're better off putting that ribozymes  unit on the 5 prime UTR region, so that if it's cleaved, the whole thing will be destroyed.  Whereas if you put it in the 3 prime UTR, maybe even if it is cleaved, protein translation may  still happen. But sometimes you can also be successful, still be successful by putting  on the 3 prime UTR. Can you guys tell me why that may be the case? So the reason will be  very different. Yeah, if you want to. Poly-Ate. Yeah, very good. So if you remove the poly-Ate  in any nucleotide mRNA, you're in danger of losing the mRNA. Why? Because remember, poly-Ate unit is  essential in the nature to allow this mRNA stability and mRNA export from the nuclear  to the cytoplasm before mRNA translation to protein. Why? So if you have the self cleaving  ribozyme installed in this region, you may lose the poly-Ate and losing of the poly-Ate  may affect the ultimate protein translation if the mRNA has disappeared from the cell due to the lack  of poly-Ate. Disappear means it gets cleaved and because it's unstable, it gets degraded. Okay,  good. All right. Is this ribozyme cis or transacting by the way?  Cis acting, yeah, good. All right, so one additional thing they do is that they actually  screen the small molecule signals that will bind to the ribozyme active site and in that way  prevent the ribozyme or deactivate the ribozyme so it can no longer self cleave so you could get the  protein expression. That's easy to understand. So before we break up, let's think about the  very last example where you essentially use arrival regulators or so-called anti-switches.  Right, so here's still ligand control context. So I'm showing you off-state and then on-state,  on the left and the right. Right, a little bit small but you could expand perhaps in your screen.  So again, this time in yeast as well. So there is this ligand responsive off-state, off-anti-switch  that you introduce and engineer in the system. Essentially transcription happens and essentially  you get to the unit of this particular structure. Right, and you engineer and screen small molecules  and realize that you can have a system where when this ligand bound, the RNA that you design is this  state but whereas there is no ligand bound RNA that you design is on the state on the left.  And based on the confirmation of the design engineer arrival regulator, you'll be able to  either interrupt the ribosome translation or you'll be able to not interrupt here.  But there is one additional unit that they introduced, the blue units, known as ribozyme  pruning. And on the state in terms of on-anti-switches, very similar,  there's essentially different structures that you get from a kind of a design system. And depending  on the confirmation of the ligand bound versus non-bound state, you'll be able to interrupt the  fluorescent protein, YFB protein translation. So let's go through a few questions and we'll  break off. So anti-switches typically have two domains, antisense affected domain and ribosome  like aptameric domain. Can you guys find out these domains and how they coordinate  the control translation? So which one is a ribosome like aptameric domain?  The stem loop structure, right? Because?  Where the ligand binding happens. Yeah, brilliant answer. So we know that aptamers,  essentially are structure shaped DNA or RNA. We know what ribosome switch is, is a ligand  gauges system. So clearly ribosome switch like aptameric domain has to be the stem loop structure.  That means process of elimination. The other thing has to be antisense affected domain. And  indeed this red part is that what interacts to block the ribosome binding. Very good.  And how they coordinate to control translation. Comment on anti-switch binding site based on  shown figure, what general challenges could you imagine from this requirement?  So remember, these are targeting endogenous RNA. So just shown by this a GFP. But based on my  earlier slides in the earlier part, what do endogenous RNA is, especially at the five prime UGR?  What do they typically have? We show general architecture. Yeah, five prime cap as well as  additional 3D shape structures, right? Typically stem loop structure. If you go back to a good  answer, so if you go back to the earlier slides where I gave you the zip code and all the other  things, you will see that in the five prime UGR typically is really dense region with a lot of 3D  shape structure. And if you are designing anything to target this kind of region, that could be  difficult. Remember, they are not targeting the actual green region also, but they're targeting  this kind of a bridging unit, the bridge between the five prime UGR, non-coding upstream region,  and then the star site, ABG star site. So in doing so, you have to be very careful. That may be the  challenge because there could be complex endogenous RNA structures that have this 3D shape structure.  There could be complex endogenous RNA structures that have this 3D scaffold that may make it very  difficult to achieve binding. So what is the purpose of HHRZ sequence at the five prime and  three prime that you see in each case? So HHRZ, if you read the footnote, it is the hammerhead  ribozyme. It's a type of class of ribozyme that is self-cleaving. So why do you think that these  engineers install these additional units? Yeah. Yes, there's one possibility. And also,  even in the absence of ligand, these hammerhead ribozyme are self-cleaving ribozyme. That means  that if you think about as soon as they get to the RNA state, they're becoming catalytic,  they are becoming RNA catalysts, and they self-cleave. That means whatever additional  base pairs that you may have would be cleaved off. And that is what they mean by ribozyme  through me, through me is cutting, you know, like cutting your hedges in the garden, for instance.  And that essentially cutting may give you a more finer sequence to make sure that what you want as  a ligand responsive anti-switch unit is really the precise sequences because the HHRZ domain should  be cleaved off once the system gets to the RNA level. Okay. I'd like to also clarify,  I know Alex, one of our teaching fellows has done this with email, but some of you may be wondering  what is going on with the problems as well. We give a timeline so that we want you guys to,  you know, at least have tackled these problems as at least for our three a week series. And  you will be very generous. So if you hand in on time by the designated timeline, you can see on  the Moodle, we will give 1% bonus point. 1% is still a big thing in my view because, you know,  I mean, it's for the whole course. So for the three lectures, you are getting essentially,  technically speaking, altogether 3% one out of the 100% total. It is not necessary to get all  the answers correct in order to gain points. So the really, the spirit here is to be able  to participate. And if you have questions, you know, bring to the Thursday session and in that  way, you really will, in that way, really help you, you know, be ready for the final exam and  you'll ace through it. Okay. I think we are good. So on time, any questions here?  Unfortunately, I only have three weeks time for this course and it's always sad for me. We have  one more class to see you guys, but there will be a CRISPR and another sort of a, you know,  game-changing systems that you'll learn next week. And come to Thursday sessions, right?  If you have questions, bring to us. And if you understand what we discuss here,  you will find problems as really quite straightforward. And I look forward to seeing  you guys again next Monday. Okay. Good. Have a good, nice rest of the week.\""
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "detected_frames = pd.read_csv(\"output/detected_frames.txt\", sep=\" \", header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>frame:0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pts:1321984</td>\n",
       "      <td>pts_time:103.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lavfi.scene_score=0.140596</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>frame:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pts:1323520</td>\n",
       "      <td>pts_time:103.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lavfi.scene_score=0.221395</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>frame:2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pts:1325568</td>\n",
       "      <td>pts_time:103.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>lavfi.scene_score=0.145158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>frame:31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pts:61705216</td>\n",
       "      <td>pts_time:4820.72</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>lavfi.scene_score=0.107849</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>frame:32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pts:70086656</td>\n",
       "      <td>pts_time:5475.52</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>lavfi.scene_score=0.172250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>66 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             0   1   2             3                 4   \n",
       "0                      frame:0 NaN NaN           NaN       pts:1321984  \\\n",
       "1   lavfi.scene_score=0.140596 NaN NaN           NaN               NaN   \n",
       "2                      frame:1 NaN NaN           NaN       pts:1323520   \n",
       "3   lavfi.scene_score=0.221395 NaN NaN           NaN               NaN   \n",
       "4                      frame:2 NaN NaN           NaN       pts:1325568   \n",
       "..                         ...  ..  ..           ...               ...   \n",
       "61  lavfi.scene_score=0.145158 NaN NaN           NaN               NaN   \n",
       "62                    frame:31 NaN NaN  pts:61705216  pts_time:4820.72   \n",
       "63  lavfi.scene_score=0.107849 NaN NaN           NaN               NaN   \n",
       "64                    frame:32 NaN NaN  pts:70086656  pts_time:5475.52   \n",
       "65  lavfi.scene_score=0.172250 NaN NaN           NaN               NaN   \n",
       "\n",
       "                  5  \n",
       "0   pts_time:103.28  \n",
       "1               NaN  \n",
       "2    pts_time:103.4  \n",
       "3               NaN  \n",
       "4   pts_time:103.56  \n",
       "..              ...  \n",
       "61              NaN  \n",
       "62              NaN  \n",
       "63              NaN  \n",
       "64              NaN  \n",
       "65              NaN  \n",
       "\n",
       "[66 rows x 6 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "detected_frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>frame:0</td>\n",
       "      <td>pts_time:103.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lavfi.scene_score=0.140596</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>frame:1</td>\n",
       "      <td>pts_time:103.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lavfi.scene_score=0.221395</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>frame:2</td>\n",
       "      <td>pts_time:103.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>lavfi.scene_score=0.145158</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>frame:31</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>lavfi.scene_score=0.107849</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>frame:32</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>lavfi.scene_score=0.172250</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>66 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             0                5\n",
       "0                      frame:0  pts_time:103.28\n",
       "1   lavfi.scene_score=0.140596              NaN\n",
       "2                      frame:1   pts_time:103.4\n",
       "3   lavfi.scene_score=0.221395              NaN\n",
       "4                      frame:2  pts_time:103.56\n",
       "..                         ...              ...\n",
       "61  lavfi.scene_score=0.145158              NaN\n",
       "62                    frame:31              NaN\n",
       "63  lavfi.scene_score=0.107849              NaN\n",
       "64                    frame:32              NaN\n",
       "65  lavfi.scene_score=0.172250              NaN\n",
       "\n",
       "[66 rows x 2 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "detected_frames.drop(columns=[1, 2, 3, 4])  # , inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Frame Numbers:\n",
      "['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32']\n",
      "\n",
      "pts_time Numbers:\n",
      "['103', '103', '103', '103', '103', '104', '104', '105', '146', '261', '383', '731', '985', '1239', '1341', '1374', '1706', '2000', '2002', '2008', '3013', '3018', '3092', '3282', '3807', '4542', '4554', '4769', '4770', '4803', '4805', '4820', '5475']\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "# Read the text file\n",
    "with open(\"output/detected_frames.txt\", \"r\") as file:\n",
    "    content = file.read()\n",
    "\n",
    "# Extract frame numbers and pts_time numbers using regular expressions\n",
    "frame_numbers = re.findall(r\"frame:(\\d+)\", content)\n",
    "# pts_time_numbers = re.findall(r'pts_time:(\\d+\\.\\d+)', content)\n",
    "pts_time_numbers = re.findall(r\"pts_time:(\\d+)\", content)\n",
    "\n",
    "# Print the extracted numbers\n",
    "print(\"Frame Numbers:\")\n",
    "print(frame_numbers)\n",
    "\n",
    "print(\"\\npts_time Numbers:\")\n",
    "print(pts_time_numbers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert the lists to dictionariers\n",
    "frame_dict = dict(zip(frame_numbers, pts_time_numbers))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33\n",
      "33\n",
      "33\n"
     ]
    }
   ],
   "source": [
    "print(len(frame_dict))\n",
    "print(len(frame_numbers))\n",
    "print(len(pts_time_numbers))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'0': '103.28',\n",
       " '1': '103.4',\n",
       " '2': '103.56',\n",
       " '3': '103.68',\n",
       " '4': '103.84',\n",
       " '5': '104.16',\n",
       " '6': '105.04',\n",
       " '7': '146.36',\n",
       " '8': '261.28',\n",
       " '9': '383.72',\n",
       " '10': '985.32',\n",
       " '11': '1239.04',\n",
       " '12': '1341.68',\n",
       " '13': '1374.44',\n",
       " '14': '1706.2',\n",
       " '15': '2000.64',\n",
       " '16': '2002.12',\n",
       " '17': '2008.44',\n",
       " '18': '3018.6',\n",
       " '19': '3092.6',\n",
       " '20': '3282.36',\n",
       " '21': '3807.24',\n",
       " '22': '4554.12',\n",
       " '23': '4769.8',\n",
       " '24': '4770.48',\n",
       " '25': '4803.88',\n",
       " '26': '4805.24',\n",
       " '27': '4820.72',\n",
       " '28': '5475.52'}"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frame_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "speaker-notes",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
